<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Robust Outlier-aware Estimation of Composition and Heterogeneity for Single-cell Data • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Robust Outlier-aware Estimation of Composition and Heterogeneity for Single-cell Data">
<meta property="og:description" content="A robust and outlier-aware method for testing differential tissue composition from single-cell data. This model can infer changes in tissue composition and heterogeneity, and can produce realistic data simulations based on any existing dataset. This model can also transfer knowledge from a large set of integrated datasets to increase accuracy further.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the sccomp package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<p>sccomp - Outlier-aware and count-based compositional analysis of single-cell data ================</p>
<!-- badges: start -->

<!-- badges: end -->
<p><a href="https://www.youtube.com/watch?v=R_lt58We9nA&amp;ab_channel=RConsortium" target="_blank" class="external-link"> <img src="https://img.youtube.com/vi/R_lt58We9nA/mqdefault.jpg" alt="Watch the video" width="280" height="180" border="10"></a></p>
<div class="section level1">
<div class="page-header"><h1 id="section">
<img src="inst/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h1></div>
<p><code>sccomp</code> tests differences in cell type proportions from single-cell data. It is robust against outliers, it models continuous and discrete factors, and capable of random-effect/intercept modelling.</p>
<p>Please cite <a href="https://www.pnas.org/doi/full/10.1073/pnas.2203828120" class="external-link">PNAS - sccomp: Robust differential composition and variability analysis for single-cell data</a></p>
<div class="section level2">
<h2 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h2>
<ul>
<li>Complex linear models with continuous and categorical covariates</li>
<li>Multilevel modelling, with population fixed and random effects/intercept</li>
<li>Modelling data from counts</li>
<li>Testing differences in cell-type proportionality</li>
<li>Testing differences in cell-type specific variability</li>
<li>Cell-type information share for variability adaptive shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code>sccomp_estimate</code></td>
<td>Fit the model onto the data, and estimate the coefficients</td>
</tr>
<tr>
<td><code>sccomp_remove_outliers</code></td>
<td>Identify outliers probabilistically based on the model fit, and exclude them from the estimation</td>
</tr>
<tr>
<td><code>sccomp_test</code></td>
<td>Calculate the probability that the coefficients are outside the H0 interval (i.e. test_composition_above_logit_fold_change)</td>
</tr>
<tr>
<td><code>sccomp_replicate</code></td>
<td>Simulate data from the model, or part of the model</td>
</tr>
<tr>
<td><code>sccomp_predict</code></td>
<td>Predicts proportions, based on the mode, or part of the model</td>
</tr>
<tr>
<td><code>sccomp_remove_unwanted_variation</code></td>
<td>Removes the variability for unwanted factors</td>
</tr>
<tr>
<td><code>plot</code></td>
<td>Plors summary plots to asses significance</td>
</tr>
</tbody>
</table>
</div>
<div class="section level1">
<h1 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h1>
<p><code>sccomp</code> can model changes in composition and variability. By default, the formula for variability is either <code>~1</code>, which assumes that the cell-group variability is independent of any covariate or <code>~ factor_of_interest</code>, which assumes that the model is dependent on the factor of interest only. The variability model must be a subset of the model for composition.</p>
<div class="section level2">
<h2 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h2>
<p>Of the output table, the estimate columns start with the prefix <code>c_</code> indicate <code>composition</code>, or with <code>v_</code> indicate <code>variability</code> (when formula_variability is set).</p>
<div class="section level3">
<h3 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">single_cell_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Here you see the results of the fit, the effects of the factor on composition and variability. You also can see the uncertainty around those effects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 72 × 18[39m</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="do">##    cell_group parameter  factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m      [3m[90m&lt;chr&gt;[39m[23m      [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="do">## [90m 1[39m B1         (Intercep… [31mNA[39m      0.882     1.12   1.34   0   [90m [39m   0   [90m [39m     [4m4[24m406.</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="do">## [90m 2[39m B1         typecancer type   -[31m1[39m[31m.[39m[31m17[39m     -[31m0[39m[31m.[39m[31m761[39m -[31m0[39m[31m.[39m[31m362[39m  5.00[90me[39m[31m-4[39m 1.07[90me[39m[31m-4[39m   [4m4[24m311.</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="do">## [90m 3[39m B2         (Intercep… [31mNA[39m      0.415     0.707  0.987  0   [90m [39m   0   [90m [39m     [4m5[24m063.</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="do">## [90m 4[39m B2         typecancer type   -[31m1[39m[31m.[39m[31m22[39m     -[31m0[39m[31m.[39m[31m724[39m -[31m0[39m[31m.[39m[31m259[39m  3.  [90me[39m[31m-3[39m 6.11[90me[39m[31m-4[39m   [4m4[24m277.</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="do">## [90m 5[39m B3         (Intercep… [31mNA[39m     -[31m0[39m[31m.[39m[31m676[39m    -[31m0[39m[31m.[39m[31m388[39m -[31m0[39m[31m.[39m[31m110[39m  2.00[90me[39m[31m-2[39m 1.52[90me[39m[31m-3[39m   [4m5[24m519.</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="do">## [90m 6[39m B3         typecancer type   -[31m0[39m[31m.[39m[31m732[39m    -[31m0[39m[31m.[39m[31m316[39m  0.088[4m2[24m 1.45[90me[39m[31m-1[39m 4.70[90me[39m[31m-2[39m   [4m4[24m677.</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="do">## [90m 7[39m BM         (Intercep… [31mNA[39m     -[31m1[39m[31m.[39m[31m31[39m     -[31m1[39m[31m.[39m[31m0[39m[31m3[39m  -[31m0[39m[31m.[39m[31m738[39m  0   [90m [39m   0   [90m [39m     [4m5[24m619.</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="do">## [90m 8[39m BM         typecancer type   -[31m0[39m[31m.[39m[31m742[39m    -[31m0[39m[31m.[39m[31m319[39m  0.073[4m0[24m 1.36[90me[39m[31m-1[39m 4.26[90me[39m[31m-2[39m   [4m5[24m845.</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="do">## [90m 9[39m CD4 1      (Intercep… [31mNA[39m      0.079[4m1[24m    0.307  0.523  3.42[90me[39m[31m-2[39m 3.60[90me[39m[31m-3[39m   [4m4[24m435.</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="do">## [90m10[39m CD4 1      typecancer type   -[31m0[39m[31m.[39m[31m103[39m     0.183  0.469  2.88[90me[39m[31m-1[39m 6.81[90me[39m[31m-2[39m   [4m3[24m753.</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="do">## [90m# ℹ 62 more rows[39m</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="do">## [90m# ℹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,[39m</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="do">## [90m#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,[39m</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="do">## [90m#   count_data &lt;list&gt;[39m</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="an-aid-to-result-interpretation-and-communication">An aid to result interpretation and communication<a class="anchor" aria-label="anchor" href="#an-aid-to-result-interpretation-and-communication"></a>
</h2>
<p>The estimated effects are expressed in the unconstrained space of the parameters. Similarly, to differential expression analysis that express change in terms of log fold change. However, for differences, in proportion, logit foold change must be used. This measure is harder to interpret and understand.</p>
<p>Therefore, we provide a more intuitive proportion, full change, that can be easier understood. However, these cannot be used to infer significance (use sccomp_test() instead), and a lot of care must be taken given the nonlinearity of these measure (1 fold increase from 0.0001 to 0.0002 carried a different weight that 1 fold increase from 0.4 to 0.8).</p>
<p>From your estimates, you can state which effects you are interested about (this can be a part of the full model, in case you want to not consider unwanted effects), and the two points you would like to compare.</p>
<p>In case of a chategorical variable, the starting and ending points are categories.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu"><a href="reference/sccomp_proportional_fold_change.html">sccomp_proportional_fold_change</a></span><span class="op">(</span></span>
<span>     formula_composition <span class="op">=</span> <span class="op">~</span>  <span class="va">type</span>,</span>
<span>     from <span class="op">=</span>  <span class="st">"healthy"</span>, </span>
<span>     to <span class="op">=</span> <span class="st">"cancer"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">cell_group</span>, <span class="va">statement</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 36 × 2[39m</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">##    cell_group statement                                </span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m      [3m[90m&lt;glue&gt;[39m[23m                                   </span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="do">## [90m 1[39m B1         2.1-fold decrease (from 0.0565 to 0.0265)</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="do">## [90m 2[39m B2         2-fold decrease (from 0.0371 to 0.0186)  </span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## [90m 3[39m B3         1.4-fold decrease (from 0.0127 to 0.0091)</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">## [90m 4[39m BM         1.4-fold decrease (from 0.0067 to 0.0048)</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">## [90m 5[39m CD4 1      1.2-fold increase (from 0.0252 to 0.03)  </span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="do">## [90m 6[39m CD4 2      1.5-fold increase (from 0.0498 to 0.073) </span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="do">## [90m 7[39m CD4 3      2.7-fold decrease (from 0.0872 to 0.0325)</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="do">## [90m 8[39m CD4 4      1-fold decrease (from 0.0016 to 0.0016)  </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="do">## [90m 9[39m CD4 5      1-fold increase (from 0.0301 to 0.0311)  </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="do">## [90m10[39m CD8 1      1.2-fold increase (from 0.1054 to 0.1229)</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="do">## [90m# ℹ 26 more rows[39m</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h2>
<p>A plot of group proportion, faceted by groups. The blue boxplots represent the posterior predictive check. If the model is likely to be descriptively adequate to the data, the blue box plot should roughly overlay with the black box plot, which represents the observed data. The outliers are coloured in red. A box plot will be returned for every (discrete) covariate present in <code>formula_composition</code>. The colour coding represents the significant associations for composition and/or variability.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">## [1m[22mJoining with `by = join_by(cell_group, sample)`</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="do">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-12-1.png"><!-- --></p>
<p>A plot of estimates of differential composition (c_) on the x-axis and differential variability (v_) on the y-axis. The error bars represent 95% credible intervals. The dashed lines represent the minimal effect that the hypothesis test is based on. An effect is labelled as significant if bigger than the minimal effect according to the 95% credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/plot_1d_intervals.html">plot_1D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-13-1.png"><!-- --></p>
<p>We can plot the relationship between abundance and variability. As we can see below, they are positively correlated, you also appreciate that this relationship is by model for single cell RNA sequencing data.</p>
<p><code>sccomp</code> models, these relationship to obtain a shrinkage effect on the estimates of both the abundance and the variability. This shrinkage is adaptive as it is modelled jointly, thanks for Bayesian inference.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/plot_2d_intervals.html">plot_2D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-14-1.png"><!-- --></p>
<p>You can produce the series of plots calling the <code>plot</code> method.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 60 × 18[39m</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="do">##    cell_group  parameter factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m       [3m[90m&lt;chr&gt;[39m[23m     [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="do">## [90m 1[39m B immature  typecanc… [31mNA[39m      -[31m1[39m[31m.[39m[31m89[39m    -[31m1[39m[31m.[39m[31m39[39m   -[31m0[39m[31m.[39m[31m893[39m 0       0            [31mNA[39m</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="do">## [90m 2[39m B immature  typeheal… [31mNA[39m       0.893    1.39    1.89  0       0            [31mNA[39m</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="do">## [90m 3[39m B mem       typecanc… [31mNA[39m      -[31m2[39m[31m.[39m[31m32[39m    -[31m1[39m[31m.[39m[31m72[39m   -[31m1[39m[31m.[39m[31m0[39m[31m6[39m  0       0            [31mNA[39m</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="do">## [90m 4[39m B mem       typeheal… [31mNA[39m       1.06     1.72    2.32  0       0            [31mNA[39m</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="do">## [90m 5[39m CD4 cm S10… typecanc… [31mNA[39m      -[31m1[39m[31m.[39m[31m49[39m    -[31m1[39m[31m.[39m[31m0[39m[31m3[39m   -[31m0[39m[31m.[39m[31m601[39m 0       0            [31mNA[39m</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="do">## [90m 6[39m CD4 cm S10… typeheal… [31mNA[39m       0.601    1.03    1.49  0       0            [31mNA[39m</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="do">## [90m 7[39m CD4 cm hig… typecanc… [31mNA[39m       0.821    1.75    2.86  0       0            [31mNA[39m</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="do">## [90m 8[39m CD4 cm hig… typeheal… [31mNA[39m      -[31m2[39m[31m.[39m[31m86[39m    -[31m1[39m[31m.[39m[31m75[39m   -[31m0[39m[31m.[39m[31m821[39m 0       0            [31mNA[39m</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="do">## [90m 9[39m CD4 cm rib… typecanc… [31mNA[39m       0.283    0.993   1.70  0.005[4m5[24m[4m0[24m 0.001[4m3[24m[4m0[24m      [31mNA[39m</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="do">## [90m10[39m CD4 cm rib… typeheal… [31mNA[39m      -[31m1[39m[31m.[39m[31m70[39m    -[31m0[39m[31m.[39m[31m993[39m  -[31m0[39m[31m.[39m[31m283[39m 0.005[4m5[24m[4m0[24m 0.001[4m3[24m[4m0[24m      [31mNA[39m</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="do">## [90m# ℹ 50 more rows[39m</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="do">## [90m# ℹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,[39m</span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="do">## [90m#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,[39m</span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="do">## [90m#   count_data &lt;list&gt;[39m</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="categorical-factor-eg-bayesian-anova">Categorical factor (e.g. Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-eg-bayesian-anova"></a>
</h2>
<p>This is achieved through model comparison with <code>loo</code>. In the following example, the model with association with factors better fits the data compared to the baseline model with no factor association. For comparisons <code>check_outliers</code> must be set to FALSE as the leave-one-out must work with the same amount of data, while outlier elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5 <code>se_diff</code> difference of 5, we are confident that a model is better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>. In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> , </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>  <span class="va">model_with_factor_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">model_without_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h2>
<p>We can model the cell-group variability also dependent on the type, and so test differences in variability</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="do">## [90m# A tibble: 60 × 14[39m</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="do">##    cell_group       parameter factor c_lower c_effect  c_upper c_n_eff c_R_k_hat</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m            [3m[90m&lt;chr&gt;[39m[23m     [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m   [3m[90m&lt;dbl&gt;[39m[23m     [3m[90m&lt;dbl&gt;[39m[23m</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="do">## [90m 1[39m B immature       (Interce… [31mNA[39m      0.353     0.760  1.18      [4m6[24m192.      1.00</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="do">## [90m 2[39m B immature       typeheal… type    0.851     1.44   1.99      [4m5[24m723.      1.00</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="do">## [90m 3[39m B mem            (Interce… [31mNA[39m     -[31m1[39m[31m.[39m[31m49[39m     -[31m0[39m[31m.[39m[31m855[39m -[31m0[39m[31m.[39m[31m209[39m     [4m4[24m427.      1.00</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="do">## [90m 4[39m B mem            typeheal… type    1.07      1.85   2.61      [4m4[24m030.      1.00</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="do">## [90m 5[39m CD4 cm S100A4    (Interce… [31mNA[39m      1.31      1.66   1.99      [4m7[24m820.      1.00</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="do">## [90m 6[39m CD4 cm S100A4    typeheal… type    0.488     0.937  1.40      [4m5[24m677.      1.00</span></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="do">## [90m 7[39m CD4 cm high cyt… (Interce… [31mNA[39m     -[31m1[39m[31m.[39m[31m0[39m[31m5[39m     -[31m0[39m[31m.[39m[31m536[39m  0.046[4m0[24m    [4m5[24m171.      1.00</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="do">## [90m 8[39m CD4 cm high cyt… typeheal… type   -[31m3[39m[31m.[39m[31m0[39m[31m6[39m     -[31m1[39m[31m.[39m[31m26[39m   1.17      [4m3[24m388.      1.00</span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="do">## [90m 9[39m CD4 cm ribosome  (Interce… [31mNA[39m     -[31m0[39m[31m.[39m[31m0[39m[31m43[4m5[24m[39m    0.311  0.680     [4m4[24m391.      1.00</span></span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a><span class="do">## [90m10[39m CD4 cm ribosome  typeheal… type   -[31m1[39m[31m.[39m[31m80[39m     -[31m0[39m[31m.[39m[31m956[39m  0.001[4m0[24m[4m6[24m   [4m5[24m167.      1.00</span></span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a><span class="do">## [90m# ℹ 50 more rows[39m</span></span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a><span class="do">## [90m# ℹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,[39m</span></span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a><span class="do">## [90m#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;[39m</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h1>
<div class="section level2">
<h2 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h2>
<p>We recommend setting <code>bimodal_mean_variability_association  = TRUE</code>. The bimodality of the mean-variability association can be confirmed from the plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level2">
<h2 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h2>
<p>We recommend setting <code>bimodal_mean_variability_association  = FALSE</code> (Default).</p>
</div>
<div class="section level2">
<h2 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h2>
<p>It is possible to directly evaluate the posterior distribution. In this example, we plot the Monte Carlo chain for the slope parameter of the first cell type. We can see that it has converged and is negative with probability 1.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-19-1.png"><!-- --></p>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="do">## [1m[22mJoining with `by = join_by(cell_group, sample)`</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="do">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-20-1.png"><!-- --></p>
<p>Plot 2D significance plot. Data points are cell groups. Error bars are the 95% credible interval. The dashed lines represent the default threshold fold change for which the probabilities (c_pH0, v_pH0) are calculated. pH0 of 0 represent the rejection of the null hypothesis that no effect is observed.</p>
<p>This plot is provided only if differential variability has been tested. The differential variability estimates are reliable only if the linear association between mean and variability for <code>(intercept)</code> (left-hand side facet) is satisfied. A scatterplot (besides the Intercept) is provided for each category of interest. The for each category of interest, the composition and variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-21-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="the-old-framework">The old framework<a class="anchor" aria-label="anchor" href="#the-old-framework"></a>
</h2>
<p>The new tidy framework was introduced in 2024, two, understand the differences and improvements. Compared to the old framework, please read this <a href="https://tidyomics.github.io/tidyomicsBlog/post/2023-12-07-tidy-sccomp/" class="external-link">blog post</a>.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/sccomp" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/stemangiola/sccomp/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/stemangiola/sccomp/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sccomp</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
