<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Robust Outlier-aware Estimation of Composition and Heterogeneity for Single-cell Data • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Robust Outlier-aware Estimation of Composition and Heterogeneity for Single-cell Data">
<meta property="og:description" content="A robust and outlier-aware method for testing differential tissue composition from single-cell data. This model can infer changes in tissue composition and heterogeneity, and can produce realistic data simulations based on any existing dataset. This model can also transfer knowledge from a large set of integrated datasets to increase accuracy further.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the sccomp package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MangiolaLaboratory/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<p>sccomp - Tests differences in cell type proportions and variability from single-cell data ================</p>
<!-- badges: start -->

<!-- badges: end -->
<p>Cellular omics such as single-cell genomics, proteomics, and microbiomics allow the characterization of tissue and microbial community composition, which can be compared between conditions to identify biological drivers. This strategy has been critical to unveiling markers of disease progression in conditions such as cancer and pathogen infections.</p>
<p>For cellular omic data, no method for differential variability analysis exists, and methods for differential composition analysis only take a few fundamental data properties into account. Here we introduce <strong>sccomp</strong>, a generalised method for differential composition and variability analyses capable of jointly modelling data count distribution, compositionality, group-specific variability, and proportion mean-variability association, while being robust to outliers.</p>
<p><strong>sccomp</strong> is an extensive analysis framework that allows realistic data simulation and cross-study knowledge transfer. We demonstrate that mean-variability association is ubiquitous across technologies, highlighting the inadequacy of the very popular Dirichlet-multinomial modeling and providing essential principles for differential variability analysis.</p>
<p><a href="https://www.youtube.com/watch?v=R_lt58We9nA&amp;ab_channel=RConsortium" target="_blank" class="external-link"> <img src="https://img.youtube.com/vi/R_lt58We9nA/mqdefault.jpg" alt="Watch the video" width="280" height="180" border="10"></a></p>
<div class="section level3">
<h3 id="cite">Cite<a class="anchor" aria-label="anchor" href="#cite"></a>
</h3>
<p>Mangiola, Stefano, Alexandra J. Roth-Schulze, Marie Trussart, Enrique Zozaya-Valdés, Mengyao Ma, Zijie Gao, Alan F. Rubin, Terence P. Speed, Heejung Shim, and Anthony T. Papenfuss. 2023. “Sccomp: Robust Differential Composition and Variability Analysis for Single-Cell Data.” Proceedings of the National Academy of Sciences of the United States of America 120 (33): e2203828120. <a href="https://doi.org/10.1073/pnas.2203828120" class="external-link uri">https://doi.org/10.1073/pnas.2203828120</a> <a href="https://www.pnas.org/doi/full/10.1073/pnas.2203828120" class="external-link">PNAS - sccomp: Robust differential composition and variability analysis for single-cell data</a></p>
</div>
<div class="section level1">
<div class="page-header"><h1 id="section">
<img src="inst/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h1></div>
<p><code>sccomp</code> tests differences in cell type proportions from single-cell data. It is robust against outliers, it models continuous and discrete factors, and capable of random-effect/intercept modelling.</p>
<div class="section level2">
<h2 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h2>
<ul>
<li>Complex linear models with continuous and categorical covariates</li>
<li>Multilevel modelling, with population fixed and random effects/intercept</li>
<li>Modelling data from counts</li>
<li>Testing differences in cell-type proportionality</li>
<li>Testing differences in cell-type specific variability</li>
<li>Cell-type information share for variability adaptive shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p><code>sccomp</code> is based on <code>cmdstanr</code> which provides the latest version of <code>cmdstan</code> the Bayesian modelling tool. <code>cmdstanr</code> is not on CRAN, so we need to have 3 simple step process (that will be prompted to the user is forgot).</p>
<ol style="list-style-type: decimal">
<li>R installation of <code>sccomp</code>
</li>
<li>R installation of <code>cmdstanr</code>
</li>
<li>
<code>cmdstanr</code> call to <code>cmdstan</code> installation</li>
</ol>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 1</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cmdstanr"</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://stan-dev.r-universe.dev/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"repos"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 3</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">check_cmdstan_toolchain</a></span><span class="op">(</span>fix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Just checking system setting</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">install_cmdstan</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 1</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"MangiolaLaboratory/sccomp"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cmdstanr"</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://stan-dev.r-universe.dev/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"repos"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 3</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">check_cmdstan_toolchain</a></span><span class="op">(</span>fix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Just checking system setting</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">install_cmdstan</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code>sccomp_estimate</code></td>
<td>Fit the model onto the data, and estimate the coefficients</td>
</tr>
<tr>
<td><code>sccomp_remove_outliers</code></td>
<td>Identify outliers probabilistically based on the model fit, and exclude them from the estimation</td>
</tr>
<tr>
<td><code>sccomp_test</code></td>
<td>Calculate the probability that the coefficients are outside the H0 interval (i.e. test_composition_above_logit_fold_change)</td>
</tr>
<tr>
<td><code>sccomp_replicate</code></td>
<td>Simulate data from the model, or part of the model</td>
</tr>
<tr>
<td><code>sccomp_predict</code></td>
<td>Predicts proportions, based on the model, or part of the model</td>
</tr>
<tr>
<td><code>sccomp_remove_unwanted_variation</code></td>
<td>Removes the variability for unwanted factors</td>
</tr>
<tr>
<td><code>plot</code></td>
<td>Plots summary plots to asses significance</td>
</tr>
</tbody>
</table>
</div>
<div class="section level1">
<h1 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h1>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MangiolaLaboratory/sccomp" class="external-link">sccomp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://forcats.tidyverse.org/" class="external-link">forcats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"seurat_obj"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sce_obj"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"counts_obj"</span><span class="op">)</span></span></code></pre></div>
<p><code>sccomp</code> can model changes in composition and variability. By default, the formula for variability is either <code>~1</code>, which assumes that the cell-group variability is independent of any covariate or <code>~ factor_of_interest</code>, which assumes that the model is dependent on the factor of interest only. The variability model must be a subset of the model for composition.</p>
<div class="section level2">
<h2 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h2>
<p>Of the output table, the estimate columns start with the prefix <code>c_</code> indicate <code>composition</code>, or with <code>v_</code> indicate <code>variability</code> (when formula_variability is set).</p>
<div class="section level3">
<h3 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">sce_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Here you see the results of the fit, the effects of the factor on composition and variability. You also can see the uncertainty around those effects.</p>
<p>The output is a tibble containing the <strong>Following columns</strong></p>
<ul>
<li>
<code>cell_group</code> - The cell groups being tested.</li>
<li>
<code>parameter</code> - The parameter being estimated from the design matrix described by the input <code>formula_composition</code> and <code>formula_variability</code>.</li>
<li>
<code>factor</code> - The covariate factor in the formula, if applicable (e.g., not present for Intercept or contrasts).</li>
<li>
<code>c_lower</code> - Lower (2.5%) quantile of the posterior distribution for a composition (c) parameter.</li>
<li>
<code>c_effect</code> - Mean of the posterior distribution for a composition (c) parameter.</li>
<li>
<code>c_upper</code> - Upper (97.5%) quantile of the posterior distribution for a composition (c) parameter.</li>
<li>
<code>c_pH0</code> - Probability of the null hypothesis (no difference) for a composition (c). This is not a p-value.</li>
<li>
<code>c_FDR</code> - False-discovery rate of the null hypothesis for a composition (c).</li>
<li>
<code>v_lower</code> - Lower (2.5%) quantile of the posterior distribution for a variability (v) parameter.</li>
<li>
<code>v_effect</code> - Mean of the posterior distribution for a variability (v) parameter.</li>
<li>
<code>v_upper</code> - Upper (97.5%) quantile of the posterior distribution for a variability (v) parameter.</li>
<li>
<code>v_pH0</code> - Probability of the null hypothesis for a variability (v).</li>
<li>
<code>v_FDR</code> - False-discovery rate of the null hypothesis for a variability (v).</li>
<li>
<code>count_data</code> - Nested input count data.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h2>
<p>A plot of group proportions, faceted by groups. The blue boxplots represent the posterior predictive check. If the model is descriptively adequate for the data, the blue boxplots should roughly overlay the black boxplots, which represent the observed data. The outliers are coloured in red. A boxplot will be returned for every (discrete) covariate present in formula_composition. The colour coding represents the significant associations for composition and/or variability.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span></code></pre></div>
<p>A plot of estimates of differential composition (c_) on the x-axis and differential variability (v_) on the y-axis. The error bars represent 95% credible intervals. The dashed lines represent the minimal effect that the hypothesis test is based on. An effect is labelled as significant if it exceeds the minimal effect according to the 95% credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/plot_1D_intervals.html">plot_1D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the relationship between abundance and variability. As we can see below, they are positively correlated. sccomp models this relationship to obtain a shrinkage effect on the estimates of both the abundance and the variability. This shrinkage is adaptive as it is modelled jointly, thanks to Bayesian inference.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/plot_2D_intervals.html">plot_2D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can produce the series of plots calling the <code>plot</code> method.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-proportions-directly-eg-from-deconvolution">Model proportions directly (e.g. from deconvolution)<a class="anchor" aria-label="anchor" href="#model-proportions-directly-eg-from-deconvolution"></a>
</h2>
<p><strong>Note:</strong> If counts are available, we strongly discourage the use of proportions, as an important source of uncertainty (i.e., for rare groups/cell types) is not modeled.</p>
<p>The use of proportions is better suited for modelling deconvolution results (e.g., of bulk RNA data), in which case counts are not available.</p>
<p>Proportions should be greater than 0. Assuming that zeros derive from a precision threshold (e.g., deconvolution), zeros are converted to the smallest non-zero value.</p>
</div>
<div class="section level2">
<h2 id="continuous-factor">Continuous factor<a class="anchor" aria-label="anchor" href="#continuous-factor"></a>
</h2>
<p><code>sccomp</code> is able to fit erbitrary complex models. In this example we have a continuous and binary covariate.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span></span>
<span>    <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="va">continuous_covariate</span>, </span>
<span>      .sample <span class="op">=</span> <span class="va">sample</span>, .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose<span class="op">=</span><span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="random-effect-modeling">Random Effect Modeling<a class="anchor" aria-label="anchor" href="#random-effect-modeling"></a>
</h2>
<p><code>sccomp</code> supports multilevel modeling by allowing the inclusion of random effects in the compositional and variability formulas. This is particularly useful when your data has hierarchical or grouped structures, such as measurements nested within subjects, batches, or experimental units. By incorporating random effects, sccomp can account for variability at different levels of your data, improving model fit and inference accuracy.</p>
<div class="section level3">
<h3 id="random-intercept-model">Random Intercept Model<a class="anchor" aria-label="anchor" href="#random-intercept-model"></a>
</h3>
<p>In this example, we demonstrate how to fit a random intercept model using sccomp. We’ll model the cell-type proportions with both fixed effects (e.g., treatment) and random effects (e.g., subject-specific variability).</p>
<p>Here is the input data</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">## Loading required package: SeuratObject</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="do">## Loading required package: sp</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="do">## 'SeuratObject' was built under R 4.4.0 but the current version is</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="do">## 4.4.1; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="do">## for R may have changed</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="do">## Attaching package: 'SeuratObject'</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="do">## The following objects are masked from 'package:base':</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="do">##     intersect, t</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="do">## [90m# A tibble: 106,297 × 9[39m</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="do">##    cell_group nCount_RNA nFeature_RNA group__ group__wrong sample type  group2__</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="do">##    [3m[90m&lt;chr&gt;[39m[23m           [3m[90m&lt;dbl&gt;[39m[23m        [3m[90m&lt;int&gt;[39m[23m [3m[90m&lt;chr&gt;[39m[23m   [3m[90m&lt;chr&gt;[39m[23m        [3m[90m&lt;chr&gt;[39m[23m  [3m[90m&lt;chr&gt;[39m[23m [3m[90m&lt;chr&gt;[39m[23m   </span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="do">## [90m 1[39m CD4 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="do">## [90m 2[39m Mono clas…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="do">## [90m 3[39m CD4 cm S1…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="do">## [90m 4[39m B immature          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="do">## [90m 5[39m CD8 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a><span class="do">## [90m 6[39m CD4 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a><span class="do">## [90m 7[39m Mono clas…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a><span class="do">## [90m 8[39m CD4 cm S1…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a><span class="do">## [90m 9[39m CD4 cm hi…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a><span class="do">## [90m10[39m B immature          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a><span class="do">## [90m# ℹ 106,287 more rows[39m</span></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a><span class="do">## [90m# ℹ 1 more variable: continuous_covariate &lt;dbl&gt;[39m</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> </span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="random-effect-model-random-slopes">Random Effect Model (random slopes)<a class="anchor" aria-label="anchor" href="#random-effect-model-random-slopes"></a>
</h3>
<p><code>sccomp</code> can model random slopes. We providean example below.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="va">type</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span>,</span>
<span>      .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>      .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>      bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="nested-random-effects">Nested Random Effects<a class="anchor" aria-label="anchor" href="#nested-random-effects"></a>
</h3>
<p>If you have a more complex hierarchy, such as measurements nested within subjects and subjects nested within batches, you can include multiple grouping variables. Here <code>group2__</code> is nested within <code>group__</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="va">type</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">group2__</span><span class="op">)</span>,</span>
<span>      .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>      .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>      bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="an-aid-to-result-interpretation-and-communication">An aid to result interpretation and communication<a class="anchor" aria-label="anchor" href="#an-aid-to-result-interpretation-and-communication"></a>
</h2>
<p>The estimated effects are expressed in the unconstrained space of the parameters, similar to differential expression analysis that expresses changes in terms of log fold change. However, for differences in proportion, logit fold change must be used, which is harder to interpret and understand.</p>
<p>Therefore, we provide a more intuitive proportional fold change that can be more easily understood. However, these cannot be used to infer significance (use sccomp_test() instead), and a lot of care must be taken given the nonlinearity of these measures (a 1-fold increase from 0.0001 to 0.0002 carries a different weight than a 1-fold increase from 0.4 to 0.8).</p>
<p>From your estimates, you can specify which effects you are interested in (this can be a subset of the full model if you wish to exclude unwanted effects), and the two points you would like to compare.</p>
<p>In the case of a categorical variable, the starting and ending points are categories.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu"><a href="reference/sccomp_proportional_fold_change.html">sccomp_proportional_fold_change</a></span><span class="op">(</span></span>
<span>     formula_composition <span class="op">=</span> <span class="op">~</span>  <span class="va">type</span>,</span>
<span>     from <span class="op">=</span>  <span class="st">"healthy"</span>, </span>
<span>     to <span class="op">=</span> <span class="st">"cancer"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_group</span>, <span class="va">statement</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="categorical-factor-eg-bayesian-anova">Categorical factor (e.g. Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-eg-bayesian-anova"></a>
</h2>
<p>This is achieved through model comparison with <code>loo</code>. In the following example, the model with association with factors better fits the data compared to the baseline model with no factor association. For comparisons <code>check_outliers</code> must be set to FALSE as the leave-one-out must work with the same amount of data, while outlier elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5 <code>se_diff</code> difference of 5, we are confident that a model is better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>. In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    inference_method <span class="op">=</span> <span class="st">"hmc"</span>,</span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    inference_method <span class="op">=</span> <span class="st">"hmc"</span>,</span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">model_with_factor_association</span>, <span class="st">"fit"</span><span class="op">)</span><span class="op">$</span><span class="fu">loo</span><span class="op">(</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">model_without_association</span>, <span class="st">"fit"</span><span class="op">)</span><span class="op">$</span><span class="fu">loo</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h2>
<p>We can model the cell-group variability also dependent on the type, and so test differences in variability</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<p><strong>Plot 1D significance plot</strong></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><strong>Plot 2D significance plot</strong> Data points are cell groups. Error bars are the 95% credible interval. The dashed lines represent the default threshold fold change for which the probabilities (c_pH0, v_pH0) are calculated. pH0 of 0 represent the rejection of the null hypothesis that no effect is observed.</p>
<p>This plot is provided only if differential variability has been tested. The differential variability estimates are reliable only if the linear association between mean and variability for <code>(intercept)</code> (left-hand side facet) is satisfied. A scatterplot (besides the Intercept) is provided for each category of interest. For each category of interest, the composition and variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h1>
<div class="section level2">
<h2 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h2>
<p>We recommend setting <code>bimodal_mean_variability_association  = TRUE</code>. The bimodality of the mean-variability association can be confirmed from the plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level2">
<h2 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h2>
<p>We recommend setting <code>bimodal_mean_variability_association  = FALSE</code> (Default).</p>
</div>
<div class="section level2">
<h2 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h2>
<p>It is possible to directly evaluate the posterior distribution. In this example, we plot the Monte Carlo chain for the slope parameter of the first cell type. We can see that it has converged and is negative with probability 1.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/cmdstanr/" class="external-link">cmdstanr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/posterior/" class="external-link">posterior</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/bayesplot/" class="external-link">bayesplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assuming res contains the fit object from cmdstanr</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract draws for 'beta[2,1]'</span></span>
<span><span class="va">draws</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mc-stan.org/posterior/reference/draws_array.html" class="external-link">as_draws_array</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="fu">draws</span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a traceplot for 'beta[2,1]'</span></span>
<span><span class="fu"><a href="https://mc-stan.org/bayesplot/reference/MCMC-traces.html" class="external-link">mcmc_trace</a></span><span class="op">(</span><span class="va">draws</span>, pars <span class="op">=</span> <span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-old-framework">The old framework<a class="anchor" aria-label="anchor" href="#the-old-framework"></a>
</h2>
<p>The new tidy framework was introduced in 2024, two, understand the differences and improvements. Compared to the old framework, please read this <a href="https://tidyomics.github.io/tidyomicsBlog/post/2023-12-07-tidy-sccomp/" class="external-link">blog post</a>.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/sccomp" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/MangiolaLaboratory/sccomp/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MangiolaLaboratory/sccomp/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sccomp</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/MangiolaLaboratory/sccomp/actions/" class="external-link"><img src="https://github.com/MangiolaLaboratory/sccomp/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
