---
title: "sccomp - Outlier-aware and count-based compositional analysis of single-cell data"
output: github_document
always_allow_html: true
editor_options: 
  markdown: 
    wrap: 72
---

<!-- badges: start -->

[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![R build
status](https://github.com/MangiolaLaboratory/sccomp/workflows/R-CMD-check/badge.svg)](https://github.com/MangiolaLaboratory/sccomp/actions/)

<!-- badges: end -->

Cell omics such as single-cell genomics, proteomics, and microbiomics allow the characterisation of tissue and microbial community composition, which can be compared between conditions to identify biological drivers. This strategy has been critical to unveiling markers of disease progression, such as cancer and pathogen infection.

For cell omic data, no method for differential variability analysis exists, and methods for differential composition analysis only take a few fundamental data properties into account. Here we introduce sccomp, a generalised method for differential composition and variability analyses capable of jointly modelling data count distribution, compositionality, group-specific variability, and proportion mean-variability association, with awareness against outliers.

Sccomp is an extensive analysis framework that allows realistic data simulation and cross-study knowledge transfer. We demonstrate that mean-variability association is ubiquitous across technologies, highlighting the inadequacy of the very popular Dirichlet-multinomial modelling and providing essential principles for differential variability analysis.

We show that sccomp accurately fits experimental data, with a 50% incremental improvement over state-of-the-art algorithms. Using sccomp, we identified novel differential constraints and composition in the microenvironment of primary breast cancer.

<a href="https://www.youtube.com/watch?v=R_lt58We9nA&ab_channel=RConsortium" target="_blank">
 <img src="https://img.youtube.com/vi/R_lt58We9nA/mqdefault.jpg" alt="Watch the video" width="280" height="180" border="10" />
</a>

```{r echo=FALSE}
knitr::opts_chunk$set( fig.path = "inst/figures/")
```


```{r child="man/fragments/intro.Rmd"}
```

## The old framework

The new tidy framework was introduced in 2024, two, understand the differences and improvements. Compared to the old framework, please read this [blog post](https://tidyomics.github.io/tidyomicsBlog/post/2023-12-07-tidy-sccomp/).