---
title: "sccomp - Outlier-aware and count-based compositional analysis of single-cell data"
output: github_document
always_allow_html: true
---

<!-- badges: start -->
[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing) [![R build status](https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg)](https://github.com/stemangiola/tidyseurat/actions/)
<!-- badges: end -->


#  <img src="inst/logo-01.png" height="139px" width="120px" />

```{r echo=FALSE}
knitr::opts_chunk$set( fig.path = "inst/figures/")
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(sccomp)
library(ggplot2)
library(forcats)
library(tidyr)
library(rstan)
data("seurat_obj")
data("sce_obj")
data("counts_obj")
```

Cell omics such as single-cell genomics, proteomics and microbiomics allow the characterisation of tissue and microbial community composition, which can be compared between conditions to identify biological drivers. This strategy has been critical to unveiling markers of disease progression such as cancer and pathogen infection. For cell omic data, no method for differential variability analysis exists, and methods for differential composition analysis only take a few fundamental data properties into account. Here we introduce sccomp, a generalised method for differential composition and variability analyses able to jointly model data count distribution, compositionality, group-specific variability and proportion mean-variability association, with awareness against outliers. Sccomp is an extensive analysis framework that allows realistic data simulation and cross-study knowledge transfer. Here, we demonstrate that mean-variability association is ubiquitous across technologies showing the inadequacy of the very popular Dirichlet-multinomial modelling and provide mandatory principles for differential variability analysis. We show that sccomp accurately fits experimental data, with a 50% incremental improvement over state-of-the-art algorithms. Using sccomp, we identified novel differential constraints and composition in the microenvironment of primary breast cancer. 

# Installation 

## (simple) Suggested for single-cell and CyTOF analyses

**Bioconductor**

```{r eval=FALSE}
if (!requireNamespace("BiocManager")) install.packages("BiocManager")
BiocManager::install("sccomp")
```

**Github**

```{r eval=FALSE}
devtools::install_github("stemangiola/sccomp")
```

## (more complex and efficient, until further optimisation of the default installation) Suggested for microbiomics

**Github**

```{r eval=FALSE}
install.packages("cmdstanr", repos = c("https://mc-stan.org/r-packages/", getOption("repos")))
check_cmdstan_toolchain()
install_cmdstan(cores = 2)
# Then, check the correct cmdstanr installation here
# https://mc-stan.org/cmdstanr/articles/cmdstanr.html

# Then install sccomp with the cmdstanr branch
devtools::install_github("stemangiola/sccomp@cmdstanr")
```

# Analysis

`sccomp` can model changes in composition and variability. Normally the furmula for variability is either `~1`, which assumes that the cell-group variability is independent on any covariate, or `~ factor_of_interest`, which assumes that the model is dependent on the factor of interest only. However, more complex models for variability are possible, is the sample size is large. In any case the model for variability must be a subset of the model for composition.

## From Seurat, SingleCellExperiment, metadata objects 

```{r eval=FALSE}

single_cell_object |>
  sccomp_glm( 
   formula_composition = ~ type, 
   .sample =  sample, 
   .cell_group = cell_group 
  )

```

## From counts

```{r warning=FALSE}

res =
  counts_obj |>
  sccomp_glm( 
    formula_composition = ~ type, 
    .sample = sample,
    .cell_group = cell_group,
    .count = count
  )

res
```

Of the output table, the estimate columns star twith the prefix `c_` indicate `composition`, or with `v_` indicate `variability` (when formula_variability is set).


## Differential variability

```{r}

seurat_obj |>
  sccomp_glm( 
   formula_composition = ~ type, 
    formula_variability = ~ type,
    .sample = sample,
    .cell_group = cell_group,
  )

```

## With contrasts

```{r warning=FALSE}

seurat_obj |>
  sccomp_glm( 
    formula_composition = ~ 0 + type, 
    contrasts =  c("typecancer - typebenign", "typebenign - typecancer"),
    .sample = sample,
    .cell_group = cell_group
  )

```

## Model comparison (e.g. Bayesian ANOVA) with `loo`

In the following example the model with association with factors better fits the data compares to the baseline model with no factor association. For comparisons `check_outliers` must be set to FALSE as the leave-one-out must work with the same amount of data, while outlier elimination does not guarantee it.

If `elpd_diff` is away from zero of > 5 `se_diff` difference of 5 we are confident that a model is better than the other [reference](https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola). In this case -79.9 / 11.5 = -6.9, therefore we can conclude that model one, the one with factor association, is better than model two.

```{r}
library(loo)

# Fit first model
model_with_factor_association = 
  seurat_obj |>
  sccomp_glm( 
   formula_composition = ~ type, 
   .sample =  sample, 
   .cell_group = cell_group, 
   check_outliers = FALSE
  )

# Fit second model
model_without_association = 
  seurat_obj |>
  sccomp_glm( 
   formula_composition = ~ 1, 
   .sample =  sample, 
   .cell_group = cell_group, 
   check_outliers = FALSE
  )

loo_with_factor_association <- loo(model_with_factor_association |> attr("fit"))
loo_without_association <- loo(model_without_association |> attr("fit"))

# Compare models
loo_compare(loo_with_factor_association, loo_without_association)

```


## Suggested settings for single-cell RNA sequencing

We recommend to set `bimodal_mean_variability_association  = TRUE`. The bimodality of the mean-variability association can be confirmed from the plots$credible_intervals_2D (see below).

## Suggested settings for CyTOF and microbiome data

We recommend to set `bimodal_mean_variability_association  = FALSE` (Default).

## Visualise data + inference

```{r, out.height="200%"}
plots = plot_summary(res) 


```

Plot of group proportion, faceted by groups. The blue boxplots represent the posterior predictive check. If the model is likely be descriptively adequate to the data, the blue box plot should roughly overlay with the black box plot, which represent the observed data. The outliers are coloured in red. A box plot will be returned for every (discrete) covariates present in `formula_composition`. The color coding represent the significant associations for composition and/or variability.


```{r}
plots$boxplot

```

Plot of estimates of differential composition (c_) on the x axis, and differential variability (v_) on the y axis. The error bars represent 95% credible intervals. The dashed lines represent the minimal effect that the hypothesis test is based on. An effect is labelled as significant if bigger than the minimal effect according to the 95% credible interval. Facets represent the covariates in the model.

```{r}
plots$credible_intervals_1D

```

## Visualisation of the MCMC chains from the posterior distribution

It is possible to directly evaluate the posterior distribution. In this example we plot the Monte Carlo chain for the slope parameter of the first cell type. We can see that has converged and is negative with probability 1.

```{r}
res %>% attr("fit") %>% rstan::traceplot("beta[2,1]")
```


## Differential variability

We can model the cell-group variability also dependent on type, and so test differences in variability

```{r warning=FALSE}

res = 
  counts_obj |>
  sccomp_glm( 
    formula_composition = ~ type, 
    formula_variability = ~ type, 
    .sample = sample,
    .cell_group = cell_group,
    .count = count
  )

res
```

Plot 1D significance plot

```{r}
plots = plot_summary(res)

plots$credible_intervals_1D
```

Plot 2D significance plot. Data points are cell groups. Error bars are the 95% credible interval. The dashed lines represent the default threshold fold change for which the probabilities (c_pH0, v_pH0) are calculated. pH0 of 0 represent the rejection of the null hypothesis, that no effect is observed.

This plot is provided only if differential variability has been tested. The differential variability estimates are reliable only if the linear association between mean and variability for `(intercept)` (left-hand side facet) is satisfied. A scatterplot (beside the Intercept) is provided for each of the categories of interest. The for each category of interest, the composition and variability effects should be generally uncorrelated. 

```{r}
plots$credible_intervals_2D
```
