<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>sccomp: Differential Composition and Variability Analysis for Single-Cell Data — sccomp-package • sccomp</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="sccomp: Differential Composition and Variability Analysis for Single-Cell Data — sccomp-package"><meta name="description" content="Comprehensive R package for differential composition and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data. Provides robust Bayesian modeling with outlier detection, random effects, and advanced statistical methods for cell type proportion analysis. Features include probabilistic outlier identification, mixed-effect modeling, differential variability testing, and comprehensive visualization tools. Perfect for cancer research, immunology, developmental biology, and single-cell genomics applications.
The sccomp package provides comprehensive tools for differential composition
and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data.
It implements robust Bayesian modeling with outlier detection, random effects,
and advanced statistical methods for cell type proportion analysis."><meta property="og:description" content="Comprehensive R package for differential composition and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data. Provides robust Bayesian modeling with outlier detection, random effects, and advanced statistical methods for cell type proportion analysis. Features include probabilistic outlier identification, mixed-effect modeling, differential variability testing, and comprehensive visualization tools. Perfect for cancer research, immunology, developmental biology, and single-cell genomics applications.
The sccomp package provides comprehensive tools for differential composition
and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data.
It implements robust Bayesian modeling with outlier detection, random effects,
and advanced statistical methods for cell type proportion analysis."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sccomp</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.21</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/introduction.html">sccomp: Differential Composition and Variability Analysis for Single-Cell Data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MangiolaLaboratory/sccomp/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>sccomp: Differential Composition and Variability Analysis for Single-Cell Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/MangiolaLaboratory/sccomp/blob/master/R/sccomp-package.R" class="external-link"><code>R/sccomp-package.R</code></a></small>
      <div class="d-none name"><code>sccomp-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Comprehensive R package for differential composition and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data. Provides robust Bayesian modeling with outlier detection, random effects, and advanced statistical methods for cell type proportion analysis. Features include probabilistic outlier identification, mixed-effect modeling, differential variability testing, and comprehensive visualization tools. Perfect for cancer research, immunology, developmental biology, and single-cell genomics applications.</p>
<p>The sccomp package provides comprehensive tools for differential composition
and variability analysis in single-cell RNA sequencing, CyTOF, and microbiome data.
It implements robust Bayesian modeling with outlier detection, random effects,
and advanced statistical methods for cell type proportion analysis.</p>
    </div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The main functions are:</p>
<ul><li><p><code><a href="sccomp_estimate.html">sccomp_estimate</a></code> - Perform differential composition and variability analysis</p></li>
<li><p><code><a href="sccomp_test.html">sccomp_test</a></code> - Test for differential composition with significance testing</p></li>
<li><p><code><a href="sccomp_remove_outliers.html">sccomp_remove_outliers</a></code> - Identify and remove outlier samples</p></li>
<li><p><code><a href="sccomp_predict.html">sccomp_predict</a></code> - Predict cell type proportions for new samples</p></li>
<li><p><code><a href="sccomp_remove_unwanted_effects.html">sccomp_remove_unwanted_effects</a></code> - Remove unwanted variation from data</p></li>
<li><p><code><a href="sccomp_proportional_fold_change.html">sccomp_proportional_fold_change</a></code> - Calculate proportional fold changes</p></li>
<li><p>Plotting functions: <code><a href="plot.sccomp_tbl.html">plot.sccomp_tbl</a></code>, <code><a href="sccomp_boxplot.html">sccomp_boxplot</a></code>, <code><a href="plot_1D_intervals.html">plot_1D_intervals</a></code>, <code><a href="plot_2D_intervals.html">plot_2D_intervals</a></code></p></li>
</ul><p>For detailed information on usage, see the package vignettes:</p>
<p><code>browseVignettes("sccomp")</code></p>
<p>All software-related questions should be posted to the GitHub Issues page:</p>
<p><a href="https://github.com/MangiolaLaboratory/sccomp/issues" class="external-link">https://github.com/MangiolaLaboratory/sccomp/issues</a></p>
<p>The code can be viewed at the GitHub repository:</p>
<p><a href="https://github.com/MangiolaLaboratory/sccomp" class="external-link">https://github.com/MangiolaLaboratory/sccomp</a></p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Mangiola, S., Roth-Schulze, A.J., Trussart, M., Zozaya-Valdés, E., Ma, M.,
Gao, Z., Rubin, A.F., Speed, T.P., Shim, H., &amp; Papenfuss, A.T. (2023).
sccomp: Robust differential composition and variability analysis for single-cell data.
Proceedings of the National Academy of Sciences, 120(33), e2203828120.
<a href="https://doi.org/10.1073/pnas.2203828120" class="external-link">doi:10.1073/pnas.2203828120</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://github.com/MangiolaLaboratory/sccomp" class="external-link">https://github.com/MangiolaLaboratory/sccomp</a></p></li>
<li><p><a href="https://mangiolalaboratory.github.io/sccomp/">https://mangiolalaboratory.github.io/sccomp/</a></p></li>
<li><p>Report bugs at <a href="https://github.com/MangiolaLaboratory/sccomp/issues" class="external-link">https://github.com/MangiolaLaboratory/sccomp/issues</a></p></li>
</ul><p>Useful links:</p><ul><li><p><a href="https://github.com/MangiolaLaboratory/sccomp" class="external-link">https://github.com/MangiolaLaboratory/sccomp</a></p></li>
<li><p>Report bugs at <a href="https://github.com/MangiolaLaboratory/sccomp/issues" class="external-link">https://github.com/MangiolaLaboratory/sccomp/issues</a></p></li>
</ul></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><strong>Maintainer</strong>: Stefano Mangiola <a href="mailto:stefano.mangiola@unimelb.edu.au">stefano.mangiola@unimelb.edu.au</a></p>
<p>Authors:</p><ul><li><p>Alexandra J. Roth-Schulze</p></li>
<li><p>Marie Trussart</p></li>
<li><p>Enrique Zozaya-Valdés</p></li>
<li><p>Mengyao Ma</p></li>
<li><p>Zijie Gao</p></li>
<li><p>Alan F. Rubin</p></li>
<li><p>Terence P. Speed</p></li>
<li><p>Heejung Shim</p></li>
<li><p>Anthony T. Papenfuss</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stefano Mangiola, Alexandra J. Roth-Schulze, Marie Trussart, Enrique Zozaya-Valdés, Mengyao Ma, Zijie Gao, Alan F. Rubin, Terence P. Speed, Heejung Shim, Anthony T. Papenfuss.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

