<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the sccomp package • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the sccomp package">
<meta property="og:description" content="sccomp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the sccomp package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2022-03-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<div class="section level2">
<h2 id="section">
<img src="figures/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h2>
<p>Single-cell transcriptomics allows the unbiased characterisation of
the cellular composition of tissues. The cellular composition can be
compared between biological or clinical conditions to identify potential
cellular drivers. This strategy has been critical to unveil drivers of
immune response in cancer and pathogen infection from single-cell data.
Developing a robust statistical method for differential composition
analyses from single-cell data is crucial for driving discoveries. The
compositional data from single-cell experiments has four main
properties. The data is in count form; counts underlie inversely
correlated proportions that sum to one; larger cell groups are more
variable across samples than small groups; real-world data is rich in
outlier observation. A model that covers more than two of these
properties is currently lacking. <strong>Here, we present a robust and
outlier-aware method for testing differential tissue composition from
single-cell data. This model can also transfer knowledge from a large
set of integrated datasets to increase accuracy further. We present how
this model can be applied to identify novel compositional and
heterogeneity changes in existing studies.</strong></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
   <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
 <span class="op">}</span>
 <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<div class="section level3">
<h3 id="from-seurat-object">From Seurat Object<a class="anchor" aria-label="anchor" href="#from-seurat-object"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  seurat_obj <span class="sc">|&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>   formula_composition <span class="ot">=</span> <span class="er">~</span> type, </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    formula_variability <span class="ot">=</span> <span class="er">~</span> <span class="dv">1</span>, </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    sample, </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    cell_group </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="er">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  sce_obj <span class="sc">|&gt;</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    formula_composition <span class="ot">=</span> <span class="er">~</span> type, </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    formula_variability <span class="ot">=</span> <span class="er">~</span> <span class="dv">1</span>, </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    sample, </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    cell_group </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="er">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-data-frame">From data.frame<a class="anchor" aria-label="anchor" href="#from-data-frame"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">=</span>
  <span class="va">seurat_obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> |&gt;
  <span class="fu"><a href="../reference/sccomp_glm.html">sccomp_glm</a></span><span class="op">(</span>
    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, 
    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, 
    <span class="va">sample</span>, 
    <span class="va">cell_group</span> 
  <span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">=</span>
  <span class="va">counts_obj</span> |&gt;
  <span class="fu"><a href="../reference/sccomp_glm.html">sccomp_glm</a></span><span class="op">(</span> 
    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, 
    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, 
    .sample <span class="op">=</span> <span class="va">sample</span>,
    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,
    .count <span class="op">=</span> <span class="va">count</span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## sccomp says: outlier identification first pass - step 1/3 [ETA: ~20s]</span></code></pre>
<pre><code><span class="co">## sccomp says: outlier identification second pass - step 2/3 [ETA: ~60s]</span></code></pre>
<pre><code><span class="co">## sccomp says: outlier-free model fitting - step 3/3 [ETA: ~20s]</span></code></pre>
<pre><code><span class="co">## sccomp says: the composition design matrix has columns: (Intercept), typecancer</span></code></pre>
<pre><code><span class="co">## sccomp says: the variability design matrix has columns: (Intercept)</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 72 × 9</span></span>
<span class="co">##    cell_group parameter   covariate c_lower c_effect c_upper  c_pH0   c_FDR</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercept) <span style="color: #BB0000;">NA</span>         0.544     0.702  0.862  0      0      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer  type      -<span style="color: #BB0000;">1.20</span>     -<span style="color: #BB0000;">0.893</span> -<span style="color: #BB0000;">0.590</span>  0      0      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercept) <span style="color: #BB0000;">NA</span>         0.114     0.367  0.622  0.092<span style="text-decoration: underline;">5</span> 0.004<span style="text-decoration: underline;">29</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer  type      -<span style="color: #BB0000;">1.11</span>     -<span style="color: #BB0000;">0.624</span> -<span style="color: #BB0000;">0.159</span>  0.037<span style="text-decoration: underline;">3</span> 0.004<span style="text-decoration: underline;">89</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercept) <span style="color: #BB0000;">NA</span>        -<span style="color: #BB0000;">0.606</span>    -<span style="color: #BB0000;">0.417</span> -<span style="color: #BB0000;">0.208</span>  0.020<span style="text-decoration: underline;">8</span> 0.001<span style="text-decoration: underline;">14</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer  type      -<span style="color: #BB0000;">0.581</span>    -<span style="color: #BB0000;">0.204</span>  0.171  0.491  0.137  </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercept) <span style="color: #BB0000;">NA</span>        -<span style="color: #BB0000;">1.33</span>     -<span style="color: #BB0000;">1.11</span>  -<span style="color: #BB0000;">0.890</span>  0      0      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer  type      -<span style="color: #BB0000;">0.754</span>    -<span style="color: #BB0000;">0.341</span>  0.058<span style="text-decoration: underline;">1</span> 0.235  0.066<span style="text-decoration: underline;">8</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercept) <span style="color: #BB0000;">NA</span>         0.349     0.483  0.617  0      0      </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer  type      -<span style="color: #BB0000;">0.075</span><span style="color: #BB0000; text-decoration: underline;">7</span>    0.166  0.402  0.615  0.193  </span>
<span class="co">## <span style="color: #949494;"># … with 62 more rows, and 1 more variable: count_data &lt;list&gt;</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualise-data-inference">Visualise data + inference<a class="anchor" aria-label="anchor" href="#visualise-data-inference"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## Joining, by = c("sample", "cell_group")</span>
<span class="co">## Joining, by = c("cell_group", "type")</span></code></pre>
<pre><code><span class="co">## Warning: Ignoring unknown aesthetics: label</span></code></pre>
<p>Plot of group proportion, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is likely be
descriptively adequate to the data, the blue boxplot should roughly
overlay with the black boxplot, which represent the observed data. The
outliers are coloured in red.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">$</span><span class="va">boxplot</span></code></pre></div>
<pre><code><span class="co">## [[1]]</span></code></pre>
<p><img src="inst/figures/unnamed-chunk-10-1.png" width="700"></p>
<p>Plot of estimates of differential composition (c_) on the x axis, and
differential variability (v_) on the y axis. The error bars represent
95% credible intervals. The dashed lines represent the minimal effect
that the hypothesis test is based on. An effect is labelled as
significant if bigger than the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior
distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that has converged and is negative with
probability 1.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="differential-variability">Differential variability<a class="anchor" aria-label="anchor" href="#differential-variability"></a>
</h3>
<p>We can model the cell-group variability also dependent on type, and
so test differences in variability</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">=</span> 
  <span class="va">counts_obj</span> |&gt;
  <span class="fu"><a href="../reference/sccomp_glm.html">sccomp_glm</a></span><span class="op">(</span> 
    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, 
    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, 
    .sample <span class="op">=</span> <span class="va">sample</span>,
    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,
    .count <span class="op">=</span> <span class="va">count</span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## sccomp says: outlier identification first pass - step 1/3 [ETA: ~20s]</span></code></pre>
<pre><code><span class="co">## sccomp says: outlier identification second pass - step 2/3 [ETA: ~60s]</span></code></pre>
<pre><code><span class="co">## sccomp says: outlier-free model fitting - step 3/3 [ETA: ~20s]</span></code></pre>
<pre><code><span class="co">## sccomp says: the composition design matrix has columns: (Intercept), typecancer</span></code></pre>
<pre><code><span class="co">## sccomp says: the variability design matrix has columns: (Intercept), typecancer</span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 72 × 14</span></span>
<span class="co">##    cell_group parameter   covariate c_lower c_effect c_upper  c_pH0    c_FDR</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercept) <span style="color: #BB0000;">NA</span>         0.584     0.748   0.912 0      0       </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer  type      -<span style="color: #BB0000;">1.17</span>     -<span style="color: #BB0000;">0.843</span>  -<span style="color: #BB0000;">0.510</span> 0      0       </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercept) <span style="color: #BB0000;">NA</span>         0.235     0.453   0.674 0.013<span style="text-decoration: underline;">2</span> 0.000<span style="text-decoration: underline;">786</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer  type      -<span style="color: #BB0000;">1.11</span>     -<span style="color: #BB0000;">0.648</span>  -<span style="color: #BB0000;">0.212</span> 0.022<span style="text-decoration: underline;">0</span> 0.003<span style="text-decoration: underline;">20</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercept) <span style="color: #BB0000;">NA</span>        -<span style="color: #BB0000;">0.617</span>    -<span style="color: #BB0000;">0.404</span>  -<span style="color: #BB0000;">0.172</span> 0.034<span style="text-decoration: underline;">7</span> 0.001<span style="text-decoration: underline;">96</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer  type      -<span style="color: #BB0000;">0.633</span>    -<span style="color: #BB0000;">0.201</span>   0.240 0.498  0.146   </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercept) <span style="color: #BB0000;">NA</span>        -<span style="color: #BB0000;">1.28</span>     -<span style="color: #BB0000;">1.05</span>   -<span style="color: #BB0000;">0.825</span> 0      0       </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer  type      -<span style="color: #BB0000;">0.738</span>    -<span style="color: #BB0000;">0.289</span>   0.155 0.350  0.107   </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercept) <span style="color: #BB0000;">NA</span>         0.387     0.531   0.683 0      0       </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer  type      -<span style="color: #BB0000;">0.043</span><span style="color: #BB0000; text-decoration: underline;">2</span>    0.241   0.523 0.379  0.118   </span>
<span class="co">## <span style="color: #949494;"># … with 62 more rows, and 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, count_data &lt;list&gt;</span></span></code></pre>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Joining, by = c("sample", "cell_group")</span>
<span class="co">## Joining, by = c("cell_group", "type")</span></code></pre>
<pre><code><span class="co">## Warning: Ignoring unknown aesthetics: label</span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-14-1.png" width="700"></p>
<p>Plot 2D significance plot. This is possible if only differential
variability has been tested</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></code></pre></div>
<p><img src="inst/figures/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] rstan_2.21.3         StanHeaders_2.21.0-7 tidyr_1.2.0         </span>
<span class="co">## [4] forcats_0.5.1        ggplot2_3.3.5        sccomp_0.99.20      </span>
<span class="co">## [7] dplyr_1.0.8          BiocStyle_2.22.0    </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] bitops_1.0-7                matrixStats_0.61.0         </span>
<span class="co">##  [3] fs_1.5.2                    RColorBrewer_1.1-2         </span>
<span class="co">##  [5] rprojroot_2.0.2             GenomeInfoDb_1.30.1        </span>
<span class="co">##  [7] tools_4.1.2                 bslib_0.3.1                </span>
<span class="co">##  [9] utf8_1.2.2                  R6_2.5.1                   </span>
<span class="co">## [11] BiocGenerics_0.40.0         colorspace_2.0-3           </span>
<span class="co">## [13] withr_2.5.0                 tidyselect_1.1.2           </span>
<span class="co">## [15] gridExtra_2.3               prettyunits_1.1.1          </span>
<span class="co">## [17] processx_3.5.2              compiler_4.1.2             </span>
<span class="co">## [19] textshaping_0.3.6           cli_3.2.0                  </span>
<span class="co">## [21] Biobase_2.54.0              desc_1.4.1                 </span>
<span class="co">## [23] DelayedArray_0.20.0         labeling_0.4.2             </span>
<span class="co">## [25] bookdown_0.24               sass_0.4.0                 </span>
<span class="co">## [27] scales_1.1.1                readr_2.1.2                </span>
<span class="co">## [29] callr_3.7.0                 pkgdown_2.0.2.9000         </span>
<span class="co">## [31] systemfonts_1.0.4           stringr_1.4.0              </span>
<span class="co">## [33] digest_0.6.29               rmarkdown_2.12             </span>
<span class="co">## [35] XVector_0.34.0              pkgconfig_2.0.3            </span>
<span class="co">## [37] htmltools_0.5.2             MatrixGenerics_1.6.0       </span>
<span class="co">## [39] highr_0.9                   fastmap_1.1.0              </span>
<span class="co">## [41] rlang_1.0.2                 farver_2.1.0               </span>
<span class="co">## [43] jquerylib_0.1.4             generics_0.1.2             </span>
<span class="co">## [45] jsonlite_1.8.0              inline_0.3.19              </span>
<span class="co">## [47] RCurl_1.98-1.6              magrittr_2.0.2             </span>
<span class="co">## [49] GenomeInfoDbData_1.2.7      loo_2.4.1                  </span>
<span class="co">## [51] patchwork_1.1.1             Matrix_1.4-0               </span>
<span class="co">## [53] Rcpp_1.0.8                  munsell_0.5.0              </span>
<span class="co">## [55] S4Vectors_0.32.3            fansi_1.0.2                </span>
<span class="co">## [57] lifecycle_1.0.1             stringi_1.7.6              </span>
<span class="co">## [59] yaml_2.3.5                  SummarizedExperiment_1.24.0</span>
<span class="co">## [61] zlibbioc_1.40.0             pkgbuild_1.3.1             </span>
<span class="co">## [63] grid_4.1.2                  parallel_4.1.2             </span>
<span class="co">## [65] ggrepel_0.9.1               crayon_1.5.0               </span>
<span class="co">## [67] lattice_0.20-45             hms_1.1.1                  </span>
<span class="co">## [69] knitr_1.37                  ps_1.6.0                   </span>
<span class="co">## [71] pillar_1.7.0                GenomicRanges_1.46.1       </span>
<span class="co">## [73] boot_1.3-28                 codetools_0.2-18           </span>
<span class="co">## [75] stats4_4.1.2                rstantools_2.1.1           </span>
<span class="co">## [77] glue_1.6.2                  evaluate_0.15              </span>
<span class="co">## [79] SeuratObject_4.0.4          BiocManager_1.30.16        </span>
<span class="co">## [81] RcppParallel_5.1.5          vctrs_0.3.8                </span>
<span class="co">## [83] tzdb_0.2.0                  gtable_0.3.0               </span>
<span class="co">## [85] purrr_0.3.4                 cachem_1.0.6               </span>
<span class="co">## [87] xfun_0.30                   ragg_1.2.2                 </span>
<span class="co">## [89] SingleCellExperiment_1.16.0 tibble_3.1.6               </span>
<span class="co">## [91] memoise_2.0.1               IRanges_2.28.0             </span>
<span class="co">## [93] ellipsis_0.3.2</span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
