<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the sccomp package â€¢ sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the sccomp package">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the sccomp package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2024-08-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/calculate_proportion_difference/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Sccomp is a generalised method for differential composition and
      variability analyses. It tests differences in cell type
      proportion, for count data. It is able of flexible
      continuous/discrete models, removal of unwanted variation, removal
      of outliers, multilevel random/effect modelling.</p>
    </div>
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<p>Sccomp is a generalised method for differential composition and
variability analyses.</p>
<div class="section level3">
<h3 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h3>
<ul>
<li>Modelling counts</li>
<li>Modelling proportionality</li>
<li>Modelling cell-type specific variability</li>
<li>Cell-type information share for variability shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p><code>sccomp</code> can model changes in composition and variability.
By default, the formula for variability is either <code>~1</code>, which
assumes that the cell-group variability is independent of any covariate
or <code>~ factor_of_interest</code>, which assumes that the model is
dependent on the factor of interest only. The variability model must be
a subset of the model for composition.</p>
<div class="section level3">
<h3 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h3>
<p>Of the output table, the estimate columns start with the prefix
<code>c_</code> indicate <code>composition</code>, or with
<code>v_</code> indicate <code>variability</code> (when
formula_variability is set).</p>
<div class="section level4">
<h4 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">single_cell_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000398 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.98 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.971 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                20.878 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                23.849 seconds (Total)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000335 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.35 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:     1 / 20299 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:   301 / 20299 [  1%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  1300 / 20299 [  6%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  2300 / 20299 [ 11%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  3300 / 20299 [ 16%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  4300 / 20299 [ 21%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  5300 / 20299 [ 26%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  6300 / 20299 [ 31%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  7300 / 20299 [ 35%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  8300 / 20299 [ 40%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  9300 / 20299 [ 45%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 10300 / 20299 [ 50%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 11300 / 20299 [ 55%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 12300 / 20299 [ 60%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 13300 / 20299 [ 65%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 14300 / 20299 [ 70%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 15300 / 20299 [ 75%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 16300 / 20299 [ 80%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 17300 / 20299 [ 85%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 18300 / 20299 [ 90%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 19300 / 20299 [ 95%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 20299 / 20299 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 3.629 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                125.488 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                129.117 seconds (Total)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000345 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.45 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 3.122 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                30.461 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                33.583 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<p>Here you see the results of the fit, the effects of the factor on
composition and variability. You also can see the uncertainty around
those effects.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 72 Ã— 18</span></span></span>
<span><span class="co">##    cell_group parameter  factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercepâ€¦ <span style="color: #BB0000;">NA</span>      0.896     1.11   1.31   0       0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">4</span>555.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer type   -<span style="color: #BB0000;">1.30</span>     -<span style="color: #BB0000;">0.899</span> -<span style="color: #BB0000;">0.536</span>  0       0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">2</span>855.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercepâ€¦ <span style="color: #BB0000;">NA</span>      0.414     0.703  0.983  0       0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">4</span>570.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer type   -<span style="color: #BB0000;">1.22</span>     -<span style="color: #BB0000;">0.724</span> -<span style="color: #BB0000;">0.270</span>  0.004<span style="text-decoration: underline;">50</span> 6.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>   <span style="text-decoration: underline;">3</span>063.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercepâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.663</span>    -<span style="color: #BB0000;">0.390</span> -<span style="color: #BB0000;">0.120</span>  0.017<span style="text-decoration: underline;">2</span>  1.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>   <span style="text-decoration: underline;">4</span>359.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer type   -<span style="color: #BB0000;">0.719</span>    -<span style="color: #BB0000;">0.310</span>  0.080<span style="text-decoration: underline;">6</span> 0.156   4.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>   <span style="text-decoration: underline;">3</span>535.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercepâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.32</span>     -<span style="color: #BB0000;">1.03</span>  -<span style="color: #BB0000;">0.753</span>  0       0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">4</span>337.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer type   -<span style="color: #BB0000;">0.745</span>    -<span style="color: #BB0000;">0.316</span>  0.097<span style="text-decoration: underline;">9</span> 0.146   3.84<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>   <span style="text-decoration: underline;">3</span>714.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercepâ€¦ <span style="color: #BB0000;">NA</span>      0.089<span style="text-decoration: underline;">5</span>    0.302  0.498  0.028<span style="text-decoration: underline;">5</span>  3.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>   <span style="text-decoration: underline;">2</span>991.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer type   -<span style="color: #BB0000;">0.096</span><span style="color: #BB0000; text-decoration: underline;">8</span>    0.181  0.471  0.282   6.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>   <span style="text-decoration: underline;">2</span>613.</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 62 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="an-aid-to-result-interpretation-and-communication">An aid to result interpretation and communication<a class="anchor" aria-label="anchor" href="#an-aid-to-result-interpretation-and-communication"></a>
</h3>
<p>The estimated effects are expressed in the unconstrained space of the
parameters. Similarly, to differential expression analysis that express
change in terms of log fold change. However, for differences, in
proportion, logit foold change must be used. This measure is harder to
interpret and understand.</p>
<p>Therefore, we provide a more intuitive proportion, full change, that
can be easier understood. However, these cannot be used to infer
significance (use sccomp_test() instead), and a lot of care must be
taken given the nonlinearity of these measure (1 fold increase from
0.0001 to 0.0002 carried a different weight that 1 fold increase from
0.4 to 0.8).</p>
<p>From your estimates, you can state which effects you are interested
about (this can be a part of the full model, in case you want to not
consider unwanted effects), and the two points you would like to
compare.</p>
<p>In case of a chategorical variable, the starting and ending points
are categories.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu"><a href="../reference/sccomp_proportional_fold_change.html">sccomp_proportional_fold_change</a></span><span class="op">(</span></span>
<span>     formula_composition <span class="op">=</span> <span class="op">~</span>  <span class="va">type</span>,</span>
<span>     from <span class="op">=</span>  <span class="st">"healthy"</span>, </span>
<span>     to <span class="op">=</span> <span class="st">"cancer"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_group</span>, <span class="va">statement</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 36 Ã— 2</span></span></span>
<span><span class="co">##    cell_group statement                                </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;glue&gt;</span>                                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         2.5-fold decrease (from 0.0565 to 0.0227)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B2         2.1-fold decrease (from 0.0376 to 0.0182)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B3         1.4-fold decrease (from 0.0125 to 0.0091)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> BM         1.4-fold decrease (from 0.0067 to 0.0048)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 1      1.2-fold increase (from 0.0251 to 0.03)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 2      1.5-fold increase (from 0.0495 to 0.074) </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 3      2.3-fold decrease (from 0.0787 to 0.0343)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 4      1-fold decrease (from 0.0016 to 0.0016)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 5      1-fold increase (from 0.0301 to 0.0313)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD8 1      1.1-fold increase (from 0.107 to 0.1227) </span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 26 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h3>
<p>A plot of group proportion, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is likely to be
descriptively adequate to the data, the blue box plot should roughly
overlay with the black box plot, which represents the observed data. The
outliers are coloured in red. A box plot will be returned for every
(discrete) covariate present in <code>formula_composition</code>. The
colour coding represents the significant associations for composition
and/or variability.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(cell_group, sample)`</span></span>
<span><span class="co">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>A plot of estimates of differential composition (c_) on the x-axis
and differential variability (v_) on the y-axis. The error bars
represent 95% credible intervals. The dashed lines represent the minimal
effect that the hypothesis test is based on. An effect is labelled as
significant if bigger than the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/plot_1d_intervals.html">plot_1D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>We can plot the relationship between abundance and variability. As we
can see below, they are positively correlated, you also appreciate that
this relationship is by model for single cell RNA sequencing data.</p>
<p><code>sccomp</code> models, these relationship to obtain a shrinkage
effect on the estimates of both the abundance and the variability. This
shrinkage is adaptive as it is modelled jointly, thanks for Bayesian
inference.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/plot_2d_intervals.html">plot_2D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>You can produce the series of plots calling the <code>plot</code>
method.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000286 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 2.86 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.211 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                18.438 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                20.649 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 Ã— 18</span></span></span>
<span><span class="co">##    cell_group  parameter factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature  typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.91</span>    -<span style="color: #BB0000;">1.39</span>   -<span style="color: #BB0000;">0.884</span> 0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature  typehealâ€¦ <span style="color: #BB0000;">NA</span>       0.884    1.39    1.91  0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem       typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.36</span>    -<span style="color: #BB0000;">1.70</span>   -<span style="color: #BB0000;">1.05</span>  0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem       typehealâ€¦ <span style="color: #BB0000;">NA</span>       1.05     1.70    2.36  0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S10â€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.48</span>    -<span style="color: #BB0000;">1.03</span>   -<span style="color: #BB0000;">0.596</span> 0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S10â€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>       0.596    1.03    1.48  0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm higâ€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>       0.857    1.76    2.85  0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm higâ€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.85</span>    -<span style="color: #BB0000;">1.76</span>   -<span style="color: #BB0000;">0.857</span> 0       0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribâ€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>       0.301    0.988   1.71  0.004<span style="text-decoration: underline;">75</span> 0.001<span style="text-decoration: underline;">11</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribâ€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.71</span>    -<span style="color: #BB0000;">0.988</span>  -<span style="color: #BB0000;">0.301</span> 0.004<span style="text-decoration: underline;">75</span> 0.001<span style="text-decoration: underline;">11</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="categorical-factor-e-g--bayesian-anova">Categorical factor (e.g.Â Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-e-g--bayesian-anova"></a>
</h3>
<p>This is achieved through model comparison with <code>loo</code>. In
the following example, the model with association with factors better
fits the data compared to the baseline model with no factor association.
For comparisons <code>check_outliers</code> must be set to FALSE as the
leave-one-out must work with the same amount of data, while outlier
elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5
<code>se_diff</code> difference of 5, we are confident that a model is
better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>.
In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model
one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> , </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>  <span class="va">model_with_factor_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">model_without_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h3>
<p>We can model the cell-group variability also dependent on the type,
and so test differences in variability</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000444 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 4.44 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 4.114 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                34.616 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                38.73 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 Ã— 14</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interceâ€¦ <span style="color: #BB0000;">NA</span>      0.351     0.771  1.18     <span style="text-decoration: underline;">5</span>518.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typehealâ€¦ type    0.889     1.43   2.01     <span style="text-decoration: underline;">4</span>662.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem             (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.46</span>     -<span style="color: #BB0000;">0.860</span> -<span style="color: #BB0000;">0.136</span>    <span style="text-decoration: underline;">3</span>144.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             typehealâ€¦ type    1.02      1.84   2.58     <span style="text-decoration: underline;">3</span>006.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4     (Interceâ€¦ <span style="color: #BB0000;">NA</span>      1.31      1.66   1.99     <span style="text-decoration: underline;">6</span>243.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100A4     typehealâ€¦ type    0.503     0.942  1.39     <span style="text-decoration: underline;">4</span>018.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high cytoâ€¦ (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.06</span>     -<span style="color: #BB0000;">0.550</span>  0.015<span style="text-decoration: underline;">4</span>   <span style="text-decoration: underline;">4</span>111.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high cytoâ€¦ typehealâ€¦ type   -<span style="color: #BB0000;">3.03</span>     -<span style="color: #BB0000;">1.23</span>   1.25     <span style="text-decoration: underline;">2</span>955.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribosome   (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.073</span><span style="color: #BB0000; text-decoration: underline;">7</span>    0.308  0.702    <span style="text-decoration: underline;">3</span>944.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribosome   typehealâ€¦ type   -<span style="color: #BB0000;">1.84</span>     -<span style="color: #BB0000;">0.968</span>  0.084<span style="text-decoration: underline;">7</span>   <span style="text-decoration: underline;">3</span>352.      1.00</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h2>
<div class="section level3">
<h3 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = TRUE</code>. The
bimodality of the mean-variability association can be confirmed from the
plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level3">
<h3 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = FALSE</code>
(Default).</p>
</div>
<div class="section level3">
<h3 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior
distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example, we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that it has converged and is negative
with probability 1.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(cell_group, sample)`</span></span>
<span><span class="co">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Plot 2D significance plot. Data points are cell groups. Error bars
are the 95% credible interval. The dashed lines represent the default
threshold fold change for which the probabilities (c_pH0, v_pH0) are
calculated. pH0 of 0 represent the rejection of the null hypothesis that
no effect is observed.</p>
<p>This plot is provided only if differential variability has been
tested. The differential variability estimates are reliable only if the
linear association between mean and variability for
<code>(intercept)</code> (left-hand side facet) is satisfied. A
scatterplot (besides the Intercept) is provided for each category of
interest. The for each category of interest, the composition and
variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rstan_2.32.6        StanHeaders_2.32.10 tidyr_1.3.1        </span></span>
<span><span class="co">## [4] forcats_1.0.0       ggplot2_3.5.1       sccomp_1.7.15      </span></span>
<span><span class="co">## [7] dplyr_1.1.4        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] gridExtra_2.3               inline_0.3.19              </span></span>
<span><span class="co">##   [3] rlang_1.1.4                 magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] matrixStats_1.3.0           compiler_4.4.1             </span></span>
<span><span class="co">##   [7] loo_2.8.0                   systemfonts_1.1.0          </span></span>
<span><span class="co">##   [9] vctrs_0.6.5                 stringr_1.5.1              </span></span>
<span><span class="co">##  [11] pkgconfig_2.0.3             crayon_1.5.3               </span></span>
<span><span class="co">##  [13] fastmap_1.2.0               XVector_0.45.0             </span></span>
<span><span class="co">##  [15] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">##  [17] rmarkdown_2.28              tzdb_0.4.0                 </span></span>
<span><span class="co">##  [19] UCSC.utils_1.1.0            ragg_1.3.2                 </span></span>
<span><span class="co">##  [21] purrr_1.0.2                 xfun_0.47                  </span></span>
<span><span class="co">##  [23] zlibbioc_1.51.1             cachem_1.1.0               </span></span>
<span><span class="co">##  [25] GenomeInfoDb_1.41.1         jsonlite_1.8.8             </span></span>
<span><span class="co">##  [27] highr_0.11                  DelayedArray_0.31.11       </span></span>
<span><span class="co">##  [29] parallel_4.4.1              R6_2.5.1                   </span></span>
<span><span class="co">##  [31] RColorBrewer_1.1-3          bslib_0.8.0                </span></span>
<span><span class="co">##  [33] stringi_1.8.4               parallelly_1.38.0          </span></span>
<span><span class="co">##  [35] boot_1.3-30                 GenomicRanges_1.57.1       </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4             Rcpp_1.0.13                </span></span>
<span><span class="co">##  [39] SummarizedExperiment_1.35.1 knitr_1.48                 </span></span>
<span><span class="co">##  [41] future.apply_1.11.2         readr_2.1.5                </span></span>
<span><span class="co">##  [43] IRanges_2.39.2              Matrix_1.7-0               </span></span>
<span><span class="co">##  [45] tidyselect_1.2.1            abind_1.4-5                </span></span>
<span><span class="co">##  [47] yaml_2.3.10                 codetools_0.2-20           </span></span>
<span><span class="co">##  [49] curl_5.2.1                  listenv_0.9.1              </span></span>
<span><span class="co">##  [51] pkgbuild_1.4.4              lattice_0.22-6             </span></span>
<span><span class="co">##  [53] tibble_3.2.1                withr_3.0.1                </span></span>
<span><span class="co">##  [55] Biobase_2.65.0              evaluate_0.24.0            </span></span>
<span><span class="co">##  [57] future_1.34.0               desc_1.4.3                 </span></span>
<span><span class="co">##  [59] RcppParallel_5.1.9          pillar_1.9.0               </span></span>
<span><span class="co">##  [61] MatrixGenerics_1.17.0       stats4_4.4.1               </span></span>
<span><span class="co">##  [63] generics_0.1.3              sp_2.1-4                   </span></span>
<span><span class="co">##  [65] S4Vectors_0.43.2            hms_1.1.3                  </span></span>
<span><span class="co">##  [67] rstantools_2.4.0            munsell_0.5.1              </span></span>
<span><span class="co">##  [69] scales_1.3.0                globals_0.16.3             </span></span>
<span><span class="co">##  [71] glue_1.7.0                  tools_4.4.1                </span></span>
<span><span class="co">##  [73] fs_1.6.4                    dotCall64_1.1-1            </span></span>
<span><span class="co">##  [75] grid_4.4.1                  QuickJSR_1.3.1             </span></span>
<span><span class="co">##  [77] colorspace_2.1-1            SingleCellExperiment_1.27.2</span></span>
<span><span class="co">##  [79] GenomeInfoDbData_1.2.12     patchwork_1.2.0            </span></span>
<span><span class="co">##  [81] cli_3.6.3                   textshaping_0.4.0          </span></span>
<span><span class="co">##  [83] spam_2.10-0                 fansi_1.0.6                </span></span>
<span><span class="co">##  [85] S4Arrays_1.5.7              V8_5.0.0                   </span></span>
<span><span class="co">##  [87] gtable_0.3.5                sass_0.4.9                 </span></span>
<span><span class="co">##  [89] digest_0.6.37               progressr_0.14.0           </span></span>
<span><span class="co">##  [91] BiocGenerics_0.51.0         SparseArray_1.5.31         </span></span>
<span><span class="co">##  [93] ggrepel_0.9.5               farver_2.1.2               </span></span>
<span><span class="co">##  [95] htmlwidgets_1.6.4           SeuratObject_5.0.2         </span></span>
<span><span class="co">##  [97] htmltools_0.5.8.1           pkgdown_2.1.0              </span></span>
<span><span class="co">##  [99] lifecycle_1.0.4             prettydoc_0.4.1            </span></span>
<span><span class="co">## [101] httr_1.4.7</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
