<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the sccomp package • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the sccomp package">
<meta property="og:description" content="sccomp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
    <li>
      <a href="../articles/new_framework.html">New sccomp framework</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the sccomp package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2024-03-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<div class="section level2">
<h2 id="section">
<img src="inst/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h2>
<p>Sccomp is a generalised method for differential composition and
variability analyses.</p>
<div class="section level3">
<h3 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h3>
<ul>
<li>Modelling counts</li>
<li>Modelling proportionality</li>
<li>Modelling cell-type specific variability</li>
<li>Cell-type information share for variability shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p><code>sccomp</code> can model changes in composition and variability.
By default, the formula for variability is either <code>~1</code>, which
assumes that the cell-group variability is independent of any covariate
or <code>~ factor_of_interest</code>, which assumes that the model is
dependent on the factor of interest only. The variability model must be
a subset of the model for composition.</p>
<div class="section level3">
<h3 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h3>
<div class="section level4">
<h4 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000471 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 4.71 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 3.151 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                21.391 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                24.542 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 72 × 14</span></span></span>
<span><span class="co">##    cell_group parameter   factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercept) <span style="color: #BB0000;">NA</span>      0.855     1.12   1.35     <span style="text-decoration: underline;">4</span>512.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer  type   -<span style="color: #BB0000;">1.06</span>     -<span style="color: #BB0000;">0.642</span> -<span style="color: #BB0000;">0.247</span>    <span style="text-decoration: underline;">3</span>213.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercept) <span style="color: #BB0000;">NA</span>      0.411     0.703  0.981    <span style="text-decoration: underline;">4</span>427.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer  type   -<span style="color: #BB0000;">1.19</span>     -<span style="color: #BB0000;">0.721</span> -<span style="color: #BB0000;">0.267</span>    <span style="text-decoration: underline;">3</span>631.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercept) <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.674</span>    -<span style="color: #BB0000;">0.396</span> -<span style="color: #BB0000;">0.131</span>    <span style="text-decoration: underline;">4</span>613.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer  type   -<span style="color: #BB0000;">0.732</span>    -<span style="color: #BB0000;">0.308</span>  0.085<span style="text-decoration: underline;">1</span>   <span style="text-decoration: underline;">5</span>557.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercept) <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.31</span>     -<span style="color: #BB0000;">1.03</span>  -<span style="color: #BB0000;">0.754</span>    <span style="text-decoration: underline;">5</span>231.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer  type   -<span style="color: #BB0000;">0.755</span>    -<span style="color: #BB0000;">0.316</span>  0.085<span style="text-decoration: underline;">9</span>   <span style="text-decoration: underline;">4</span>698.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercept) <span style="color: #BB0000;">NA</span>      0.068<span style="text-decoration: underline;">9</span>    0.297  0.501    <span style="text-decoration: underline;">3</span>790.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer  type   -<span style="color: #BB0000;">0.105</span>     0.188  0.480    <span style="text-decoration: underline;">2</span>502.      1.00</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 62 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
<p>Of the output table, the estimate columns start with the prefix
<code>c_</code> indicate <code>composition</code>, or with
<code>v_</code> indicate <code>variability</code> (when
formula_variability is set).</p>
</div>
</div>
<div class="section level3">
<h3 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000306 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.06 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.081 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                17.516 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                19.597 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 18</span></span></span>
<span><span class="co">##    cell_group   parameter factor c_lower c_effect c_upper  c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature   typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.90</span>    -<span style="color: #BB0000;">1.40</span>   -<span style="color: #BB0000;">0.901</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature   typeheal… <span style="color: #BB0000;">NA</span>       0.901    1.40    1.90  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem        typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.33</span>    -<span style="color: #BB0000;">1.71</span>   -<span style="color: #BB0000;">1.06</span>  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem        typeheal… <span style="color: #BB0000;">NA</span>       1.06     1.71    2.33  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100… typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.49</span>    -<span style="color: #BB0000;">1.03</span>   -<span style="color: #BB0000;">0.608</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100… typeheal… <span style="color: #BB0000;">NA</span>       0.608    1.03    1.49  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high… typecanc… <span style="color: #BB0000;">NA</span>       0.814    1.75    2.85  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.85</span>    -<span style="color: #BB0000;">1.75</span>   -<span style="color: #BB0000;">0.814</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribo… typecanc… <span style="color: #BB0000;">NA</span>       0.319    0.999   1.71  0.011<span style="text-decoration: underline;">0</span> 0.003<span style="text-decoration: underline;">50</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribo… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.71</span>    -<span style="color: #BB0000;">0.999</span>  -<span style="color: #BB0000;">0.319</span> 0.011<span style="text-decoration: underline;">0</span> 0.003<span style="text-decoration: underline;">50</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="categorical-factor-e-g--bayesian-anova">Categorical factor (e.g. Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-e-g--bayesian-anova"></a>
</h3>
<p>This is achieved through model comparison with <code>loo</code>. In
the following example, the model with association with factors better
fits the data compared to the baseline model with no factor association.
For comparisons <code>check_outliers</code> must be set to FALSE as the
leave-one-out must work with the same amount of data, while outlier
elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5
<code>se_diff</code> difference of 5, we are confident that a model is
better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>.
In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model
one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    check_outliers <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    check_outliers <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> , </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>  <span class="va">model_with_factor_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">model_without_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h3>
<p>We can model the cell-group variability also dependent on the type,
and so test differences in variability</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000292 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 2.92 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 4.24 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                34.389 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                38.629 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 14</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interce… <span style="color: #BB0000;">NA</span>      0.361     0.761  1.19     <span style="text-decoration: underline;">5</span>556.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typeheal… type    0.847     1.43   1.98     <span style="text-decoration: underline;">4</span>129.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem             (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.49</span>     -<span style="color: #BB0000;">0.871</span> -<span style="color: #BB0000;">0.176</span>    <span style="text-decoration: underline;">4</span>279.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             typeheal… type    1.04      1.85   2.63     <span style="text-decoration: underline;">3</span>985.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4     (Interce… <span style="color: #BB0000;">NA</span>      1.32      1.66   1.99     <span style="text-decoration: underline;">5</span>809.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100A4     typeheal… type    0.482     0.933  1.41     <span style="text-decoration: underline;">4</span>900.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high cyto… (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.06</span>     -<span style="color: #BB0000;">0.546</span>  0.032<span style="text-decoration: underline;">6</span>   <span style="text-decoration: underline;">4</span>469.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high cyto… typeheal… type   -<span style="color: #BB0000;">3.05</span>     -<span style="color: #BB0000;">1.24</span>   1.21     <span style="text-decoration: underline;">3</span>077.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribosome   (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.056</span><span style="color: #BB0000; text-decoration: underline;">2</span>    0.312  0.694    <span style="text-decoration: underline;">4</span>096.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribosome   typeheal… type   -<span style="color: #BB0000;">1.80</span>     -<span style="color: #BB0000;">0.963</span>  0.073<span style="text-decoration: underline;">2</span>   <span style="text-decoration: underline;">3</span>480.      1.00</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h2>
<div class="section level3">
<h3 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = TRUE</code>. The
bimodality of the mean-variability association can be confirmed from the
plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level3">
<h3 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = FALSE</code>
(Default).</p>
</div>
</div>
<div class="section level2">
<h2 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a>
</h2>
<div class="section level3">
<h3 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Warning: `plot_summary()` was deprecated in sccomp 1.7.1.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> sccomp says: plot_summary() is soft-deprecated. Please use sccomp_test().</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">##              cell_group   parameter factor      c_lower      c_effect</span></span>
<span><span class="co">## 1            B immature (Intercept)   &lt;NA&gt;  0.361335750  0.7610843787</span></span>
<span><span class="co">## 2            B immature typehealthy   type  0.846645226  1.4313448240</span></span>
<span><span class="co">## 3                 B mem (Intercept)   &lt;NA&gt; -1.491667363 -0.8705328282</span></span>
<span><span class="co">## 4                 B mem typehealthy   type  1.044621292  1.8464576110</span></span>
<span><span class="co">## 5         CD4 cm S100A4 (Intercept)   &lt;NA&gt;  1.315297759  1.6626472418</span></span>
<span><span class="co">## 6         CD4 cm S100A4 typehealthy   type  0.481536552  0.9328551417</span></span>
<span><span class="co">## 7  CD4 cm high cytokine (Intercept)   &lt;NA&gt; -1.055981798 -0.5455480513</span></span>
<span><span class="co">## 8  CD4 cm high cytokine typehealthy   type -3.046043839 -1.2425233937</span></span>
<span><span class="co">## 9       CD4 cm ribosome (Intercept)   &lt;NA&gt; -0.056237906  0.3117464682</span></span>
<span><span class="co">## 10      CD4 cm ribosome typehealthy   type -1.795418062 -0.9629447717</span></span>
<span><span class="co">## 11 CD4 em high cytokine (Intercept)   &lt;NA&gt; -1.160495355 -0.7124512091</span></span>
<span><span class="co">## 12 CD4 em high cytokine typehealthy   type -2.392637902 -1.3041040598</span></span>
<span><span class="co">## 13            CD4 naive (Intercept)   &lt;NA&gt;  0.789746901  1.2098129266</span></span>
<span><span class="co">## 14            CD4 naive typehealthy   type  0.455463569  1.1419744399</span></span>
<span><span class="co">## 15         CD4 ribosome (Intercept)   &lt;NA&gt; -0.316511321  0.0449776945</span></span>
<span><span class="co">## 16         CD4 ribosome typehealthy   type  1.465844336  2.0692814529</span></span>
<span><span class="co">## 17             CD8 em 1 (Intercept)   &lt;NA&gt;  0.305004180  0.6774084176</span></span>
<span><span class="co">## 18             CD8 em 1 typehealthy   type -0.056085036  0.7796162553</span></span>
<span><span class="co">## 19             CD8 em 2 (Intercept)   &lt;NA&gt; -0.852322829  0.1319217505</span></span>
<span><span class="co">## 20             CD8 em 2 typehealthy   type -3.495939955 -1.1332514627</span></span>
<span><span class="co">## 21             CD8 em 3 (Intercept)   &lt;NA&gt; -1.376888305 -0.6587100309</span></span>
<span><span class="co">## 22             CD8 em 3 typehealthy   type -1.761624627 -0.8220066434</span></span>
<span><span class="co">## 23            CD8 naive (Intercept)   &lt;NA&gt;  0.612878008  1.0227113279</span></span>
<span><span class="co">## 24            CD8 naive typehealthy   type -0.730622098  0.0081362943</span></span>
<span><span class="co">## 25     CD8 transitional (Intercept)   &lt;NA&gt;  0.921485057  1.2122637013</span></span>
<span><span class="co">## 26     CD8 transitional typehealthy   type  0.196110474  0.6797219648</span></span>
<span><span class="co">## 27                 MAIT (Intercept)   &lt;NA&gt; -1.063688975 -0.5958567156</span></span>
<span><span class="co">## 28                 MAIT typehealthy   type  0.457965724  1.2146536073</span></span>
<span><span class="co">## 29               Mac M1 (Intercept)   &lt;NA&gt;  0.620993651  0.9469437283</span></span>
<span><span class="co">## 30               Mac M1 typehealthy   type  0.009944924  0.5952306473</span></span>
<span><span class="co">## 31        Mac M1 FCER1A (Intercept)   &lt;NA&gt; -1.352662954 -0.9805367919</span></span>
<span><span class="co">## 32        Mac M1 FCER1A typehealthy   type -1.814850080 -1.0779609819</span></span>
<span><span class="co">## 33                 Mono (Intercept)   &lt;NA&gt; -1.808269928 -1.3396240981</span></span>
<span><span class="co">## 34                 Mono typehealthy   type -0.858907732 -0.1001649072</span></span>
<span><span class="co">## 35          Mono NKG7 1 (Intercept)   &lt;NA&gt; -1.507486023 -0.8651796037</span></span>
<span><span class="co">## 36          Mono NKG7 1 typehealthy   type -3.991498845 -2.8799693881</span></span>
<span><span class="co">## 37          Mono NKG7 2 (Intercept)   &lt;NA&gt; -0.413092407 -0.0006123920</span></span>
<span><span class="co">## 38          Mono NKG7 2 typehealthy   type -2.438862982 -1.2690211681</span></span>
<span><span class="co">## 39  Mono classic inflam (Intercept)   &lt;NA&gt;  1.491381700  1.7882557952</span></span>
<span><span class="co">## 40  Mono classic inflam typehealthy   type -1.253629418 -0.4731128230</span></span>
<span><span class="co">## 41     Mono non-classic (Intercept)   &lt;NA&gt;  0.616986405  0.9672676712</span></span>
<span><span class="co">## 42     Mono non-classic typehealthy   type -0.628262840  0.0008427117</span></span>
<span><span class="co">## 43    Myeloid migratory (Intercept)   &lt;NA&gt; -0.359035184  0.0147033416</span></span>
<span><span class="co">## 44    Myeloid migratory typehealthy   type -1.327486243 -0.5464070029</span></span>
<span><span class="co">## 45                   NK (Intercept)   &lt;NA&gt;  1.135281829  1.4325018510</span></span>
<span><span class="co">## 46                   NK typehealthy   type -0.053131837  0.5443258722</span></span>
<span><span class="co">## 47           NK cycling (Intercept)   &lt;NA&gt; -1.654773213 -1.2153089284</span></span>
<span><span class="co">## 48           NK cycling typehealthy   type -0.811196824 -0.1805742824</span></span>
<span><span class="co">## 49     NK high cytokine (Intercept)   &lt;NA&gt; -2.191038253 -1.6320773684</span></span>
<span><span class="co">## 50     NK high cytokine typehealthy   type -1.194242846  0.0456987120</span></span>
<span><span class="co">## 51                 Stem (Intercept)   &lt;NA&gt; -1.092679243 -0.5750920505</span></span>
<span><span class="co">## 52                 Stem typehealthy   type -3.213392800 -2.2399877567</span></span>
<span><span class="co">## 53                T gd1 (Intercept)   &lt;NA&gt; -1.073860129 -0.6898489711</span></span>
<span><span class="co">## 54                T gd1 typehealthy   type  0.040046344  0.9267160202</span></span>
<span><span class="co">## 55                T gd2 (Intercept)   &lt;NA&gt; -1.238421686 -0.6644650242</span></span>
<span><span class="co">## 56                T gd2 typehealthy   type -0.392518985  0.5164781590</span></span>
<span><span class="co">## 57                 cDC2 (Intercept)   &lt;NA&gt; -0.786043102 -0.2921548064</span></span>
<span><span class="co">## 58                 cDC2 typehealthy   type -0.380050576  0.3243495149</span></span>
<span><span class="co">## 59                  pDC (Intercept)   &lt;NA&gt; -1.148425930 -0.6763980865</span></span>
<span><span class="co">## 60                  pDC typehealthy   type -0.089065820  0.6732108058</span></span>
<span><span class="co">##        c_upper   c_pH0        c_FDR  c_n_eff c_R_k_hat     v_lower     v_effect</span></span>
<span><span class="co">## 1   1.19082303 0.00325 4.230769e-04 5555.660 0.9998081 -4.49186224 -3.827272207</span></span>
<span><span class="co">## 2   1.97751054 0.00000 0.000000e+00 4128.725 0.9997506 -1.67274410 -0.490730778</span></span>
<span><span class="co">## 3  -0.17554074 0.02975 4.779412e-03 4279.296 1.0001883 -5.30318509 -4.565244119</span></span>
<span><span class="co">## 4   2.63245560 0.00000 0.000000e+00 3984.925 1.0005908 -2.06742753 -0.939821112</span></span>
<span><span class="co">## 5   1.98510366 0.00000 0.000000e+00 5809.372 1.0000078 -3.83922207 -3.202783354</span></span>
<span><span class="co">## 6   1.40721764 0.00250 6.250000e-04 4899.517 0.9998172 -2.05632280 -0.929137943</span></span>
<span><span class="co">## 7   0.03259866 0.10925 2.035870e-02 4469.169 0.9999394 -5.47834443 -4.712033028</span></span>
<span><span class="co">## 8   1.20715576 0.19725 5.845238e-02 3076.835 1.0009504 -0.43521243  1.531293936</span></span>
<span><span class="co">## 9   0.69435529 0.27075 3.444000e-02 4096.081 0.9997675 -5.35672733 -4.604764084</span></span>
<span><span class="co">## 10  0.07316059 0.06575 2.003846e-02 3480.198 1.0002291 -0.55464411  0.651979465</span></span>
<span><span class="co">## 11 -0.18310523 0.02900 3.218750e-03 4094.068 1.0000791 -5.93410382 -5.149613287</span></span>
<span><span class="co">## 12  0.23926880 0.06225 1.622917e-02 3247.258 0.9997584 -1.11260720  0.350654782</span></span>
<span><span class="co">## 13  1.65239493 0.00000 0.000000e+00 5381.033 0.9997778 -3.66736227 -3.060646053</span></span>
<span><span class="co">## 14  1.80123725 0.00425 1.350000e-03 5219.699 1.0004124 -0.83812993  0.200559741</span></span>
<span><span class="co">## 15  0.43244777 0.80750 9.155357e-02 5412.260 0.9997515 -5.68558806 -4.889662576</span></span>
<span><span class="co">## 16  2.63076283 0.00000 0.000000e+00 6213.848 0.9997607 -0.68062978  0.557819163</span></span>
<span><span class="co">## 17  1.08720539 0.00625 8.392857e-04 4342.332 0.9997510 -4.78177096 -4.069967955</span></span>
<span><span class="co">## 18  1.58261272 0.08250 2.808333e-02 4335.443 1.0000313 -0.01008463  1.084638146</span></span>
<span><span class="co">## 19  1.23037277 0.54650 6.503704e-02 3119.646 1.0002099 -3.67850530 -1.988088366</span></span>
<span><span class="co">## 20  1.23646821 0.23700 6.656818e-02 5029.945 1.0000561 -1.56585070  0.556560553</span></span>
<span><span class="co">## 21  0.21108077 0.12200 2.459375e-02 3280.247 1.0001883 -4.75057680 -3.956729613</span></span>
<span><span class="co">## 22  0.03742959 0.07825 2.419643e-02 3207.719 1.0006216 -4.06545638 -2.705331883</span></span>
<span><span class="co">## 23  1.43191325 0.00000 0.000000e+00 5480.782 0.9998635 -4.15455916 -3.506021345</span></span>
<span><span class="co">## 24  0.74135903 0.70800 1.638017e-01 5201.313 0.9999224 -0.98438838  0.033244727</span></span>
<span><span class="co">## 25  1.51027762 0.00000 0.000000e+00 4962.997 1.0003573 -4.80224189 -4.092711704</span></span>
<span><span class="co">## 26  1.17355200 0.02600 8.275000e-03 4727.549 0.9998071 -1.12784152  0.078269559</span></span>
<span><span class="co">## 27 -0.12001874 0.04675 9.950000e-03 5443.383 1.0000110 -5.86296194 -5.081376252</span></span>
<span><span class="co">## 28  2.02829971 0.00625 2.642857e-03 5411.040 0.9997743 -0.75767968  0.381894108</span></span>
<span><span class="co">## 29  1.26483078 0.00000 0.000000e+00 4952.365 0.9997539 -4.96303774 -4.231524099</span></span>
<span><span class="co">## 30  1.18944335 0.08725 3.178125e-02 5923.551 0.9997500 -0.79476240  0.314007718</span></span>
<span><span class="co">## 31 -0.57534208 0.00175 1.875000e-04 4558.019 0.9997530 -6.94876237 -5.978223962</span></span>
<span><span class="co">## 32 -0.19717460 0.02550 6.305556e-03 3995.259 0.9997574 -1.51614840 -0.065782456</span></span>
<span><span class="co">## 33 -0.79343840 0.00000 0.000000e+00 4934.361 0.9997508 -6.61524653 -5.738007952</span></span>
<span><span class="co">## 34  0.63691786 0.60150 1.443661e-01 4905.417 0.9998788 -2.11272948 -0.764814065</span></span>
<span><span class="co">## 35 -0.14811125 0.03375 6.388889e-03 4706.964 1.0000588 -5.23364789 -4.494631142</span></span>
<span><span class="co">## 36 -0.72635624 0.01275 3.906250e-03 1518.947 0.9997775 -5.28623567 -2.335633464</span></span>
<span><span class="co">## 37  0.42253295 0.83400 1.409917e-01 4489.978 1.0009650 -5.40649731 -4.687424353</span></span>
<span><span class="co">## 38  0.47928402 0.08725 3.504412e-02 3960.677 1.0005387 -0.16570649  1.198535866</span></span>
<span><span class="co">## 39  2.06602964 0.00000 0.000000e+00 4089.532 0.9997996 -4.29983259 -3.595346363</span></span>
<span><span class="co">## 40  0.33454812 0.24950 8.163542e-02 4719.467 0.9998558  0.36950805  1.447330311</span></span>
<span><span class="co">## 41  1.31288414 0.00000 0.000000e+00 4515.639 0.9997572 -4.70205604 -4.013681841</span></span>
<span><span class="co">## 42  0.71552349 0.72475 1.825000e-01 5635.379 0.9997501 -0.89468216  0.255410248</span></span>
<span><span class="co">## 43  0.41382403 0.83225 1.170948e-01 4327.038 0.9997636 -5.59345116 -4.837184422</span></span>
<span><span class="co">## 44  0.33123286 0.19250 5.151250e-02 3892.888 0.9998474 -0.76062363  0.373365278</span></span>
<span><span class="co">## 45  1.73538907 0.00000 0.000000e+00 4314.920 0.9998546 -4.52438986 -3.806325842</span></span>
<span><span class="co">## 46  1.12488395 0.13075 4.409211e-02 5457.642 1.0003145 -0.49032665  0.650288209</span></span>
<span><span class="co">## 47 -0.70981281 0.00000 0.000000e+00 5024.023 0.9997500 -6.64497166 -5.792684985</span></span>
<span><span class="co">## 48  0.39724792 0.52550 1.094615e-01 4432.611 0.9998815 -3.34921612 -1.592857383</span></span>
<span><span class="co">## 49 -0.99508206 0.00050 4.545455e-05 5021.266 0.9997502 -6.35782262 -5.558080776</span></span>
<span><span class="co">## 50  1.73387528 0.59475 1.274352e-01 3074.236 0.9998121 -0.67598186  0.704770116</span></span>
<span><span class="co">## 51 -0.02997738 0.08250 1.631818e-02 5331.699 1.0000081 -5.46070845 -4.783184277</span></span>
<span><span class="co">## 52 -0.76776315 0.00550 2.041667e-03 2093.125 0.9998001 -2.78376816 -0.834866765</span></span>
<span><span class="co">## 53 -0.28152646 0.01075 1.500000e-03 4549.574 0.9997596 -6.48609943 -5.583856580</span></span>
<span><span class="co">## 54  1.88726441 0.04975 1.204545e-02 5066.297 0.9997610  0.13954618  1.330438753</span></span>
<span><span class="co">## 55 -0.01256852 0.07750 1.316667e-02 4443.899 0.9998362 -5.28874914 -4.567020095</span></span>
<span><span class="co">## 56  1.51109373 0.24525 7.433696e-02 4364.434 0.9997874 -1.22300039 -0.104424234</span></span>
<span><span class="co">## 57  0.22486805 0.34850 4.651923e-02 5203.704 0.9997613 -5.39054319 -4.632316184</span></span>
<span><span class="co">## 58  1.01648162 0.36125 9.282000e-02 5333.302 0.9998791 -1.77647771 -0.643090318</span></span>
<span><span class="co">## 59 -0.15935470 0.03725 8.013158e-03 5684.462 0.9998258 -5.79810031 -5.050842692</span></span>
<span><span class="co">## 60  1.53257700 0.11125 3.927778e-02 6155.554 0.9998523 -1.10315696  0.004002058</span></span>
<span><span class="co">##        v_upper   v_pH0        v_FDR  v_n_eff v_R_k_hat   count_data</span></span>
<span><span class="co">## 1  -3.01872776 0.00000 0.0000000000 5019.271 0.9998999 c("10x_6....</span></span>
<span><span class="co">## 2   0.73225613 0.32225 0.1457500000 5561.400 0.9999999 c("10x_6....</span></span>
<span><span class="co">## 3  -3.56055266 0.00000 0.0000000000 3888.401 1.0002409 c("10x_6....</span></span>
<span><span class="co">## 4   0.24241821 0.11425 0.0575000000 3794.083 0.9998946 c("10x_6....</span></span>
<span><span class="co">## 5  -2.47696889 0.00000 0.0000000000 5286.253 1.0000057 c("10x_6....</span></span>
<span><span class="co">## 6   0.31475477 0.11775 0.0635250000 5228.213 0.9999928 c("10x_6....</span></span>
<span><span class="co">## 7  -3.75892875 0.00000 0.0000000000 6645.589 0.9997837 c("10x_6....</span></span>
<span><span class="co">## 8   3.49609452 0.10175 0.0504062500 3578.167 1.0007517 c("10x_6....</span></span>
<span><span class="co">## 9  -3.70575897 0.00000 0.0000000000 5855.623 0.9997973 c("10x_6....</span></span>
<span><span class="co">## 10  1.84732734 0.23050 0.1089821429 4579.874 1.0007839 c("10x_6....</span></span>
<span><span class="co">## 11 -4.16640248 0.00000 0.0000000000 5285.853 1.0001452 c("10x_6....</span></span>
<span><span class="co">## 12  1.93836798 0.42850 0.1997173913 3374.151 0.9997913 c("10x_6....</span></span>
<span><span class="co">## 13 -2.29771923 0.00000 0.0000000000 3462.562 1.0003972 c("10x_6....</span></span>
<span><span class="co">## 14  1.27773179 0.49925 0.2223000000 4156.498 1.0004262 c("10x_6....</span></span>
<span><span class="co">## 15 -3.99300750 0.00000 0.0000000000 5669.882 0.9997701 c("10x_6....</span></span>
<span><span class="co">## 16  1.80000579 0.28500 0.1353676471 6103.866 0.9997988 c("10x_6....</span></span>
<span><span class="co">## 17 -3.23653068 0.00000 0.0000000000 5902.873 0.9997624 c("10x_6....</span></span>
<span><span class="co">## 18  2.21527199 0.05475 0.0280500000 5098.396 0.9997843 c("10x_6....</span></span>
<span><span class="co">## 19 -0.64475576 0.00325 0.0001083333 2423.262 1.0002025 c("10x_6....</span></span>
<span><span class="co">## 20  2.65580991 0.37125 0.1576184211 4084.963 0.9998504 c("10x_6....</span></span>
<span><span class="co">## 21 -2.61731403 0.00000 0.0000000000 2115.026 1.0005751 c("10x_6....</span></span>
<span><span class="co">## 22 -1.34933753 0.00025 0.0002500000 4226.670 0.9997701 c("10x_6....</span></span>
<span><span class="co">## 23 -2.73340092 0.00000 0.0000000000 6326.152 0.9998076 c("10x_6....</span></span>
<span><span class="co">## 24  1.04888860 0.63450 0.2727500000 5300.137 0.9997500 c("10x_6....</span></span>
<span><span class="co">## 25 -3.25116445 0.00000 0.0000000000 5271.065 0.9998494 c("10x_6....</span></span>
<span><span class="co">## 26  1.31050005 0.58250 0.2598303571 5340.942 0.9997522 c("10x_6....</span></span>
<span><span class="co">## 27 -4.14932018 0.00000 0.0000000000 5653.588 0.9998666 c("10x_6....</span></span>
<span><span class="co">## 28  1.61574886 0.37175 0.1683250000 4778.894 0.9997724 c("10x_6....</span></span>
<span><span class="co">## 29 -3.37867607 0.00000 0.0000000000 5629.958 1.0001482 c("10x_6....</span></span>
<span><span class="co">## 30  1.50364031 0.42150 0.1893181818 4955.277 0.9997556 c("10x_6....</span></span>
<span><span class="co">## 31 -4.89145548 0.00000 0.0000000000 4620.841 0.9998009 c("10x_6....</span></span>
<span><span class="co">## 32  1.48695275 0.57050 0.2478796296 3910.155 1.0000160 c("10x_6....</span></span>
<span><span class="co">## 33 -4.69476932 0.00000 0.0000000000 4668.317 0.9997796 c("10x_6....</span></span>
<span><span class="co">## 34  0.60818154 0.20375 0.0762727273 4366.205 0.9997514 c("10x_6....</span></span>
<span><span class="co">## 35 -3.52046636 0.00000 0.0000000000 5402.457 0.9998934 c("10x_6....</span></span>
<span><span class="co">## 36  0.75043242 0.09000 0.0430714286 2304.954 0.9998608 c("10x_6....</span></span>
<span><span class="co">## 37 -3.81527454 0.00000 0.0000000000 8091.325 0.9997626 c("10x_6....</span></span>
<span><span class="co">## 38  2.73776606 0.07125 0.0352500000 3975.093 1.0008964 c("10x_6....</span></span>
<span><span class="co">## 39 -2.78482253 0.00000 0.0000000000 5945.671 0.9997571 c("10x_6....</span></span>
<span><span class="co">## 40  2.47225645 0.01150 0.0058750000 5068.737 0.9998013 c("10x_6....</span></span>
<span><span class="co">## 41 -3.20519019 0.00000 0.0000000000 6441.522 0.9997620 c("10x_6....</span></span>
<span><span class="co">## 42  1.45055764 0.46475 0.2107604167 4394.715 0.9997623 c("10x_6....</span></span>
<span><span class="co">## 43 -3.95129166 0.00000 0.0000000000 6108.715 1.0000434 c("10x_6....</span></span>
<span><span class="co">## 44  1.60048592 0.37700 0.1782619048 4002.354 0.9997909 c("10x_6....</span></span>
<span><span class="co">## 45 -3.01436927 0.00000 0.0000000000 6366.406 1.0000545 c("10x_6....</span></span>
<span><span class="co">## 46  1.79647407 0.22750 0.0888750000 5052.937 0.9997503 c("10x_6....</span></span>
<span><span class="co">## 47 -4.81969943 0.00000 0.0000000000 4849.228 0.9997580 c("10x_6....</span></span>
<span><span class="co">## 48 -0.03902555 0.04075 0.0213750000 4911.462 0.9998478 c("10x_6....</span></span>
<span><span class="co">## 49 -4.53942333 0.00000 0.0000000000 6369.055 0.9997564 c("10x_6....</span></span>
<span><span class="co">## 50  2.14981215 0.23675 0.1175000000 3830.718 0.9998986 c("10x_6....</span></span>
<span><span class="co">## 51 -3.88666788 0.00000 0.0000000000 5674.877 0.9997692 c("10x_6....</span></span>
<span><span class="co">## 52  0.95947865 0.25375 0.1260156250 2979.964 0.9997661 c("10x_6....</span></span>
<span><span class="co">## 53 -4.61994781 0.00000 0.0000000000 5289.690 0.9997532 c("10x_6....</span></span>
<span><span class="co">## 54  2.61106324 0.03300 0.0149166667 4854.778 0.9998050 c("10x_6....</span></span>
<span><span class="co">## 55 -3.58166206 0.00000 0.0000000000 4523.290 0.9997514 c("10x_6....</span></span>
<span><span class="co">## 56  1.04929961 0.56475 0.2354711538 4031.894 0.9997596 c("10x_6....</span></span>
<span><span class="co">## 57 -3.66604759 0.00000 0.0000000000 5721.000 0.9997644 c("10x_6....</span></span>
<span><span class="co">## 58  0.58076649 0.22875 0.0996346154 4789.019 0.9997569 c("10x_6....</span></span>
<span><span class="co">## 59 -4.14612563 0.00000 0.0000000000 6335.201 0.9997672 c("10x_6....</span></span>
<span><span class="co">## 60  1.14173270 0.63650 0.2848750000 4577.894 0.9997670 c("10x_6....</span></span></code></pre>
<p>A plot of group proportion, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is likely to be
descriptively adequate to the data, the blue box plot should roughly
overlay with the black box plot, which represents the observed data. The
outliers are coloured in red. A box plot will be returned for every
(discrete) covariate present in <code>formula_composition</code>. The
colour coding represents the significant associations for composition
and/or variability.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">boxplot</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `boxplot`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<p>A plot of estimates of differential composition (c_) on the x-axis
and differential variability (v_) on the y-axis. The error bars
represent 95% credible intervals. The dashed lines represent the minimal
effect that the hypothesis test is based on. An effect is labelled as
significant if bigger than the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_1D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior
distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example, we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that it has converged and is negative
with probability 1.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              cell_group   parameter factor      c_lower      c_effect</span></span>
<span><span class="co">## 1            B immature (Intercept)   &lt;NA&gt;  0.361335750  0.7610843787</span></span>
<span><span class="co">## 2            B immature typehealthy   type  0.846645226  1.4313448240</span></span>
<span><span class="co">## 3                 B mem (Intercept)   &lt;NA&gt; -1.491667363 -0.8705328282</span></span>
<span><span class="co">## 4                 B mem typehealthy   type  1.044621292  1.8464576110</span></span>
<span><span class="co">## 5         CD4 cm S100A4 (Intercept)   &lt;NA&gt;  1.315297759  1.6626472418</span></span>
<span><span class="co">## 6         CD4 cm S100A4 typehealthy   type  0.481536552  0.9328551417</span></span>
<span><span class="co">## 7  CD4 cm high cytokine (Intercept)   &lt;NA&gt; -1.055981798 -0.5455480513</span></span>
<span><span class="co">## 8  CD4 cm high cytokine typehealthy   type -3.046043839 -1.2425233937</span></span>
<span><span class="co">## 9       CD4 cm ribosome (Intercept)   &lt;NA&gt; -0.056237906  0.3117464682</span></span>
<span><span class="co">## 10      CD4 cm ribosome typehealthy   type -1.795418062 -0.9629447717</span></span>
<span><span class="co">## 11 CD4 em high cytokine (Intercept)   &lt;NA&gt; -1.160495355 -0.7124512091</span></span>
<span><span class="co">## 12 CD4 em high cytokine typehealthy   type -2.392637902 -1.3041040598</span></span>
<span><span class="co">## 13            CD4 naive (Intercept)   &lt;NA&gt;  0.789746901  1.2098129266</span></span>
<span><span class="co">## 14            CD4 naive typehealthy   type  0.455463569  1.1419744399</span></span>
<span><span class="co">## 15         CD4 ribosome (Intercept)   &lt;NA&gt; -0.316511321  0.0449776945</span></span>
<span><span class="co">## 16         CD4 ribosome typehealthy   type  1.465844336  2.0692814529</span></span>
<span><span class="co">## 17             CD8 em 1 (Intercept)   &lt;NA&gt;  0.305004180  0.6774084176</span></span>
<span><span class="co">## 18             CD8 em 1 typehealthy   type -0.056085036  0.7796162553</span></span>
<span><span class="co">## 19             CD8 em 2 (Intercept)   &lt;NA&gt; -0.852322829  0.1319217505</span></span>
<span><span class="co">## 20             CD8 em 2 typehealthy   type -3.495939955 -1.1332514627</span></span>
<span><span class="co">## 21             CD8 em 3 (Intercept)   &lt;NA&gt; -1.376888305 -0.6587100309</span></span>
<span><span class="co">## 22             CD8 em 3 typehealthy   type -1.761624627 -0.8220066434</span></span>
<span><span class="co">## 23            CD8 naive (Intercept)   &lt;NA&gt;  0.612878008  1.0227113279</span></span>
<span><span class="co">## 24            CD8 naive typehealthy   type -0.730622098  0.0081362943</span></span>
<span><span class="co">## 25     CD8 transitional (Intercept)   &lt;NA&gt;  0.921485057  1.2122637013</span></span>
<span><span class="co">## 26     CD8 transitional typehealthy   type  0.196110474  0.6797219648</span></span>
<span><span class="co">## 27                 MAIT (Intercept)   &lt;NA&gt; -1.063688975 -0.5958567156</span></span>
<span><span class="co">## 28                 MAIT typehealthy   type  0.457965724  1.2146536073</span></span>
<span><span class="co">## 29               Mac M1 (Intercept)   &lt;NA&gt;  0.620993651  0.9469437283</span></span>
<span><span class="co">## 30               Mac M1 typehealthy   type  0.009944924  0.5952306473</span></span>
<span><span class="co">## 31        Mac M1 FCER1A (Intercept)   &lt;NA&gt; -1.352662954 -0.9805367919</span></span>
<span><span class="co">## 32        Mac M1 FCER1A typehealthy   type -1.814850080 -1.0779609819</span></span>
<span><span class="co">## 33                 Mono (Intercept)   &lt;NA&gt; -1.808269928 -1.3396240981</span></span>
<span><span class="co">## 34                 Mono typehealthy   type -0.858907732 -0.1001649072</span></span>
<span><span class="co">## 35          Mono NKG7 1 (Intercept)   &lt;NA&gt; -1.507486023 -0.8651796037</span></span>
<span><span class="co">## 36          Mono NKG7 1 typehealthy   type -3.991498845 -2.8799693881</span></span>
<span><span class="co">## 37          Mono NKG7 2 (Intercept)   &lt;NA&gt; -0.413092407 -0.0006123920</span></span>
<span><span class="co">## 38          Mono NKG7 2 typehealthy   type -2.438862982 -1.2690211681</span></span>
<span><span class="co">## 39  Mono classic inflam (Intercept)   &lt;NA&gt;  1.491381700  1.7882557952</span></span>
<span><span class="co">## 40  Mono classic inflam typehealthy   type -1.253629418 -0.4731128230</span></span>
<span><span class="co">## 41     Mono non-classic (Intercept)   &lt;NA&gt;  0.616986405  0.9672676712</span></span>
<span><span class="co">## 42     Mono non-classic typehealthy   type -0.628262840  0.0008427117</span></span>
<span><span class="co">## 43    Myeloid migratory (Intercept)   &lt;NA&gt; -0.359035184  0.0147033416</span></span>
<span><span class="co">## 44    Myeloid migratory typehealthy   type -1.327486243 -0.5464070029</span></span>
<span><span class="co">## 45                   NK (Intercept)   &lt;NA&gt;  1.135281829  1.4325018510</span></span>
<span><span class="co">## 46                   NK typehealthy   type -0.053131837  0.5443258722</span></span>
<span><span class="co">## 47           NK cycling (Intercept)   &lt;NA&gt; -1.654773213 -1.2153089284</span></span>
<span><span class="co">## 48           NK cycling typehealthy   type -0.811196824 -0.1805742824</span></span>
<span><span class="co">## 49     NK high cytokine (Intercept)   &lt;NA&gt; -2.191038253 -1.6320773684</span></span>
<span><span class="co">## 50     NK high cytokine typehealthy   type -1.194242846  0.0456987120</span></span>
<span><span class="co">## 51                 Stem (Intercept)   &lt;NA&gt; -1.092679243 -0.5750920505</span></span>
<span><span class="co">## 52                 Stem typehealthy   type -3.213392800 -2.2399877567</span></span>
<span><span class="co">## 53                T gd1 (Intercept)   &lt;NA&gt; -1.073860129 -0.6898489711</span></span>
<span><span class="co">## 54                T gd1 typehealthy   type  0.040046344  0.9267160202</span></span>
<span><span class="co">## 55                T gd2 (Intercept)   &lt;NA&gt; -1.238421686 -0.6644650242</span></span>
<span><span class="co">## 56                T gd2 typehealthy   type -0.392518985  0.5164781590</span></span>
<span><span class="co">## 57                 cDC2 (Intercept)   &lt;NA&gt; -0.786043102 -0.2921548064</span></span>
<span><span class="co">## 58                 cDC2 typehealthy   type -0.380050576  0.3243495149</span></span>
<span><span class="co">## 59                  pDC (Intercept)   &lt;NA&gt; -1.148425930 -0.6763980865</span></span>
<span><span class="co">## 60                  pDC typehealthy   type -0.089065820  0.6732108058</span></span>
<span><span class="co">##        c_upper   c_pH0        c_FDR  c_n_eff c_R_k_hat     v_lower     v_effect</span></span>
<span><span class="co">## 1   1.19082303 0.00325 4.230769e-04 5555.660 0.9998081 -4.49186224 -3.827272207</span></span>
<span><span class="co">## 2   1.97751054 0.00000 0.000000e+00 4128.725 0.9997506 -1.67274410 -0.490730778</span></span>
<span><span class="co">## 3  -0.17554074 0.02975 4.779412e-03 4279.296 1.0001883 -5.30318509 -4.565244119</span></span>
<span><span class="co">## 4   2.63245560 0.00000 0.000000e+00 3984.925 1.0005908 -2.06742753 -0.939821112</span></span>
<span><span class="co">## 5   1.98510366 0.00000 0.000000e+00 5809.372 1.0000078 -3.83922207 -3.202783354</span></span>
<span><span class="co">## 6   1.40721764 0.00250 6.250000e-04 4899.517 0.9998172 -2.05632280 -0.929137943</span></span>
<span><span class="co">## 7   0.03259866 0.10925 2.035870e-02 4469.169 0.9999394 -5.47834443 -4.712033028</span></span>
<span><span class="co">## 8   1.20715576 0.19725 5.845238e-02 3076.835 1.0009504 -0.43521243  1.531293936</span></span>
<span><span class="co">## 9   0.69435529 0.27075 3.444000e-02 4096.081 0.9997675 -5.35672733 -4.604764084</span></span>
<span><span class="co">## 10  0.07316059 0.06575 2.003846e-02 3480.198 1.0002291 -0.55464411  0.651979465</span></span>
<span><span class="co">## 11 -0.18310523 0.02900 3.218750e-03 4094.068 1.0000791 -5.93410382 -5.149613287</span></span>
<span><span class="co">## 12  0.23926880 0.06225 1.622917e-02 3247.258 0.9997584 -1.11260720  0.350654782</span></span>
<span><span class="co">## 13  1.65239493 0.00000 0.000000e+00 5381.033 0.9997778 -3.66736227 -3.060646053</span></span>
<span><span class="co">## 14  1.80123725 0.00425 1.350000e-03 5219.699 1.0004124 -0.83812993  0.200559741</span></span>
<span><span class="co">## 15  0.43244777 0.80750 9.155357e-02 5412.260 0.9997515 -5.68558806 -4.889662576</span></span>
<span><span class="co">## 16  2.63076283 0.00000 0.000000e+00 6213.848 0.9997607 -0.68062978  0.557819163</span></span>
<span><span class="co">## 17  1.08720539 0.00625 8.392857e-04 4342.332 0.9997510 -4.78177096 -4.069967955</span></span>
<span><span class="co">## 18  1.58261272 0.08250 2.808333e-02 4335.443 1.0000313 -0.01008463  1.084638146</span></span>
<span><span class="co">## 19  1.23037277 0.54650 6.503704e-02 3119.646 1.0002099 -3.67850530 -1.988088366</span></span>
<span><span class="co">## 20  1.23646821 0.23700 6.656818e-02 5029.945 1.0000561 -1.56585070  0.556560553</span></span>
<span><span class="co">## 21  0.21108077 0.12200 2.459375e-02 3280.247 1.0001883 -4.75057680 -3.956729613</span></span>
<span><span class="co">## 22  0.03742959 0.07825 2.419643e-02 3207.719 1.0006216 -4.06545638 -2.705331883</span></span>
<span><span class="co">## 23  1.43191325 0.00000 0.000000e+00 5480.782 0.9998635 -4.15455916 -3.506021345</span></span>
<span><span class="co">## 24  0.74135903 0.70800 1.638017e-01 5201.313 0.9999224 -0.98438838  0.033244727</span></span>
<span><span class="co">## 25  1.51027762 0.00000 0.000000e+00 4962.997 1.0003573 -4.80224189 -4.092711704</span></span>
<span><span class="co">## 26  1.17355200 0.02600 8.275000e-03 4727.549 0.9998071 -1.12784152  0.078269559</span></span>
<span><span class="co">## 27 -0.12001874 0.04675 9.950000e-03 5443.383 1.0000110 -5.86296194 -5.081376252</span></span>
<span><span class="co">## 28  2.02829971 0.00625 2.642857e-03 5411.040 0.9997743 -0.75767968  0.381894108</span></span>
<span><span class="co">## 29  1.26483078 0.00000 0.000000e+00 4952.365 0.9997539 -4.96303774 -4.231524099</span></span>
<span><span class="co">## 30  1.18944335 0.08725 3.178125e-02 5923.551 0.9997500 -0.79476240  0.314007718</span></span>
<span><span class="co">## 31 -0.57534208 0.00175 1.875000e-04 4558.019 0.9997530 -6.94876237 -5.978223962</span></span>
<span><span class="co">## 32 -0.19717460 0.02550 6.305556e-03 3995.259 0.9997574 -1.51614840 -0.065782456</span></span>
<span><span class="co">## 33 -0.79343840 0.00000 0.000000e+00 4934.361 0.9997508 -6.61524653 -5.738007952</span></span>
<span><span class="co">## 34  0.63691786 0.60150 1.443661e-01 4905.417 0.9998788 -2.11272948 -0.764814065</span></span>
<span><span class="co">## 35 -0.14811125 0.03375 6.388889e-03 4706.964 1.0000588 -5.23364789 -4.494631142</span></span>
<span><span class="co">## 36 -0.72635624 0.01275 3.906250e-03 1518.947 0.9997775 -5.28623567 -2.335633464</span></span>
<span><span class="co">## 37  0.42253295 0.83400 1.409917e-01 4489.978 1.0009650 -5.40649731 -4.687424353</span></span>
<span><span class="co">## 38  0.47928402 0.08725 3.504412e-02 3960.677 1.0005387 -0.16570649  1.198535866</span></span>
<span><span class="co">## 39  2.06602964 0.00000 0.000000e+00 4089.532 0.9997996 -4.29983259 -3.595346363</span></span>
<span><span class="co">## 40  0.33454812 0.24950 8.163542e-02 4719.467 0.9998558  0.36950805  1.447330311</span></span>
<span><span class="co">## 41  1.31288414 0.00000 0.000000e+00 4515.639 0.9997572 -4.70205604 -4.013681841</span></span>
<span><span class="co">## 42  0.71552349 0.72475 1.825000e-01 5635.379 0.9997501 -0.89468216  0.255410248</span></span>
<span><span class="co">## 43  0.41382403 0.83225 1.170948e-01 4327.038 0.9997636 -5.59345116 -4.837184422</span></span>
<span><span class="co">## 44  0.33123286 0.19250 5.151250e-02 3892.888 0.9998474 -0.76062363  0.373365278</span></span>
<span><span class="co">## 45  1.73538907 0.00000 0.000000e+00 4314.920 0.9998546 -4.52438986 -3.806325842</span></span>
<span><span class="co">## 46  1.12488395 0.13075 4.409211e-02 5457.642 1.0003145 -0.49032665  0.650288209</span></span>
<span><span class="co">## 47 -0.70981281 0.00000 0.000000e+00 5024.023 0.9997500 -6.64497166 -5.792684985</span></span>
<span><span class="co">## 48  0.39724792 0.52550 1.094615e-01 4432.611 0.9998815 -3.34921612 -1.592857383</span></span>
<span><span class="co">## 49 -0.99508206 0.00050 4.545455e-05 5021.266 0.9997502 -6.35782262 -5.558080776</span></span>
<span><span class="co">## 50  1.73387528 0.59475 1.274352e-01 3074.236 0.9998121 -0.67598186  0.704770116</span></span>
<span><span class="co">## 51 -0.02997738 0.08250 1.631818e-02 5331.699 1.0000081 -5.46070845 -4.783184277</span></span>
<span><span class="co">## 52 -0.76776315 0.00550 2.041667e-03 2093.125 0.9998001 -2.78376816 -0.834866765</span></span>
<span><span class="co">## 53 -0.28152646 0.01075 1.500000e-03 4549.574 0.9997596 -6.48609943 -5.583856580</span></span>
<span><span class="co">## 54  1.88726441 0.04975 1.204545e-02 5066.297 0.9997610  0.13954618  1.330438753</span></span>
<span><span class="co">## 55 -0.01256852 0.07750 1.316667e-02 4443.899 0.9998362 -5.28874914 -4.567020095</span></span>
<span><span class="co">## 56  1.51109373 0.24525 7.433696e-02 4364.434 0.9997874 -1.22300039 -0.104424234</span></span>
<span><span class="co">## 57  0.22486805 0.34850 4.651923e-02 5203.704 0.9997613 -5.39054319 -4.632316184</span></span>
<span><span class="co">## 58  1.01648162 0.36125 9.282000e-02 5333.302 0.9998791 -1.77647771 -0.643090318</span></span>
<span><span class="co">## 59 -0.15935470 0.03725 8.013158e-03 5684.462 0.9998258 -5.79810031 -5.050842692</span></span>
<span><span class="co">## 60  1.53257700 0.11125 3.927778e-02 6155.554 0.9998523 -1.10315696  0.004002058</span></span>
<span><span class="co">##        v_upper   v_pH0        v_FDR  v_n_eff v_R_k_hat   count_data</span></span>
<span><span class="co">## 1  -3.01872776 0.00000 0.0000000000 5019.271 0.9998999 c("10x_6....</span></span>
<span><span class="co">## 2   0.73225613 0.32225 0.1457500000 5561.400 0.9999999 c("10x_6....</span></span>
<span><span class="co">## 3  -3.56055266 0.00000 0.0000000000 3888.401 1.0002409 c("10x_6....</span></span>
<span><span class="co">## 4   0.24241821 0.11425 0.0575000000 3794.083 0.9998946 c("10x_6....</span></span>
<span><span class="co">## 5  -2.47696889 0.00000 0.0000000000 5286.253 1.0000057 c("10x_6....</span></span>
<span><span class="co">## 6   0.31475477 0.11775 0.0635250000 5228.213 0.9999928 c("10x_6....</span></span>
<span><span class="co">## 7  -3.75892875 0.00000 0.0000000000 6645.589 0.9997837 c("10x_6....</span></span>
<span><span class="co">## 8   3.49609452 0.10175 0.0504062500 3578.167 1.0007517 c("10x_6....</span></span>
<span><span class="co">## 9  -3.70575897 0.00000 0.0000000000 5855.623 0.9997973 c("10x_6....</span></span>
<span><span class="co">## 10  1.84732734 0.23050 0.1089821429 4579.874 1.0007839 c("10x_6....</span></span>
<span><span class="co">## 11 -4.16640248 0.00000 0.0000000000 5285.853 1.0001452 c("10x_6....</span></span>
<span><span class="co">## 12  1.93836798 0.42850 0.1997173913 3374.151 0.9997913 c("10x_6....</span></span>
<span><span class="co">## 13 -2.29771923 0.00000 0.0000000000 3462.562 1.0003972 c("10x_6....</span></span>
<span><span class="co">## 14  1.27773179 0.49925 0.2223000000 4156.498 1.0004262 c("10x_6....</span></span>
<span><span class="co">## 15 -3.99300750 0.00000 0.0000000000 5669.882 0.9997701 c("10x_6....</span></span>
<span><span class="co">## 16  1.80000579 0.28500 0.1353676471 6103.866 0.9997988 c("10x_6....</span></span>
<span><span class="co">## 17 -3.23653068 0.00000 0.0000000000 5902.873 0.9997624 c("10x_6....</span></span>
<span><span class="co">## 18  2.21527199 0.05475 0.0280500000 5098.396 0.9997843 c("10x_6....</span></span>
<span><span class="co">## 19 -0.64475576 0.00325 0.0001083333 2423.262 1.0002025 c("10x_6....</span></span>
<span><span class="co">## 20  2.65580991 0.37125 0.1576184211 4084.963 0.9998504 c("10x_6....</span></span>
<span><span class="co">## 21 -2.61731403 0.00000 0.0000000000 2115.026 1.0005751 c("10x_6....</span></span>
<span><span class="co">## 22 -1.34933753 0.00025 0.0002500000 4226.670 0.9997701 c("10x_6....</span></span>
<span><span class="co">## 23 -2.73340092 0.00000 0.0000000000 6326.152 0.9998076 c("10x_6....</span></span>
<span><span class="co">## 24  1.04888860 0.63450 0.2727500000 5300.137 0.9997500 c("10x_6....</span></span>
<span><span class="co">## 25 -3.25116445 0.00000 0.0000000000 5271.065 0.9998494 c("10x_6....</span></span>
<span><span class="co">## 26  1.31050005 0.58250 0.2598303571 5340.942 0.9997522 c("10x_6....</span></span>
<span><span class="co">## 27 -4.14932018 0.00000 0.0000000000 5653.588 0.9998666 c("10x_6....</span></span>
<span><span class="co">## 28  1.61574886 0.37175 0.1683250000 4778.894 0.9997724 c("10x_6....</span></span>
<span><span class="co">## 29 -3.37867607 0.00000 0.0000000000 5629.958 1.0001482 c("10x_6....</span></span>
<span><span class="co">## 30  1.50364031 0.42150 0.1893181818 4955.277 0.9997556 c("10x_6....</span></span>
<span><span class="co">## 31 -4.89145548 0.00000 0.0000000000 4620.841 0.9998009 c("10x_6....</span></span>
<span><span class="co">## 32  1.48695275 0.57050 0.2478796296 3910.155 1.0000160 c("10x_6....</span></span>
<span><span class="co">## 33 -4.69476932 0.00000 0.0000000000 4668.317 0.9997796 c("10x_6....</span></span>
<span><span class="co">## 34  0.60818154 0.20375 0.0762727273 4366.205 0.9997514 c("10x_6....</span></span>
<span><span class="co">## 35 -3.52046636 0.00000 0.0000000000 5402.457 0.9998934 c("10x_6....</span></span>
<span><span class="co">## 36  0.75043242 0.09000 0.0430714286 2304.954 0.9998608 c("10x_6....</span></span>
<span><span class="co">## 37 -3.81527454 0.00000 0.0000000000 8091.325 0.9997626 c("10x_6....</span></span>
<span><span class="co">## 38  2.73776606 0.07125 0.0352500000 3975.093 1.0008964 c("10x_6....</span></span>
<span><span class="co">## 39 -2.78482253 0.00000 0.0000000000 5945.671 0.9997571 c("10x_6....</span></span>
<span><span class="co">## 40  2.47225645 0.01150 0.0058750000 5068.737 0.9998013 c("10x_6....</span></span>
<span><span class="co">## 41 -3.20519019 0.00000 0.0000000000 6441.522 0.9997620 c("10x_6....</span></span>
<span><span class="co">## 42  1.45055764 0.46475 0.2107604167 4394.715 0.9997623 c("10x_6....</span></span>
<span><span class="co">## 43 -3.95129166 0.00000 0.0000000000 6108.715 1.0000434 c("10x_6....</span></span>
<span><span class="co">## 44  1.60048592 0.37700 0.1782619048 4002.354 0.9997909 c("10x_6....</span></span>
<span><span class="co">## 45 -3.01436927 0.00000 0.0000000000 6366.406 1.0000545 c("10x_6....</span></span>
<span><span class="co">## 46  1.79647407 0.22750 0.0888750000 5052.937 0.9997503 c("10x_6....</span></span>
<span><span class="co">## 47 -4.81969943 0.00000 0.0000000000 4849.228 0.9997580 c("10x_6....</span></span>
<span><span class="co">## 48 -0.03902555 0.04075 0.0213750000 4911.462 0.9998478 c("10x_6....</span></span>
<span><span class="co">## 49 -4.53942333 0.00000 0.0000000000 6369.055 0.9997564 c("10x_6....</span></span>
<span><span class="co">## 50  2.14981215 0.23675 0.1175000000 3830.718 0.9998986 c("10x_6....</span></span>
<span><span class="co">## 51 -3.88666788 0.00000 0.0000000000 5674.877 0.9997692 c("10x_6....</span></span>
<span><span class="co">## 52  0.95947865 0.25375 0.1260156250 2979.964 0.9997661 c("10x_6....</span></span>
<span><span class="co">## 53 -4.61994781 0.00000 0.0000000000 5289.690 0.9997532 c("10x_6....</span></span>
<span><span class="co">## 54  2.61106324 0.03300 0.0149166667 4854.778 0.9998050 c("10x_6....</span></span>
<span><span class="co">## 55 -3.58166206 0.00000 0.0000000000 4523.290 0.9997514 c("10x_6....</span></span>
<span><span class="co">## 56  1.04929961 0.56475 0.2354711538 4031.894 0.9997596 c("10x_6....</span></span>
<span><span class="co">## 57 -3.66604759 0.00000 0.0000000000 5721.000 0.9997644 c("10x_6....</span></span>
<span><span class="co">## 58  0.58076649 0.22875 0.0996346154 4789.019 0.9997569 c("10x_6....</span></span>
<span><span class="co">## 59 -4.14612563 0.00000 0.0000000000 6335.201 0.9997672 c("10x_6....</span></span>
<span><span class="co">## 60  1.14173270 0.63650 0.2848750000 4577.894 0.9997670 c("10x_6....</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_1D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<p>Plot 2D significance plot. Data points are cell groups. Error bars
are the 95% credible interval. The dashed lines represent the default
threshold fold change for which the probabilities (c_pH0, v_pH0) are
calculated. pH0 of 0 represent the rejection of the null hypothesis that
no effect is observed.</p>
<p>This plot is provided only if differential variability has been
tested. The differential variability estimates are reliable only if the
linear association between mean and variability for
<code>(intercept)</code> (left-hand side facet) is satisfied. A
scatterplot (besides the Intercept) is provided for each category of
interest. The for each category of interest, the composition and
variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_2D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-02-28 r85999)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rstan_2.32.6       StanHeaders_2.32.6 tidyr_1.3.1        forcats_1.0.0     </span></span>
<span><span class="co">## [5] ggplot2_3.5.0      sccomp_1.7.6       dplyr_1.1.4        BiocStyle_2.31.0  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gridExtra_2.3               inline_0.3.19              </span></span>
<span><span class="co">##  [3] rlang_1.1.3                 magrittr_2.0.3             </span></span>
<span><span class="co">##  [5] matrixStats_1.2.0           compiler_4.4.0             </span></span>
<span><span class="co">##  [7] loo_2.7.0                   systemfonts_1.0.5          </span></span>
<span><span class="co">##  [9] vctrs_0.6.5                 stringr_1.5.1              </span></span>
<span><span class="co">## [11] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">## [13] fastmap_1.1.1               XVector_0.43.1             </span></span>
<span><span class="co">## [15] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">## [17] rmarkdown_2.26              tzdb_0.4.0                 </span></span>
<span><span class="co">## [19] ragg_1.2.7                  purrr_1.0.2                </span></span>
<span><span class="co">## [21] xfun_0.42                   zlibbioc_1.49.0            </span></span>
<span><span class="co">## [23] cachem_1.0.8                GenomeInfoDb_1.39.7        </span></span>
<span><span class="co">## [25] jsonlite_1.8.8              highr_0.10                 </span></span>
<span><span class="co">## [27] DelayedArray_0.29.9         parallel_4.4.0             </span></span>
<span><span class="co">## [29] R6_2.5.1                    bslib_0.6.1                </span></span>
<span><span class="co">## [31] stringi_1.8.3               parallelly_1.37.1          </span></span>
<span><span class="co">## [33] boot_1.3-30                 GenomicRanges_1.55.3       </span></span>
<span><span class="co">## [35] jquerylib_0.1.4             Rcpp_1.0.12                </span></span>
<span><span class="co">## [37] bookdown_0.38               SummarizedExperiment_1.33.3</span></span>
<span><span class="co">## [39] knitr_1.45                  future.apply_1.11.1        </span></span>
<span><span class="co">## [41] readr_2.1.5                 IRanges_2.37.1             </span></span>
<span><span class="co">## [43] Matrix_1.6-5                tidyselect_1.2.0           </span></span>
<span><span class="co">## [45] abind_1.4-5                 yaml_2.3.8                 </span></span>
<span><span class="co">## [47] codetools_0.2-19            curl_5.2.1                 </span></span>
<span><span class="co">## [49] listenv_0.9.1               pkgbuild_1.4.3             </span></span>
<span><span class="co">## [51] lattice_0.22-5              tibble_3.2.1               </span></span>
<span><span class="co">## [53] withr_3.0.0                 Biobase_2.63.0             </span></span>
<span><span class="co">## [55] evaluate_0.23               future_1.33.1              </span></span>
<span><span class="co">## [57] desc_1.4.3                  RcppParallel_5.1.7         </span></span>
<span><span class="co">## [59] pillar_1.9.0                BiocManager_1.30.22        </span></span>
<span><span class="co">## [61] MatrixGenerics_1.15.0       stats4_4.4.0               </span></span>
<span><span class="co">## [63] generics_0.1.3              sp_2.1-3                   </span></span>
<span><span class="co">## [65] S4Vectors_0.41.3            hms_1.1.3                  </span></span>
<span><span class="co">## [67] rstantools_2.4.0            munsell_0.5.0              </span></span>
<span><span class="co">## [69] scales_1.3.0                globals_0.16.2             </span></span>
<span><span class="co">## [71] glue_1.7.0                  tools_4.4.0                </span></span>
<span><span class="co">## [73] fs_1.6.3                    dotCall64_1.1-1            </span></span>
<span><span class="co">## [75] grid_4.4.0                  QuickJSR_1.1.3             </span></span>
<span><span class="co">## [77] colorspace_2.1-0            SingleCellExperiment_1.25.0</span></span>
<span><span class="co">## [79] GenomeInfoDbData_1.2.11     patchwork_1.2.0            </span></span>
<span><span class="co">## [81] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">## [83] spam_2.10-0                 fansi_1.0.6                </span></span>
<span><span class="co">## [85] S4Arrays_1.3.5              V8_4.4.2                   </span></span>
<span><span class="co">## [87] gtable_0.3.4                sass_0.4.8                 </span></span>
<span><span class="co">## [89] digest_0.6.34               progressr_0.14.0           </span></span>
<span><span class="co">## [91] BiocGenerics_0.49.1         SparseArray_1.3.4          </span></span>
<span><span class="co">## [93] ggrepel_0.9.5               farver_2.1.1               </span></span>
<span><span class="co">## [95] SeuratObject_5.0.1          memoise_2.0.1              </span></span>
<span><span class="co">## [97] htmltools_0.5.7             pkgdown_2.0.7              </span></span>
<span><span class="co">## [99] lifecycle_1.0.4</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
