<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>sccomp: Differential Composition and Variability Analysis for Single-Cell Data • sccomp</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="sccomp: Differential Composition and Variability Analysis for Single-Cell Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sccomp</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.21</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction.html">sccomp: Differential Composition and Variability Analysis for Single-Cell Data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MangiolaLaboratory/sccomp/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>sccomp: Differential Composition and Variability Analysis for Single-Cell Data</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2025-11-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MangiolaLaboratory/sccomp/blob/master/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Sccomp is a comprehensive R package for differential
      composition and variability analysis in single-cell RNA
      sequencing, CyTOF, and microbiome data. It provides robust
      Bayesian modeling with outlier detection, random effects, and
      advanced statistical methods for cell type proportion analysis.
      Perfect for cancer research, immunology, and single-cell
      genomics.</p>
    </div>
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/sccomp/actions/" class="external-link"><img src="https://github.com/stemangiola/sccomp/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<div class="section level2">
<h2 id="section">
<img src="../inst/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h2>
<p><strong>sccomp</strong> is a powerful R package designed for
comprehensive differential composition and variability analysis in
single-cell genomics, proteomics, and microbiomics data.</p>
<div class="section level3">
<h3 id="why-sccomp">Why sccomp?<a class="anchor" aria-label="anchor" href="#why-sccomp"></a>
</h3>
<p>For cellular omic data, no method for differential variability
analysis exists, and methods for differential composition analysis only
take a few fundamental data properties into account. Here we introduce
<strong>sccomp</strong>, a generalised method for differential
composition and variability analyses capable of jointly modelling data
count distribution, compositionality, group-specific variability, and
proportion mean-variability association, while being robust to
outliers.</p>
<p><img src="../inst/cartoon_methods.png" width="100%"></p>
<div class="section level4">
<h4 id="comprehensive-method-comparison">Comprehensive Method Comparison<a class="anchor" aria-label="anchor" href="#comprehensive-method-comparison"></a>
</h4>
<ul>
<li>
<strong>I</strong>: Data are modelled as counts.</li>
<li>
<strong>II</strong>: Group proportions are modelled as
compositional.</li>
<li>
<strong>III</strong>: The proportion variability is modelled as
cell-type specific.</li>
<li>
<strong>IV</strong>: Information sharing across cell types,
mean–variability association.</li>
<li>
<strong>V</strong>: Outlier detection or robustness.</li>
<li>
<strong>VI</strong>: Differential variability analysis.</li>
<li>
<strong>VII</strong> Mixed effect modelling</li>
<li>
<strong>VIII</strong> Removal unwanted effects</li>
</ul>
<table class="table">
<colgroup>
<col width="17%">
<col width="7%">
<col width="37%">
<col width="3%">
<col width="4%">
<col width="6%">
<col width="4%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th>Method</th>
<th>Year</th>
<th>Model</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
<th>VI</th>
<th>VII</th>
<th>VIII</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>sccomp</strong></td>
<td>2023</td>
<td>Sum-constrained Beta-binomial</td>
<td>●</td>
<td>●</td>
<td>●</td>
<td>●</td>
<td>●</td>
<td>●</td>
<td>●</td>
<td>●</td>
</tr>
<tr class="even">
<td><strong>scCODA</strong></td>
<td>2021</td>
<td>Dirichlet-multinomial</td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>quasi-binom.</strong></td>
<td>2021</td>
<td>Quasi-binomial</td>
<td>●</td>
<td></td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>rlm</strong></td>
<td>2021</td>
<td>Robust-log-linear</td>
<td></td>
<td>●</td>
<td></td>
<td></td>
<td>●</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>propeller</strong></td>
<td>2021</td>
<td>Logit-linear + limma</td>
<td></td>
<td>●</td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>ANCOM-BC</strong></td>
<td>2020</td>
<td>Log-linear</td>
<td></td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>corncob</strong></td>
<td>2020</td>
<td>Beta-binomial</td>
<td>●</td>
<td></td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>scDC</strong></td>
<td>2019</td>
<td>Log-linear</td>
<td></td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>dmbvs</strong></td>
<td>2017</td>
<td>Dirichlet-multinomial</td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><strong>MixMC</strong></td>
<td>2016</td>
<td>Zero-inflated Log-linear</td>
<td></td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><strong>ALDEx2</strong></td>
<td>2014</td>
<td>Dirichlet-multinomial</td>
<td>●</td>
<td>●</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="scientific-citation">Scientific Citation<a class="anchor" aria-label="anchor" href="#scientific-citation"></a>
</h4>
<p>Mangiola, Stefano, Alexandra J. Roth-Schulze, Marie Trussart, Enrique
Zozaya-Valdés, Mengyao Ma, Zijie Gao, Alan F. Rubin, Terence P. Speed,
Heejung Shim, and Anthony T. Papenfuss. 2023. “Sccomp: Robust
Differential Composition and Variability Analysis for Single-Cell Data.”
Proceedings of the National Academy of Sciences of the United States of
America 120 (33): e2203828120. <a href="https://doi.org/10.1073/pnas.2203828120" class="external-link uri">https://doi.org/10.1073/pnas.2203828120</a> <a href="https://www.pnas.org/doi/full/10.1073/pnas.2203828120" class="external-link">PNAS -
sccomp: Robust differential composition and variability analysis for
single-cell data</a></p>
</div>
<div class="section level4">
<h4 id="talk">Talk<a class="anchor" aria-label="anchor" href="#talk"></a>
</h4>
<p><a href="https://www.youtube.com/watch?v=R_lt58We9nA&amp;ab_channel=RConsortium" target="_blank" class="external-link">
<img src="https://img.youtube.com/vi/R_lt58We9nA/mqdefault.jpg" alt="Watch the video" width="280" height="180" border="10"></a></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="installation-guide">Installation Guide<a class="anchor" aria-label="anchor" href="#installation-guide"></a>
</h2>
<p><code>sccomp</code> is based on <code>cmdstanr</code> which provides
the latest version of <code>cmdstan</code> the Bayesian modelling tool.
<code>cmdstanr</code> is not on CRAN, so we need to have 3 simple step
process (that will be prompted to the user is forgot).</p>
<ol style="list-style-type: decimal">
<li>R installation of <code>sccomp</code>
</li>
<li>R installation of <code>cmdstanr</code>
</li>
<li>
<code>cmdstanr</code> call to <code>cmdstan</code> installation</li>
</ol>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 1</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cmdstanr"</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://stan-dev.r-universe.dev/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"repos"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 3</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">check_cmdstan_toolchain</a></span><span class="op">(</span>fix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Just checking system setting</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">install_cmdstan</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 1</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"MangiolaLaboratory/sccomp"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cmdstanr"</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://stan-dev.r-universe.dev/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"repos"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 3</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">check_cmdstan_toolchain</a></span><span class="op">(</span>fix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Just checking system setting</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/cmdstanr/reference/install_cmdstan.html" class="external-link">install_cmdstan</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="core-functions">Core Functions<a class="anchor" aria-label="anchor" href="#core-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>sccomp_estimate</code></td>
<td>Fit the model onto the data, and estimate the coefficients</td>
</tr>
<tr class="even">
<td><code>sccomp_remove_outliers</code></td>
<td>Identify outliers probabilistically based on the model fit, and
exclude them from the estimation</td>
</tr>
<tr class="odd">
<td><code>sccomp_test</code></td>
<td>Calculate the probability that the coefficients are outside the H0
interval (i.e. test_composition_above_logit_fold_change)</td>
</tr>
<tr class="even">
<td><code>sccomp_replicate</code></td>
<td>Simulate data from the model, or part of the model</td>
</tr>
<tr class="odd">
<td><code>sccomp_predict</code></td>
<td>Predicts proportions, based on the model, or part of the model</td>
</tr>
<tr class="even">
<td><code>sccomp_remove_unwanted_effects</code></td>
<td>Removes the variability for unwanted factors</td>
</tr>
<tr class="odd">
<td><code>plot</code></td>
<td>Plots summary plots to assess significance</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="analysis-tutorial">Analysis Tutorial<a class="anchor" aria-label="anchor" href="#analysis-tutorial"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MangiolaLaboratory/sccomp" class="external-link">sccomp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://forcats.tidyverse.org/" class="external-link">forcats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"seurat_obj"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sce_obj"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"counts_obj"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="binary-factor-analysis">Binary Factor Analysis<a class="anchor" aria-label="anchor" href="#binary-factor-analysis"></a>
</h3>
<p>Of the output table, the estimate columns start with the prefix
<code>c_</code> indicate <code>composition</code>, or with
<code>v_</code> indicate <code>variability</code> (when
formula_variability is set).</p>
<div class="section level4">
<h4 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">sce_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>    abundance <span class="op">=</span> <span class="st">"count"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Here you see the results of the fit, the effects of the factor on
composition and variability. You also can see the uncertainty around
those effects.</p>
<p>The output is a tibble containing the <strong>Following
columns</strong></p>
<ul>
<li>
<code>cell_group</code> - The cell groups being tested.</li>
<li>
<code>parameter</code> - The parameter being estimated from the
design matrix described by the input <code>formula_composition</code>
and <code>formula_variability</code>.</li>
<li>
<code>factor</code> - The covariate factor in the formula, if
applicable (e.g., not present for Intercept or contrasts).</li>
<li>
<code>c_lower</code> - Lower (2.5%) quantile of the posterior
distribution for a composition (c) parameter.</li>
<li>
<code>c_effect</code> - Mean of the posterior distribution for a
composition (c) parameter.</li>
<li>
<code>c_upper</code> - Upper (97.5%) quantile of the posterior
distribution for a composition (c) parameter.</li>
<li>
<code>c_pH0</code> - Probability of the null hypothesis (no
difference) for a composition (c). This is not a p-value.</li>
<li>
<code>c_FDR</code> - False-discovery rate of the null hypothesis for
a composition (c).</li>
<li>
<code>v_lower</code> - Lower (2.5%) quantile of the posterior
distribution for a variability (v) parameter.</li>
<li>
<code>v_effect</code> - Mean of the posterior distribution for a
variability (v) parameter.</li>
<li>
<code>v_upper</code> - Upper (97.5%) quantile of the posterior
distribution for a variability (v) parameter.</li>
<li>
<code>v_pH0</code> - Probability of the null hypothesis for a
variability (v).</li>
<li>
<code>v_FDR</code> - False-discovery rate of the null hypothesis for
a variability (v).</li>
<li>
<code>count_data</code> - Nested input count data.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 36 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 72 × 19</span></span></span>
<span><span class="co">##    cell_group parameter  factor  c_lower c_effect c_upper   c_pH0   c_FDR c_rhat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercep… <span style="color: #BB0000;">NA</span>      0.954      1.20   1.46   0       0   <span style="color: #949494;"> </span>    1.000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer type   -<span style="color: #BB0000;">0.927</span>     -<span style="color: #BB0000;">0.615</span> -<span style="color: #BB0000;">0.322</span>  0       0   <span style="color: #949494;"> </span>    1.000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercep… <span style="color: #BB0000;">NA</span>      0.511      0.774  1.03   0       0   <span style="color: #949494;"> </span>    1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer type   -<span style="color: #BB0000;">0.976</span>     -<span style="color: #BB0000;">0.673</span> -<span style="color: #BB0000;">0.352</span>  0.001<span style="text-decoration: underline;">50</span> 1.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>  1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercep… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.583</span>     -<span style="color: #BB0000;">0.332</span> -<span style="color: #BB0000;">0.070</span><span style="color: #BB0000; text-decoration: underline;">1</span> 0.042<span style="text-decoration: underline;">0</span>  3.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>  1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer type   -<span style="color: #BB0000;">0.573</span>     -<span style="color: #BB0000;">0.276</span>  0.034<span style="text-decoration: underline;">0</span> 0.124   2.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.000</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercep… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.23</span>      -<span style="color: #BB0000;">0.971</span> -<span style="color: #BB0000;">0.725</span>  0       0   <span style="color: #949494;"> </span>    1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer type   -<span style="color: #BB0000;">0.586</span>     -<span style="color: #BB0000;">0.285</span>  0.020<span style="text-decoration: underline;">4</span> 0.115   2.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercep… <span style="color: #BB0000;">NA</span>      0.196      0.366  0.530  0       0   <span style="color: #949494;"> </span>    1.00 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer type   -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">99</span>    0.209  0.429  0.158   3.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>  1.000</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 62 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 10 more variables: c_ess_bulk &lt;dbl&gt;, c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_rhat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="outlier-identification">Outlier Identification<a class="anchor" aria-label="anchor" href="#outlier-identification"></a>
</h3>
<p><code>sccomp</code> can identify outliers probabilistically and
exclude them from the estimation.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>    abundance <span class="op">=</span> <span class="st">"count"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># max_sampling_iterations is used to reduce draw file sizes</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualization-and-summary-plots">Visualization and Summary Plots<a class="anchor" aria-label="anchor" href="#visualization-and-summary-plots"></a>
</h3>
<p>A plot of group proportions, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is descriptively
adequate for the data, the blue boxplots should roughly overlay the
black boxplots, which represent the observed data. The outliers are
coloured in red. A boxplot will be returned for every (discrete)
covariate present in formula_composition. The colour coding represents
the significant associations for composition and/or variability.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<p><img src="introduction_files/figure-html/boxplot-basic-1.png" class="r-plt" width="700"></p>
<p>You can plot proportions adjusted for unwanted effects. This is
helpful especially for complex models, where multiple factors can
significantly impact the proportions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span>, remove_unwanted_effects <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<p><img src="introduction_files/figure-html/boxplot-adjusted-1.png" class="r-plt" width="700"></p>
<div class="section level4">
<h4 id="adding-custom-layers-to-boxplots">Adding Custom Layers to Boxplots<a class="anchor" aria-label="anchor" href="#adding-custom-layers-to-boxplots"></a>
</h4>
<p>Since <code><a href="../reference/sccomp_boxplot.html">sccomp_boxplot()</a></code> returns a ggplot object, you can
add additional layers to customize the plot. For example, you can add
labels showing FDR values or other statistics:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>  <span class="co"># Extendd the ylim to accomodate the label</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">0.35</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"FDR: %.3f"</span>, <span class="va">c_FDR</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             x <span class="op">=</span> <span class="fl">1.5</span>, y <span class="op">=</span> <span class="fl">0.45</span>,</span>
<span>             size <span class="op">=</span> <span class="fl">2</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<p><img src="introduction_files/figure-html/boxplot-fdr-labels-1.png" class="r-plt" width="700"></p>
<p>You can also add text annotations for p-values:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_boxplot.html">sccomp_boxplot</a></span><span class="op">(</span>factor <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>  <span class="co"># Extendd the ylim to accomodate the label</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">0.35</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"p: %.3f"</span>, <span class="va">c_pH0</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             x <span class="op">=</span> <span class="fl">1.5</span>, y <span class="op">=</span> <span class="fl">0.45</span>,</span>
<span>             size <span class="op">=</span> <span class="fl">2</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<p><img src="introduction_files/figure-html/boxplot-pvalue-labels-1.png" class="r-plt" width="700"></p>
<p>A plot of estimates of differential composition (c_) on the x-axis
and differential variability (v_) on the y-axis. The error bars
represent 95% credible intervals. The dashed lines represent the minimal
effect that the hypothesis test is based on. An effect is labelled as
significant if it exceeds the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/plot_1D_intervals.html">plot_1D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot-1d-intervals-1.png" class="r-plt" width="700"></p>
<p>We can plot the relationship between abundance and variability. As we
can see below, they are positively correlated. sccomp models this
relationship to obtain a shrinkage effect on the estimates of both the
abundance and the variability. This shrinkage is adaptive as it is
modelled jointly, thanks to Bayesian inference.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/plot_2D_intervals.html">plot_2D_intervals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot-2d-intervals-1.png" class="r-plt" width="700"></p>
<p>You can produce the series of plots calling the <code>plot</code>
method.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="model-proportions-directly-e-g--from-deconvolution">Model Proportions Directly (e.g. from deconvolution)<a class="anchor" aria-label="anchor" href="#model-proportions-directly-e-g--from-deconvolution"></a>
</h3>
<p><strong>Note:</strong> If counts are available, we strongly
discourage the use of proportions, as an important source of uncertainty
(i.e., for rare groups/cell types) is not modeled.</p>
<p>The use of proportions is better suited for modelling deconvolution
results (e.g., of bulk RNA data), in which case counts are not
available.</p>
<p>Proportions should be greater than 0. Assuming that zeros derive from
a precision threshold (e.g., deconvolution), zeros are converted to the
smallest non-zero value.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>    abundance <span class="op">=</span> <span class="st">"proportion"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="continuous-factor-analysis">Continuous Factor Analysis<a class="anchor" aria-label="anchor" href="#continuous-factor-analysis"></a>
</h3>
<p><code>sccomp</code> is able to fit arbitrary complex models. In this
example we have a continuous and binary covariate.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span></span>
<span>    <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="va">continuous_covariate</span>, </span>
<span>      sample <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>      cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>      max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type + continuous_covariate </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 90 × 15</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_rhat c_ess_bulk</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interce… <span style="color: #BB0000;">NA</span>       0.556   0.836    1.12   1.00       <span style="text-decoration: underline;">1</span>820.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typeheal… type     1.03    1.35     1.67   1.000      <span style="text-decoration: underline;">1</span>982.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B immature        continuo… conti…  -<span style="color: #BB0000;">0.270</span>   0.047<span style="text-decoration: underline;">2</span>   0.364  1.00       <span style="text-decoration: underline;">2</span>226.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             (Interce… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">0.984</span>  -<span style="color: #BB0000;">0.675</span>   -<span style="color: #BB0000;">0.354</span>  1.00       <span style="text-decoration: underline;">2</span>097.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B mem             typeheal… type     1.19    1.57     1.93   1.00       <span style="text-decoration: underline;">1</span>891.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B mem             continuo… conti…  -<span style="color: #BB0000;">0.270</span>   0.074<span style="text-decoration: underline;">0</span>   0.404  1.000      <span style="text-decoration: underline;">2</span>050.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm S100A4     (Interce… <span style="color: #BB0000;">NA</span>       1.32    1.55     1.79   1.00       <span style="text-decoration: underline;">2</span>135.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm S100A4     typeheal… type     0.862   1.13     1.40   1.000      <span style="text-decoration: underline;">2</span>165.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm S100A4     continuo… conti…  -<span style="color: #BB0000;">0.107</span>   0.177    0.448  1.00       <span style="text-decoration: underline;">1</span>876.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm high cyto… (Interce… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">0.924</span>  -<span style="color: #BB0000;">0.585</span>   -<span style="color: #BB0000;">0.235</span>  1.000      <span style="text-decoration: underline;">1</span>939.</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 80 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_rhat &lt;dbl&gt;, v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="random-effect-modeling-mixed-effect-modeling">Random Effect Modeling (Mixed-Effect Modeling)<a class="anchor" aria-label="anchor" href="#random-effect-modeling-mixed-effect-modeling"></a>
</h3>
<p><code>sccomp</code> supports multilevel modeling by allowing the
inclusion of random effects in the compositional and variability
formulas. This is particularly useful when your data has hierarchical or
grouped structures, such as measurements nested within subjects,
batches, or experimental units. By incorporating random effects, sccomp
can account for variability at different levels of your data, improving
model fit and inference accuracy.</p>
<div class="section level4">
<h4 id="random-intercept-model">Random Intercept Model<a class="anchor" aria-label="anchor" href="#random-intercept-model"></a>
</h4>
<p>In this example, we demonstrate how to fit a random intercept model
using sccomp. We’ll model the cell-type proportions with both fixed
effects (e.g., treatment) and random effects (e.g., subject-specific
variability).</p>
<p>Here is the input data</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 106,297 × 9</span></span></span>
<span><span class="co">##    cell_group nCount_RNA nFeature_RNA group__ group__wrong sample type  group2__</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> CD4 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Mono clas…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CD4 cm S1…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B immature          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD8 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 naive           0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Mono clas…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm S1…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm hi…          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> B immature          0            0 GROUP1  1            SI-GA… canc… GROUP21 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 106,287 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: continuous_covariate &lt;dbl&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span> </span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type + (1 | group__) </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 180 × 15</span></span></span>
<span><span class="co">##    cell_group parameter        factor c_lower c_effect c_upper c_rhat c_ess_bulk</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature (Intercept)      <span style="color: #BB0000;">NA</span>      0.552   0.871    1.21     1.01       252.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature typehealthy      type    0.819   1.19     1.68     1.00       122.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.407</span>   0.038<span style="text-decoration: underline;">0</span>   0.467    1.00       136.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.275</span>   0.214    0.661    1.00       106.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.091</span><span style="color: #BB0000; text-decoration: underline;">1</span>  0.229    0.585    1.00       191.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.721</span>  -<span style="color: #BB0000;">0.279</span>    0.059<span style="text-decoration: underline;">1</span>   1.00       155.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> B mem      (Intercept)      <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.842</span>  -<span style="color: #BB0000;">0.472</span>   -<span style="color: #BB0000;">0.054</span><span style="color: #BB0000; text-decoration: underline;">5</span>   1.00       156.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> B mem      typehealthy      type    0.730   1.26     1.71     1.01       107.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> B mem      (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.362</span>   0.009<span style="text-decoration: underline;">74</span>  0.486    1.01       138.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> B mem      (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.024</span><span style="color: #BB0000; text-decoration: underline;">4</span>  0.277    0.790    1.02       135.</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 170 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_rhat &lt;dbl&gt;, v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="random-effect-model-random-slopes">Random Effect Model (random slopes)<a class="anchor" aria-label="anchor" href="#random-effect-model-random-slopes"></a>
</h4>
<p><code>sccomp</code> can model random slopes. We provide an example
below.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="va">type</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span>,</span>
<span>      sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>      cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>      bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type + (type | group__) </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 240 × 15</span></span></span>
<span><span class="co">##    cell_group parameter        factor c_lower c_effect c_upper c_rhat c_ess_bulk</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature (Intercept)      <span style="color: #BB0000;">NA</span>      0.458   0.849    1.28    1.02       108. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature typehealthy      type    0.729   1.22     1.67    1.00        87.9</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.223</span>   0.021<span style="text-decoration: underline;">0</span>   0.347   1.00       153. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B immature typehealthy___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.227</span>   0.002<span style="text-decoration: underline;">07</span>  0.295   1.00       200. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.133</span>   0.124    0.481   1.02       133. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B immature typehealthy___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.119</span>   0.099<span style="text-decoration: underline;">8</span>   0.443   1.00       168. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.039</span><span style="color: #BB0000; text-decoration: underline;">6</span>  0.200    0.560   1.01       123. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.656</span>  -<span style="color: #BB0000;">0.222</span>    0.032<span style="text-decoration: underline;">8</span>  1.000      120. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> B mem      (Intercept)      <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.04</span>   -<span style="color: #BB0000;">0.544</span>   -<span style="color: #BB0000;">0.158</span>   1.00       137. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> B mem      typehealthy      type    0.775   1.32     1.86    1.01        89.0</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 230 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_rhat &lt;dbl&gt;, v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="nested-random-effects">Nested Random Effects<a class="anchor" aria-label="anchor" href="#nested-random-effects"></a>
</h4>
<p>If you have a more complex hierarchy, such as measurements nested
within subjects and subjects nested within batches, you can include
multiple grouping variables. Here <code>group2__</code> is nested within
<code>group__</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span></span>
<span>      formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="op">(</span><span class="va">type</span> <span class="op">|</span> <span class="va">group__</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">group2__</span><span class="op">)</span>,</span>
<span>      sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>      cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>      bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type + (type | group__) + (1 | group2__) </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 300 × 15</span></span></span>
<span><span class="co">##    cell_group parameter        factor c_lower c_effect c_upper c_rhat c_ess_bulk</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature (Intercept)      <span style="color: #BB0000;">NA</span>      0.431    0.787   1.28     1.02      111. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature typehealthy      type    0.742    1.18    1.67     1.01      102. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.192</span>    0.088<span style="text-decoration: underline;">1</span>  0.494    1.02       74.6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B immature typehealthy___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.162</span>    0.079<span style="text-decoration: underline;">4</span>  0.501    1.00      114. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.191</span>    0.051<span style="text-decoration: underline;">8</span>  0.378    1.00      139. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B immature typehealthy___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.184</span>    0.049<span style="text-decoration: underline;">8</span>  0.366    1.02      131. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.095</span><span style="color: #BB0000; text-decoration: underline;">8</span>   0.209   0.625    1.01       80.2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.788</span>   -<span style="color: #BB0000;">0.274</span>   0.016<span style="text-decoration: underline;">0</span>   1.02       77.7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.432</span>   -<span style="color: #BB0000;">0.101</span>   0.129    1.04       98.3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> B immature (Intercept)___G… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.027</span><span style="color: #BB0000; text-decoration: underline;">6</span>   0.231   0.598    1.00       84.1</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 290 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_rhat &lt;dbl&gt;, v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="result-interpretation-and-communication">Result Interpretation and Communication<a class="anchor" aria-label="anchor" href="#result-interpretation-and-communication"></a>
</h3>
<p>The estimated effects are expressed in the unconstrained space of the
parameters, similar to differential expression analysis that expresses
changes in terms of log fold change. However, for differences in
proportion, logit fold change must be used, which is harder to interpret
and understand.</p>
<p>Therefore, we provide a more intuitive proportional fold change that
can be more easily understood. However, these cannot be used to infer
significance (use sccomp_test() instead), and a lot of care must be
taken given the nonlinearity of these measures (a 1-fold increase from
0.0001 to 0.0002 carries a different weight than a 1-fold increase from
0.4 to 0.8).</p>
<p>From your estimates, you can specify which effects you are interested
in (this can be a subset of the full model if you wish to exclude
unwanted effects), and the two points you would like to compare.</p>
<p>In the case of a categorical variable, the starting and ending points
are categories.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">|&gt;</span> </span>
<span>   <span class="fu"><a href="../reference/sccomp_proportional_fold_change.html">sccomp_proportional_fold_change</a></span><span class="op">(</span></span>
<span>     formula_composition <span class="op">=</span> <span class="op">~</span>  <span class="va">type</span>,</span>
<span>     from <span class="op">=</span>  <span class="st">"healthy"</span>, </span>
<span>     to <span class="op">=</span> <span class="st">"cancer"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_group</span>, <span class="va">statement</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 30 × 2</span></span></span>
<span><span class="co">##    cell_group           statement                                </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;glue&gt;</span>                                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature           2-fold decrease (from 0.1025 to 0.0508)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B mem                2.3-fold decrease (from 0.0331 to 0.0146)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CD4 cm high cytokine 7.8-fold increase (from 0.0016 to 0.0124)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> CD4 cm ribosome      3.5-fold increase (from 0.0075 to 0.0261)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4        1.6-fold decrease (from 0.1592 to 0.1016)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 em high cytokine 4.8-fold increase (from 0.0023 to 0.0109)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 naive            1.5-fold decrease (from 0.1137 to 0.0774)</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 ribosome         3-fold decrease (from 0.0868 to 0.0288)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD8 em 1             1.2-fold increase (from 0.0491 to 0.06)  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD8 em 2             3.5-fold increase (from 0.0062 to 0.0218)</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 20 more rows</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="contrasts-analysis">Contrasts Analysis<a class="anchor" aria-label="anchor" href="#contrasts-analysis"></a>
</h3>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~0 + type </span></span>
<span><span class="co">##   Variability formula: ~1 </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 11</span></span></span>
<span><span class="co">##    cell_group   parameter factor c_lower c_effect c_upper   c_pH0   c_FDR c_rhat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature   typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.89</span>    -<span style="color: #BB0000;">1.33</span>   -<span style="color: #BB0000;">0.820</span> 0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature   typeheal… <span style="color: #BB0000;">NA</span>       0.820    1.33    1.89  0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem        typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.26</span>    -<span style="color: #BB0000;">1.65</span>   -<span style="color: #BB0000;">1.11</span>  0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem        typeheal… <span style="color: #BB0000;">NA</span>       1.11     1.65    2.26  0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100… typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.47</span>    -<span style="color: #BB0000;">0.991</span>  -<span style="color: #BB0000;">0.503</span> 5.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 8.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>     <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100… typeheal… <span style="color: #BB0000;">NA</span>       0.503    0.991   1.47  5.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 8.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>     <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high… typecanc… <span style="color: #BB0000;">NA</span>       0.969    1.58    2.17  0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.17</span>    -<span style="color: #BB0000;">1.58</span>   -<span style="color: #BB0000;">0.969</span> 0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>       <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribo… typecanc… <span style="color: #BB0000;">NA</span>       0.362    0.948   1.54  1.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 4.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>     <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribo… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.54</span>    -<span style="color: #BB0000;">0.948</span>  -<span style="color: #BB0000;">0.362</span> 1.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 4.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>     <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: c_ess_bulk &lt;dbl&gt;, c_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="categorical-factor-analysis-bayesian-anova">Categorical Factor Analysis (Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-analysis-bayesian-anova"></a>
</h3>
<p>This is achieved through model comparison with <code>loo</code>. In
the following example, the model with association with factors better
fits the data compared to the baseline model with no factor association.
For model comparisons <code><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers()</a></code> must not be
executed as the leave-one-out must work with the same amount of data,
while outlier elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5
<code>se_diff</code> difference of 5, we are confident that a model is
better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>.
In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model
one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>, </span>
<span>    inference_method <span class="op">=</span> <span class="st">"hmc"</span>,</span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>, </span>
<span>    inference_method <span class="op">=</span> <span class="st">"hmc"</span>,</span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">model_with_factor_association</span>, <span class="st">"fit"</span><span class="op">)</span><span class="op">$</span><span class="fu">loo</span><span class="op">(</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">model_without_association</span>, <span class="st">"fit"</span><span class="op">)</span><span class="op">$</span><span class="fu">loo</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        elpd_diff se_diff</span></span>
<span><span class="co">## model1   0.0       0.0  </span></span>
<span><span class="co">## model2 -80.4      10.6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="differential-variability-analysis">Differential Variability Analysis<a class="anchor" aria-label="anchor" href="#differential-variability-analysis"></a>
</h3>
<p>We can model the cell-group variability also dependent on the type,
and so test differences in variability</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    sample <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    cell_group <span class="op">=</span> <span class="st">"cell_group"</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    max_sampling_iterations <span class="op">=</span> <span class="fl">2000</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp model</span></span>
<span><span class="co">## ============</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Model specifications:</span></span>
<span><span class="co">##   Family: multi_beta_binomial </span></span>
<span><span class="co">##   Composition formula: ~type </span></span>
<span><span class="co">##   Variability formula: ~type </span></span>
<span><span class="co">##   Inference method: pathfinder </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Data: Samples: 20   Cell groups: 30 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Column prefixes: c_ -&gt; composition parameters  v_ -&gt; variability parameters</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Convergence diagnostics:</span></span>
<span><span class="co">##   For each parameter, n_eff is the effective sample size and R_k_hat is the potential</span></span>
<span><span class="co">##   scale reduction factor on split chains (at convergence, R_k_hat = 1).</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 15</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_rhat c_ess_bulk</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interce… <span style="color: #BB0000;">NA</span>      0.508     0.818   1.12   1.000     <span style="text-decoration: underline;">1</span>790. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typeheal… type    1.04      1.39    1.74   1.000     <span style="text-decoration: underline;">1</span>497. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem             (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.05</span>     -<span style="color: #BB0000;">0.681</span>  -<span style="color: #BB0000;">0.337</span>  1.000      356. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             typeheal… type    1.21      1.61    2.02   1.000      441. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4     (Interce… <span style="color: #BB0000;">NA</span>      1.45      1.71    1.98   1.00      <span style="text-decoration: underline;">1</span>801. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100A4     typeheal… type    0.599     0.896   1.18   1.00       587. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high cyto… (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.984</span>    -<span style="color: #BB0000;">0.594</span>  -<span style="color: #BB0000;">0.199</span>  1.00      <span style="text-decoration: underline;">1</span>828. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high cyto… typeheal… type   -<span style="color: #BB0000;">1.92</span>     -<span style="color: #BB0000;">1.23</span>   -<span style="color: #BB0000;">0.584</span>  1.00        88.4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribosome   (Interce… <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.071</span><span style="color: #BB0000; text-decoration: underline;">8</span>    0.301   0.682  1.000     <span style="text-decoration: underline;">1</span>981. </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribosome   typeheal… type   -<span style="color: #BB0000;">1.35</span>     -<span style="color: #BB0000;">0.925</span>  -<span style="color: #BB0000;">0.496</span>  1.00      <span style="text-decoration: underline;">1</span>426. </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: c_ess_tail &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_rhat &lt;dbl&gt;, v_ess_bulk &lt;dbl&gt;, v_ess_tail &lt;dbl&gt;</span></span></span></code></pre>
<p><strong>Plot 1D significance plot</strong></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running standalone generated quantities after 1 MCMC chain, with 1 thread(s) per chain...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Chain 1 finished in 0.0 seconds.</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot-1d-significance-1.png" class="r-plt" width="700"></p>
<p><strong>Plot 2D significance plot</strong> Data points are cell
groups. Error bars are the 95% credible interval. The dashed lines
represent the default threshold fold change for which the probabilities
(c_pH0, v_pH0) are calculated. pH0 of 0 represent the rejection of the
null hypothesis that no effect is observed.</p>
<p>This plot is provided only if differential variability has been
tested. The differential variability estimates are reliable only if the
linear association between mean and variability for
<code>(intercept)</code> (left-hand side facet) is satisfied. A
scatterplot (besides the Intercept) is provided for each category of
interest. For each category of interest, the composition and variability
effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot-2d-significance-1.png" class="r-plt" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="recommended-settings-for-different-data-types">Recommended Settings for Different Data Types<a class="anchor" aria-label="anchor" href="#recommended-settings-for-different-data-types"></a>
</h2>
<div class="section level3">
<h3 id="for-single-cell-rna-sequencing">For Single-Cell RNA Sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = TRUE</code>. The
bimodality of the mean-variability association can be confirmed from the
plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level3">
<h3 id="for-cytof-and-microbiome-data">For CyTOF and Microbiome Data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = FALSE</code>
(Default).</p>
</div>
<div class="section level3">
<h3 id="mcmc-chain-visualization">MCMC Chain Visualization<a class="anchor" aria-label="anchor" href="#mcmc-chain-visualization"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example, we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that it has converged and is negative
with probability 1.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/cmdstanr/" class="external-link">cmdstanr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/posterior/" class="external-link">posterior</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/bayesplot/" class="external-link">bayesplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assuming res contains the fit object from cmdstanr</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract draws for 'beta[2,1]'</span></span>
<span><span class="va">draws</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mc-stan.org/posterior/reference/draws_array.html" class="external-link">as_draws_array</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="fu">draws</span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a traceplot for 'beta[2,1]'</span></span>
<span><span class="fu"><a href="https://mc-stan.org/bayesplot/reference/MCMC-traces.html" class="external-link">mcmc_trace</a></span><span class="op">(</span><span class="va">draws</span>, pars <span class="op">=</span> <span class="st">"beta[2,1]"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/mcmc-traceplot-1.png" class="r-plt" width="700"></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bayesplot_1.14.0  posterior_1.6.1   cmdstanr_0.9.0    loo_2.8.0        </span></span>
<span><span class="co">##  [5] tidyr_1.3.1       forcats_1.0.1     ggplot2_4.0.1     sccomp_2.1.21    </span></span>
<span><span class="co">##  [9] instantiate_0.2.3 dplyr_1.1.4      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.1            farver_2.1.2               </span></span>
<span><span class="co">##  [3] S7_0.2.1                    fastmap_1.2.0              </span></span>
<span><span class="co">##  [5] SingleCellExperiment_1.32.0 tensorA_0.36.2.1           </span></span>
<span><span class="co">##  [7] dotCall64_1.2               digest_0.6.39              </span></span>
<span><span class="co">##  [9] lifecycle_1.0.4             SeuratObject_5.2.0         </span></span>
<span><span class="co">## [11] processx_3.8.6              magrittr_2.0.4             </span></span>
<span><span class="co">## [13] compiler_4.5.2              rlang_1.1.6                </span></span>
<span><span class="co">## [15] sass_0.4.10                 tools_4.5.2                </span></span>
<span><span class="co">## [17] utf8_1.2.6                  yaml_2.3.10                </span></span>
<span><span class="co">## [19] data.table_1.17.8           knitr_1.50                 </span></span>
<span><span class="co">## [21] labeling_0.4.3              S4Arrays_1.10.0            </span></span>
<span><span class="co">## [23] htmlwidgets_1.6.4           sp_2.2-0                   </span></span>
<span><span class="co">## [25] DelayedArray_0.36.0         plyr_1.8.9                 </span></span>
<span><span class="co">## [27] RColorBrewer_1.1-3          abind_1.4-8                </span></span>
<span><span class="co">## [29] withr_3.0.2                 purrr_1.2.0                </span></span>
<span><span class="co">## [31] BiocGenerics_0.56.0         desc_1.4.3                 </span></span>
<span><span class="co">## [33] grid_4.5.2                  stats4_4.5.2               </span></span>
<span><span class="co">## [35] future_1.68.0               progressr_0.18.0           </span></span>
<span><span class="co">## [37] globals_0.18.0              scales_1.4.0               </span></span>
<span><span class="co">## [39] SummarizedExperiment_1.40.0 cli_3.6.5                  </span></span>
<span><span class="co">## [41] rmarkdown_2.30              crayon_1.5.3               </span></span>
<span><span class="co">## [43] ragg_1.5.0                  generics_0.1.4             </span></span>
<span><span class="co">## [45] future.apply_1.20.0         reshape2_1.4.5             </span></span>
<span><span class="co">## [47] tzdb_0.5.0                  cachem_1.1.0               </span></span>
<span><span class="co">## [49] stringr_1.6.0               parallel_4.5.2             </span></span>
<span><span class="co">## [51] XVector_0.50.0              matrixStats_1.5.0          </span></span>
<span><span class="co">## [53] vctrs_0.6.5                 Matrix_1.7-4               </span></span>
<span><span class="co">## [55] jsonlite_2.0.0              callr_3.7.6                </span></span>
<span><span class="co">## [57] IRanges_2.44.0              hms_1.1.4                  </span></span>
<span><span class="co">## [59] patchwork_1.3.2             S4Vectors_0.48.0           </span></span>
<span><span class="co">## [61] ggrepel_0.9.6               listenv_0.10.0             </span></span>
<span><span class="co">## [63] systemfonts_1.3.1           jquerylib_0.1.4            </span></span>
<span><span class="co">## [65] spam_2.11-1                 parallelly_1.45.1          </span></span>
<span><span class="co">## [67] glue_1.8.0                  pkgdown_2.2.0              </span></span>
<span><span class="co">## [69] codetools_0.2-20            ps_1.9.1                   </span></span>
<span><span class="co">## [71] distributional_0.5.0        stringi_1.8.7              </span></span>
<span><span class="co">## [73] gtable_0.3.6                GenomicRanges_1.62.0       </span></span>
<span><span class="co">## [75] tibble_3.3.0                pillar_1.11.1              </span></span>
<span><span class="co">## [77] htmltools_0.5.8.1           Seqinfo_1.0.0              </span></span>
<span><span class="co">## [79] R6_2.6.1                    textshaping_1.0.4          </span></span>
<span><span class="co">## [81] evaluate_1.0.5              lattice_0.22-7             </span></span>
<span><span class="co">## [83] Biobase_2.70.0              readr_2.1.6                </span></span>
<span><span class="co">## [85] backports_1.5.0             bslib_0.9.0                </span></span>
<span><span class="co">## [87] Rcpp_1.1.0                  SparseArray_1.10.3         </span></span>
<span><span class="co">## [89] checkmate_2.3.3             xfun_0.54                  </span></span>
<span><span class="co">## [91] fs_1.6.6                    MatrixGenerics_1.22.0      </span></span>
<span><span class="co">## [93] prettydoc_0.4.1             pkgconfig_2.0.3</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stefano Mangiola, Alexandra J. Roth-Schulze, Marie Trussart, Enrique Zozaya-Valdés, Mengyao Ma, Zijie Gao, Alan F. Rubin, Terence P. Speed, Heejung Shim, Anthony T. Papenfuss.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
