<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the sccomp package • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the sccomp package">
<meta property="og:description" content="sccomp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the sccomp package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2024-04-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Sccomp is a generalised method for differential composition and
      variability analyses. It tests differences in cell type
      proportion, for count data. It is able of flexible
      continuous/discrete models, removal of unwanted variation, removal
      of outliers, multilevel random/effect modelling.</p>
    </div>
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<p>Sccomp is a generalised method for differential composition and
variability analyses.</p>
<div class="section level3">
<h3 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h3>
<ul>
<li>Modelling counts</li>
<li>Modelling proportionality</li>
<li>Modelling cell-type specific variability</li>
<li>Cell-type information share for variability shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p><code>sccomp</code> can model changes in composition and variability.
By default, the formula for variability is either <code>~1</code>, which
assumes that the cell-group variability is independent of any covariate
or <code>~ factor_of_interest</code>, which assumes that the model is
dependent on the factor of interest only. The variability model must be
a subset of the model for composition.</p>
<div class="section level3">
<h3 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h3>
<p>Of the output table, the estimate columns start with the prefix
<code>c_</code> indicate <code>composition</code>, or with
<code>v_</code> indicate <code>variability</code> (when
formula_variability is set).</p>
<div class="section level4">
<h4 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">single_cell_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sccomp_result</span> <span class="op">=</span> </span>
<span>  <span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Optional</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: EXPERIMENTAL ALGORITHM:</span></span>
<span><span class="co">## Chain 1:   This procedure has not been thoroughly tested and may be unstable</span></span>
<span><span class="co">## Chain 1:   or buggy. The interface is subject to change.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000587 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 5.87 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin eta adaptation.</span></span>
<span><span class="co">## Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 250 / 250 [100%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Success! Found best value [eta = 0.1].</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin stochastic gradient ascent.</span></span>
<span><span class="co">## Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes </span></span>
<span><span class="co">## Chain 1:    100        -4891.421             1.000            1.000</span></span>
<span><span class="co">## Chain 1:    200        -4373.450             0.559            1.000</span></span>
<span><span class="co">## Chain 1:    300        -4167.787             0.389            0.118</span></span>
<span><span class="co">## Chain 1:    400        -4027.529             0.301            0.118</span></span>
<span><span class="co">## Chain 1:    500        -3953.345             0.244            0.049</span></span>
<span><span class="co">## Chain 1:    600        -3885.684             0.206            0.049</span></span>
<span><span class="co">## Chain 1:    700        -3839.685             0.179            0.035</span></span>
<span><span class="co">## Chain 1:    800        -3814.016             0.157            0.035</span></span>
<span><span class="co">## Chain 1:    900        -3784.410             0.141            0.019</span></span>
<span><span class="co">## Chain 1:   1000        -3766.615             0.127            0.019</span></span>
<span><span class="co">## Chain 1:   1100        -3748.480             0.027            0.017</span></span>
<span><span class="co">## Chain 1:   1200        -3738.424             0.016            0.012</span></span>
<span><span class="co">## Chain 1:   1300        -3727.991             0.011            0.008   MEDIAN ELBO CONVERGED</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Drawing a sample of size 1000 from the approximate posterior... </span></span>
<span><span class="co">## Chain 1: COMPLETED.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: EXPERIMENTAL ALGORITHM:</span></span>
<span><span class="co">## Chain 1:   This procedure has not been thoroughly tested and may be unstable</span></span>
<span><span class="co">## Chain 1:   or buggy. The interface is subject to change.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000449 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 4.49 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin eta adaptation.</span></span>
<span><span class="co">## Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Success! Found best value [eta = 1] earlier than expected.</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin stochastic gradient ascent.</span></span>
<span><span class="co">## Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes </span></span>
<span><span class="co">## Chain 1:    100        -3811.443             1.000            1.000</span></span>
<span><span class="co">## Chain 1:    200        -3534.541             0.539            1.000</span></span>
<span><span class="co">## Chain 1:    300        -3526.073             0.360            0.078</span></span>
<span><span class="co">## Chain 1:    400        -3522.602             0.270            0.078</span></span>
<span><span class="co">## Chain 1:    500        -3515.808             0.217            0.002   MEDIAN ELBO CONVERGED</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Drawing a sample of size 1000 from the approximate posterior... </span></span>
<span><span class="co">## Chain 1: COMPLETED.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: EXPERIMENTAL ALGORITHM:</span></span>
<span><span class="co">## Chain 1:   This procedure has not been thoroughly tested and may be unstable</span></span>
<span><span class="co">## Chain 1:   or buggy. The interface is subject to change.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000409 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 4.09 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin eta adaptation.</span></span>
<span><span class="co">## Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Success! Found best value [eta = 1] earlier than expected.</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin stochastic gradient ascent.</span></span>
<span><span class="co">## Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes </span></span>
<span><span class="co">## Chain 1:    100        -9466.188             1.000            1.000</span></span>
<span><span class="co">## Chain 1:    200        -2905.502             1.629            2.258</span></span>
<span><span class="co">## Chain 1:    300         2339.702             1.833            2.242</span></span>
<span><span class="co">## Chain 1:    400         2718.365             1.410            2.242</span></span>
<span><span class="co">## Chain 1:    500         2516.492             1.144            1.000</span></span>
<span><span class="co">## Chain 1:    600          235.495             2.568            2.242</span></span>
<span><span class="co">## Chain 1:    700         1180.584             2.315            1.000</span></span>
<span><span class="co">## Chain 1:    800         -947.354             2.307            2.242</span></span>
<span><span class="co">## Chain 1:    900         1069.079             2.260            1.886</span></span>
<span><span class="co">## Chain 1:   1000         1316.863             2.053            1.886</span></span>
<span><span class="co">## Chain 1:   1100          697.300             2.041            1.886   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1200          317.202             1.936            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1300         1411.313             1.789            0.889   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1400         -762.576             2.060            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1500         2494.403             2.183            1.306   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1600         2663.433             1.220            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1700          725.182             1.408            1.306   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1800         1360.345             1.230            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   1900         2253.368             1.081            0.889   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2000          744.798             1.264            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2100         2310.417             1.243            1.198   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2200          621.376             1.395            1.306   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2300          655.624             1.323            1.306   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2400          885.798             1.064            0.678   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2500         1225.384             0.961            0.467   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2600         -809.764             1.206            0.678   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2700          -16.980             5.608            0.678   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2800          935.586             5.663            1.018   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   2900         1277.672             5.650            1.018   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3000         1681.268             5.471            0.678   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3100          307.709             5.850            1.018   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3200         3224.238             5.669            0.905   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3300         2403.629             5.698            0.905   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3400          347.114             6.264            1.018   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3500         2109.195             6.320            1.018   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3600         1060.996             6.167            0.988   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3700         2522.941             1.556            0.905   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3800         2208.428             1.469            0.835   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   3900         1120.818             1.539            0.905   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4000         2248.750             1.565            0.905   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4100         3723.472             1.158            0.835   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4200          910.170             1.377            0.835   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4300          339.658             1.511            0.970   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4400         1067.184             0.987            0.835   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4500         4118.226             0.977            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4600         3465.167             0.897            0.682   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4700         1721.560             0.940            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4800         1291.834             0.960            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   4900         -741.074             1.137            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5000         -550.348             1.121            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5100        -1926.296             1.153            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5200         -879.097             0.963            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5300         -701.095             0.821            0.714   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5400          221.474             1.169            0.741   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5500         -568.733             1.234            1.013   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5600         1264.823             1.360            1.191   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5700          264.024             1.638            1.389   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5800          726.554             1.668            1.389   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   5900         1807.750             1.454            1.191   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6000         1238.488             1.465            1.191   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6100         2011.635             1.432            1.191   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6200         2137.621             1.319            0.637   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6300         1441.622             1.342            0.637   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6400         3438.823             0.983            0.598   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6500         1255.307             1.018            0.598   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6600         -131.083             1.931            0.598   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6700          -58.489             1.676            0.598   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6800         2316.926             1.715            0.598   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   6900         1163.554             1.754            0.991   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7000         2498.845             1.761            0.991   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7100          572.743             2.059            1.025   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7200         1927.987             2.124            1.025   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7300         2231.369             2.089            1.025   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7400         1673.812             2.064            1.025   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7500          518.999             2.113            1.025   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7600         2436.283             1.134            0.991   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7700         2197.541             1.021            0.787   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7800         2252.730             0.921            0.703   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   7900          486.562             1.184            0.703   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8000         1386.729             1.196            0.703   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8100          846.500             0.923            0.649   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8200         1366.006             0.891            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8300         2299.396             0.918            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8400         2330.357             0.886            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8500           41.563             6.170            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8600         -318.317             6.205            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8700         1275.856             6.319            0.649   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8800         1644.528             6.339            0.649   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   8900         1100.184             6.025            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9000          241.909             6.315            0.638   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9100         1810.424             6.338            0.866   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9200         1728.613             6.305            0.866   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9300         2176.903             6.285            0.866   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9400         1728.294             6.309            0.866   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9500         1942.485             0.814            0.495   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9600         2134.864             0.710            0.260   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9700         1972.166             0.593            0.224   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9800          676.539             0.762            0.260   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   9900         3170.102             0.791            0.260   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1:   10000         1263.247             0.587            0.260   MAY BE DIVERGING... INSPECT ELBO</span></span>
<span><span class="co">## Chain 1: Informational Message: The maximum number of iterations is reached! The algorithm may not have converged.</span></span>
<span><span class="co">## Chain 1: This variational approximation is not guaranteed to be meaningful.</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Drawing a sample of size 1000 from the approximate posterior... </span></span>
<span><span class="co">## Chain 1: COMPLETED.</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">sccomp_result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(cell_group, sample)`</span></span>
<span><span class="co">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre>
<p>A plot of group proportion, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is likely to be
descriptively adequate to the data, the blue box plot should roughly
overlay with the black box plot, which represents the observed data. The
outliers are coloured in red. A box plot will be returned for every
(discrete) covariate present in <code>formula_composition</code>. The
colour coding represents the significant associations for composition
and/or variability.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">boxplot</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>A plot of estimates of differential composition (c_) on the x-axis
and differential variability (v_) on the y-axis. The error bars
represent 95% credible intervals. The dashed lines represent the minimal
effect that the hypothesis test is based on. An effect is labelled as
significant if bigger than the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We can plot the relationship between abundance and variability. As we
can see below, they are positively correlated, you also appreciate that
this relationship is by model for single cell RNA sequencing data.</p>
<p><code>sccomp</code> models, these relationship to obtain a shrinkage
effect on the estimates of both the abundance and the variability. This
shrinkage is adaptive as it is modelled jointly, thanks for Bayesian
inference.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: EXPERIMENTAL ALGORITHM:</span></span>
<span><span class="co">## Chain 1:   This procedure has not been thoroughly tested and may be unstable</span></span>
<span><span class="co">## Chain 1:   or buggy. The interface is subject to change.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000345 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.45 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin eta adaptation.</span></span>
<span><span class="co">## Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Success! Found best value [eta = 1] earlier than expected.</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin stochastic gradient ascent.</span></span>
<span><span class="co">## Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes </span></span>
<span><span class="co">## Chain 1:    100        -3463.463             1.000            1.000</span></span>
<span><span class="co">## Chain 1:    200        -3163.286             0.547            1.000</span></span>
<span><span class="co">## Chain 1:    300        -3164.252             0.365            0.095</span></span>
<span><span class="co">## Chain 1:    400        -3157.519             0.274            0.095</span></span>
<span><span class="co">## Chain 1:    500        -3158.173             0.220            0.002   MEDIAN ELBO CONVERGED</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Drawing a sample of size 1000 from the approximate posterior... </span></span>
<span><span class="co">## Chain 1: COMPLETED.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 18</span></span></span>
<span><span class="co">##    cell_group  parameter factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature  typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.94</span>    -<span style="color: #BB0000;">1.36</span>   -<span style="color: #BB0000;">0.745</span> 0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature  typeheal… <span style="color: #BB0000;">NA</span>       0.745    1.36    1.94  0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem       typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.28</span>    -<span style="color: #BB0000;">1.64</span>   -<span style="color: #BB0000;">1.04</span>  0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem       typeheal… <span style="color: #BB0000;">NA</span>       1.04     1.64    2.28  0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S10… typecanc… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.58</span>    -<span style="color: #BB0000;">1.15</span>   -<span style="color: #BB0000;">0.769</span> 0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S10… typeheal… <span style="color: #BB0000;">NA</span>       0.769    1.15    1.58  0       0   <span style="color: #949494;"> </span>        <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm hig… typecanc… <span style="color: #BB0000;">NA</span>       0.595    1.57    2.55  0.003<span style="text-decoration: underline;">00</span> 8.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm hig… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.55</span>    -<span style="color: #BB0000;">1.57</span>   -<span style="color: #BB0000;">0.595</span> 0.003<span style="text-decoration: underline;">00</span> 8.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm rib… typecanc… <span style="color: #BB0000;">NA</span>       0.347    0.939   1.52  0.004<span style="text-decoration: underline;">00</span> 1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm rib… typeheal… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.52</span>    -<span style="color: #BB0000;">0.939</span>  -<span style="color: #BB0000;">0.347</span> 0.004<span style="text-decoration: underline;">00</span> 1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="categorical-factor-e-g--bayesian-anova">Categorical factor (e.g. Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-e-g--bayesian-anova"></a>
</h3>
<p>This is achieved through model comparison with <code>loo</code>. In
the following example, the model with association with factors better
fits the data compared to the baseline model with no factor association.
For comparisons <code>check_outliers</code> must be set to FALSE as the
leave-one-out must work with the same amount of data, while outlier
elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5
<code>se_diff</code> difference of 5, we are confident that a model is
better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>.
In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model
one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> , </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>  <span class="va">model_with_factor_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">model_without_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h3>
<p>We can model the cell-group variability also dependent on the type,
and so test differences in variability</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: EXPERIMENTAL ALGORITHM:</span></span>
<span><span class="co">## Chain 1:   This procedure has not been thoroughly tested and may be unstable</span></span>
<span><span class="co">## Chain 1:   or buggy. The interface is subject to change.</span></span>
<span><span class="co">## Chain 1: ------------------------------------------------------------</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.00036 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.6 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin eta adaptation.</span></span>
<span><span class="co">## Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Iteration: 250 / 250 [100%]  (Adaptation)</span></span>
<span><span class="co">## Chain 1: Success! Found best value [eta = 0.1].</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Begin stochastic gradient ascent.</span></span>
<span><span class="co">## Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes </span></span>
<span><span class="co">## Chain 1:    100        -4362.156             1.000            1.000</span></span>
<span><span class="co">## Chain 1:    200        -3780.244             0.577            1.000</span></span>
<span><span class="co">## Chain 1:    300        -3568.887             0.404            0.154</span></span>
<span><span class="co">## Chain 1:    400        -3445.120             0.312            0.154</span></span>
<span><span class="co">## Chain 1:    500        -3379.177             0.254            0.059</span></span>
<span><span class="co">## Chain 1:    600        -3321.402             0.214            0.059</span></span>
<span><span class="co">## Chain 1:    700        -3290.427             0.185            0.036</span></span>
<span><span class="co">## Chain 1:    800        -3261.035             0.163            0.036</span></span>
<span><span class="co">## Chain 1:    900        -3242.879             0.146            0.020</span></span>
<span><span class="co">## Chain 1:   1000        -3223.884             0.132            0.020</span></span>
<span><span class="co">## Chain 1:   1100        -3214.270             0.032            0.017</span></span>
<span><span class="co">## Chain 1:   1200        -3205.477             0.017            0.009   MEDIAN ELBO CONVERGED</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Drawing a sample of size 1000 from the approximate posterior... </span></span>
<span><span class="co">## Chain 1: COMPLETED.</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 × 14</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interce… <span style="color: #BB0000;">NA</span>       0.203    0.746   1.28      <span style="color: #BB0000;">NaN</span>      3.56</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typeheal… type     0.336    1.18    2.03      <span style="color: #BB0000;">NaN</span>      3.51</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem             (Interce… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.33</span>    -<span style="color: #BB0000;">0.782</span>  -<span style="color: #BB0000;">0.245</span>     <span style="color: #BB0000;">NaN</span>      3.50</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             typeheal… type     0.740    1.53    2.33      <span style="color: #BB0000;">NaN</span>      3.51</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4     (Interce… <span style="color: #BB0000;">NA</span>       0.935    1.60    2.29      <span style="color: #BB0000;">NaN</span>      3.46</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100A4     typeheal… type    -<span style="color: #BB0000;">0.434</span>    0.660   1.78      <span style="color: #BB0000;">NaN</span>      3.52</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high cyto… (Interce… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.07</span>    -<span style="color: #BB0000;">0.399</span>   0.282     <span style="color: #BB0000;">NaN</span>      3.50</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high cyto… typeheal… type    -<span style="color: #BB0000;">1.94</span>    -<span style="color: #BB0000;">0.851</span>   0.204     <span style="color: #BB0000;">NaN</span>      3.57</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribosome   (Interce… <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">0.257</span>    0.340   0.843     <span style="color: #BB0000;">NaN</span>      3.57</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribosome   typeheal… type    -<span style="color: #BB0000;">1.90</span>    -<span style="color: #BB0000;">1.14</span>   -<span style="color: #BB0000;">0.348</span>     <span style="color: #BB0000;">NaN</span>      3.46</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h2>
<div class="section level3">
<h3 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = TRUE</code>. The
bimodality of the mean-variability association can be confirmed from the
plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level3">
<h3 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = FALSE</code>
(Default).</p>
</div>
<div class="section level3">
<h3 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior
distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example, we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that it has converged and is negative
with probability 1.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(cell_group, sample)`</span></span>
<span><span class="co">## Joining with `by = join_by(cell_group, type)`</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Plot 2D significance plot. Data points are cell groups. Error bars
are the 95% credible interval. The dashed lines represent the default
threshold fold change for which the probabilities (c_pH0, v_pH0) are
calculated. pH0 of 0 represent the rejection of the null hypothesis that
no effect is observed.</p>
<p>This plot is provided only if differential variability has been
tested. The differential variability estimates are reliable only if the
linear association between mean and variability for
<code>(intercept)</code> (left-hand side facet) is satisfied. A
scatterplot (besides the Intercept) is provided for each category of
interest. The for each category of interest, the composition and
variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 Patched (2024-04-24 r86483)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rstan_2.32.6       StanHeaders_2.32.7 tidyr_1.3.1        forcats_1.0.0     </span></span>
<span><span class="co">## [5] ggplot2_3.5.1      sccomp_1.7.11      dplyr_1.1.4       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] gridExtra_2.3               inline_0.3.19              </span></span>
<span><span class="co">##   [3] rlang_1.1.3                 magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] matrixStats_1.3.0           compiler_4.4.0             </span></span>
<span><span class="co">##   [7] loo_2.7.0                   systemfonts_1.0.6          </span></span>
<span><span class="co">##   [9] vctrs_0.6.5                 stringr_1.5.1              </span></span>
<span><span class="co">##  [11] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">##  [13] fastmap_1.1.1               XVector_0.43.1             </span></span>
<span><span class="co">##  [15] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">##  [17] rmarkdown_2.26              tzdb_0.4.0                 </span></span>
<span><span class="co">##  [19] UCSC.utils_0.99.7           ragg_1.3.0                 </span></span>
<span><span class="co">##  [21] purrr_1.0.2                 xfun_0.43                  </span></span>
<span><span class="co">##  [23] zlibbioc_1.49.3             cachem_1.0.8               </span></span>
<span><span class="co">##  [25] GenomeInfoDb_1.39.14        jsonlite_1.8.8             </span></span>
<span><span class="co">##  [27] highr_0.10                  DelayedArray_0.29.9        </span></span>
<span><span class="co">##  [29] parallel_4.4.0              R6_2.5.1                   </span></span>
<span><span class="co">##  [31] RColorBrewer_1.1-3          bslib_0.7.0                </span></span>
<span><span class="co">##  [33] stringi_1.8.3               parallelly_1.37.1          </span></span>
<span><span class="co">##  [35] boot_1.3-30                 GenomicRanges_1.55.4       </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4             Rcpp_1.0.12                </span></span>
<span><span class="co">##  [39] SummarizedExperiment_1.33.3 knitr_1.46                 </span></span>
<span><span class="co">##  [41] future.apply_1.11.2         readr_2.1.5                </span></span>
<span><span class="co">##  [43] IRanges_2.37.1              Matrix_1.7-0               </span></span>
<span><span class="co">##  [45] tidyselect_1.2.1            abind_1.4-5                </span></span>
<span><span class="co">##  [47] yaml_2.3.8                  codetools_0.2-20           </span></span>
<span><span class="co">##  [49] curl_5.2.1                  listenv_0.9.1              </span></span>
<span><span class="co">##  [51] pkgbuild_1.4.4              lattice_0.22-6             </span></span>
<span><span class="co">##  [53] tibble_3.2.1                withr_3.0.0                </span></span>
<span><span class="co">##  [55] Biobase_2.63.1              evaluate_0.23              </span></span>
<span><span class="co">##  [57] future_1.33.2               desc_1.4.3                 </span></span>
<span><span class="co">##  [59] RcppParallel_5.1.7          pillar_1.9.0               </span></span>
<span><span class="co">##  [61] MatrixGenerics_1.15.1       stats4_4.4.0               </span></span>
<span><span class="co">##  [63] generics_0.1.3              sp_2.1-3                   </span></span>
<span><span class="co">##  [65] S4Vectors_0.41.7            hms_1.1.3                  </span></span>
<span><span class="co">##  [67] rstantools_2.4.0            munsell_0.5.1              </span></span>
<span><span class="co">##  [69] scales_1.3.0                globals_0.16.3             </span></span>
<span><span class="co">##  [71] glue_1.7.0                  tools_4.4.0                </span></span>
<span><span class="co">##  [73] fs_1.6.4                    dotCall64_1.1-1            </span></span>
<span><span class="co">##  [75] grid_4.4.0                  QuickJSR_1.1.3             </span></span>
<span><span class="co">##  [77] colorspace_2.1-0            SingleCellExperiment_1.25.1</span></span>
<span><span class="co">##  [79] GenomeInfoDbData_1.2.12     patchwork_1.2.0            </span></span>
<span><span class="co">##  [81] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">##  [83] spam_2.10-0                 fansi_1.0.6                </span></span>
<span><span class="co">##  [85] S4Arrays_1.3.7              V8_4.4.2                   </span></span>
<span><span class="co">##  [87] gtable_0.3.5                sass_0.4.9                 </span></span>
<span><span class="co">##  [89] digest_0.6.35               progressr_0.14.0           </span></span>
<span><span class="co">##  [91] BiocGenerics_0.49.1         SparseArray_1.3.5          </span></span>
<span><span class="co">##  [93] ggrepel_0.9.5               farver_2.1.1               </span></span>
<span><span class="co">##  [95] htmlwidgets_1.6.4           SeuratObject_5.0.1         </span></span>
<span><span class="co">##  [97] memoise_2.0.1               htmltools_0.5.8.1          </span></span>
<span><span class="co">##  [99] pkgdown_2.0.9               lifecycle_1.0.4            </span></span>
<span><span class="co">## [101] prettydoc_0.4.1             httr_1.4.7</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
