<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the sccomp package â€¢ sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the sccomp package">
<meta property="og:description" content="sccomp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
    <li>
      <a href="../articles/new_framework.html">New sccomp framework</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the sccomp package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2023-12-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<div class="section level2">
<h2 id="section">
<img src="inst/logo-01.png" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h2>
<p>Sccomp is a generalised method for differential composition and
variability analyses.</p>
<div class="section level3">
<h3 id="characteristics">Characteristics<a class="anchor" aria-label="anchor" href="#characteristics"></a>
</h3>
<ul>
<li>Modelling counts</li>
<li>Modelling proportionality</li>
<li>Modelling cell-type specific variability</li>
<li>Cell-type information share for variability shrinkage</li>
<li>Testing differential variability</li>
<li>Probabilistic outlier identification</li>
<li>Cross-dataset learning (hyperpriors).</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p><code>sccomp</code> can model changes in composition and variability.
By default, the formula for variability is either <code>~1</code>, which
assumes that the cell-group variability is independent of any covariate
or <code>~ factor_of_interest</code>, which assumes that the model is
dependent on the factor of interest only. The variability model must be
a subset of the model for composition.</p>
<div class="section level3">
<h3 id="binary-factor">Binary factor<a class="anchor" aria-label="anchor" href="#binary-factor"></a>
</h3>
<div class="section level4">
<h4 id="from-seurat-singlecellexperiment-metadata-objects">From Seurat, SingleCellExperiment, metadata objects<a class="anchor" aria-label="anchor" href="#from-seurat-singlecellexperiment-metadata-objects"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="from-counts">From counts<a class="anchor" aria-label="anchor" href="#from-counts"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    .count <span class="op">=</span> <span class="va">count</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000514 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 5.14 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 3.056 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                27.486 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                30.542 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 72 Ã— 14</span></span></span>
<span><span class="co">##    cell_group parameter   factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B1         (Intercept) <span style="color: #BB0000;">NA</span>      0.854     1.12   1.34     <span style="text-decoration: underline;">3</span>944.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B1         typecancer  type   -<span style="color: #BB0000;">1.04</span>     -<span style="color: #BB0000;">0.641</span> -<span style="color: #BB0000;">0.246</span>    <span style="text-decoration: underline;">2</span>282.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B2         (Intercept) <span style="color: #BB0000;">NA</span>      0.408     0.692  0.970    <span style="text-decoration: underline;">5</span>302.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B2         typecancer  type   -<span style="color: #BB0000;">1.22</span>     -<span style="color: #BB0000;">0.727</span> -<span style="color: #BB0000;">0.272</span>    <span style="text-decoration: underline;">4</span>533.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B3         (Intercept) <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.677</span>    -<span style="color: #BB0000;">0.397</span> -<span style="color: #BB0000;">0.119</span>    <span style="text-decoration: underline;">4</span>064.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B3         typecancer  type   -<span style="color: #BB0000;">0.721</span>    -<span style="color: #BB0000;">0.312</span>  0.079<span style="text-decoration: underline;">1</span>   <span style="text-decoration: underline;">3</span>692.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> BM         (Intercept) <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.33</span>     -<span style="color: #BB0000;">1.04</span>  -<span style="color: #BB0000;">0.757</span>    <span style="text-decoration: underline;">5</span>825.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> BM         typecancer  type   -<span style="color: #BB0000;">0.727</span>    -<span style="color: #BB0000;">0.316</span>  0.094<span style="text-decoration: underline;">0</span>   <span style="text-decoration: underline;">5</span>752.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 1      (Intercept) <span style="color: #BB0000;">NA</span>      0.063<span style="text-decoration: underline;">3</span>    0.294  0.502    <span style="text-decoration: underline;">3</span>897.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 1      typecancer  type   -<span style="color: #BB0000;">0.090</span><span style="color: #BB0000; text-decoration: underline;">6</span>    0.187  0.479    <span style="text-decoration: underline;">4</span>313.      1.00</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 62 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
<p>Of the output table, the estimate columns start with the prefix
<code>c_</code> indicate <code>composition</code>, or with
<code>v_</code> indicate <code>variability</code> (when
formula_variability is set).</p>
</div>
</div>
<div class="section level3">
<h3 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span> contrasts <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"typecancer - typehealthy"</span>, <span class="st">"typehealthy - typecancer"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000318 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.18 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.295 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                18.22 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                20.515 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 Ã— 18</span></span></span>
<span><span class="co">##    cell_group   parameter factor c_lower c_effect c_upper  c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature   typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.89</span>    -<span style="color: #BB0000;">1.40</span>   -<span style="color: #BB0000;">0.919</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature   typehealâ€¦ <span style="color: #BB0000;">NA</span>       0.919    1.40    1.89  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem        typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.37</span>    -<span style="color: #BB0000;">1.70</span>   -<span style="color: #BB0000;">1.05</span>  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem        typehealâ€¦ <span style="color: #BB0000;">NA</span>       1.05     1.70    2.37  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100â€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.47</span>    -<span style="color: #BB0000;">1.03</span>   -<span style="color: #BB0000;">0.596</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100â€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>       0.596    1.03    1.47  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm highâ€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>       0.853    1.77    2.83  0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm highâ€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">2.83</span>    -<span style="color: #BB0000;">1.77</span>   -<span style="color: #BB0000;">0.853</span> 0      0            <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm riboâ€¦ typecancâ€¦ <span style="color: #BB0000;">NA</span>       0.285    0.988   1.68  0.015<span style="text-decoration: underline;">8</span> 0.004<span style="text-decoration: underline;">08</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm riboâ€¦ typehealâ€¦ <span style="color: #BB0000;">NA</span>      -<span style="color: #BB0000;">1.68</span>    -<span style="color: #BB0000;">0.988</span>  -<span style="color: #BB0000;">0.285</span> 0.015<span style="text-decoration: underline;">8</span> 0.004<span style="text-decoration: underline;">08</span>      <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="categorical-factor-e-g--bayesian-anova">Categorical factor (e.g.Â Bayesian ANOVA)<a class="anchor" aria-label="anchor" href="#categorical-factor-e-g--bayesian-anova"></a>
</h3>
<p>This is achieved through model comparison with <code>loo</code>. In
the following example, the model with association with factors better
fits the data compared to the baseline model with no factor association.
For comparisons <code>check_outliers</code> must be set to FALSE as the
leave-one-out must work with the same amount of data, while outlier
elimination does not guarantee it.</p>
<p>If <code>elpd_diff</code> is away from zero of &gt; 5
<code>se_diff</code> difference of 5, we are confident that a model is
better than the other <a href="https://discourse.mc-stan.org/t/interpreting-elpd-diff-loo-package/1628/2?u=stemangiola" class="external-link">reference</a>.
In this case, -79.9 / 11.5 = -6.9, therefore we can conclude that model
one, the one with factor association, is better than model two.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/loo/" class="external-link">loo</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit first model</span></span>
<span><span class="va">model_with_factor_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    check_outliers <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit second model</span></span>
<span><span class="va">model_without_association</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    check_outliers <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> , </span>
<span>    enable_loo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare models</span></span>
<span><span class="fu"><a href="https://mc-stan.org/loo/reference/loo_compare.html" class="external-link">loo_compare</a></span><span class="op">(</span></span>
<span>  <span class="va">model_with_factor_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">model_without_association</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://mc-stan.org/loo/reference/loo.html" class="external-link">loo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-variability-binary-factor">Differential variability, binary factor<a class="anchor" aria-label="anchor" href="#differential-variability-binary-factor"></a>
</h3>
<p>We can model the cell-group variability also dependent on the type,
and so test differences in variability</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>, </span>
<span>    formula_variability <span class="op">=</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>,</span>
<span>    bimodal_mean_variability_association <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    cores <span class="op">=</span> <span class="fl">1</span> </span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000341 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.41 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 4.041 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                34.181 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                38.222 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60 Ã— 14</span></span></span>
<span><span class="co">##    cell_group        parameter factor c_lower c_effect c_upper c_n_eff c_R_k_hat</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature        (Interceâ€¦ <span style="color: #BB0000;">NA</span>      0.343     0.766  1.20     <span style="text-decoration: underline;">6</span>220.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B immature        typehealâ€¦ type    0.884     1.44   2.02     <span style="text-decoration: underline;">4</span>332.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B mem             (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.50</span>     -<span style="color: #BB0000;">0.867</span> -<span style="color: #BB0000;">0.162</span>    <span style="text-decoration: underline;">4</span>714.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B mem             typehealâ€¦ type    1.05      1.86   2.65     <span style="text-decoration: underline;">4</span>341.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm S100A4     (Interceâ€¦ <span style="color: #BB0000;">NA</span>      1.31      1.66   1.98     <span style="text-decoration: underline;">5</span>954.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 cm S100A4     typehealâ€¦ type    0.490     0.942  1.40     <span style="text-decoration: underline;">5</span>186.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 cm high cytoâ€¦ (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">1.05</span>     -<span style="color: #BB0000;">0.538</span>  0.031<span style="text-decoration: underline;">5</span>   <span style="text-decoration: underline;">5</span>062.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 cm high cytoâ€¦ typehealâ€¦ type   -<span style="color: #BB0000;">3.00</span>     -<span style="color: #BB0000;">1.25</span>   1.15     <span style="text-decoration: underline;">4</span>029.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD4 cm ribosome   (Interceâ€¦ <span style="color: #BB0000;">NA</span>     -<span style="color: #BB0000;">0.047</span><span style="color: #BB0000; text-decoration: underline;">5</span>    0.306  0.682    <span style="text-decoration: underline;">4</span>060.      1.00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD4 cm ribosome   typehealâ€¦ type   -<span style="color: #BB0000;">1.77</span>     -<span style="color: #BB0000;">0.954</span>  0.055<span style="text-decoration: underline;">3</span>   <span style="text-decoration: underline;">3</span>441.      1.00</span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 50 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># â„¹ 6 more variables: v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;, v_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;, count_data &lt;list&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="suggested-settings">Suggested settings<a class="anchor" aria-label="anchor" href="#suggested-settings"></a>
</h2>
<div class="section level3">
<h3 id="for-single-cell-rna-sequencing">For single-cell RNA sequencing<a class="anchor" aria-label="anchor" href="#for-single-cell-rna-sequencing"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = TRUE</code>. The
bimodality of the mean-variability association can be confirmed from the
plots$credible_intervals_2D (see below).</p>
</div>
<div class="section level3">
<h3 id="for-cytof-and-microbiome-data">For CyTOF and microbiome data<a class="anchor" aria-label="anchor" href="#for-cytof-and-microbiome-data"></a>
</h3>
<p>We recommend setting
<code>bimodal_mean_variability_association  = FALSE</code>
(Default).</p>
</div>
</div>
<div class="section level2">
<h2 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a>
</h2>
<div class="section level3">
<h3 id="summary-plots">Summary plots<a class="anchor" aria-label="anchor" href="#summary-plots"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Warning: `plot_summary()` was deprecated in sccomp 1.7.1.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">â„¹</span> sccomp says: plot_summary() is soft-deprecated. Please use sccomp_test().</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">##              cell_group   parameter factor     c_lower     c_effect     c_upper</span></span>
<span><span class="co">## 1            B immature (Intercept)   &lt;NA&gt;  0.34326262  0.766464042  1.19539556</span></span>
<span><span class="co">## 2            B immature typehealthy   type  0.88383339  1.443109913  2.01527658</span></span>
<span><span class="co">## 3                 B mem (Intercept)   &lt;NA&gt; -1.49625916 -0.867265848 -0.16156936</span></span>
<span><span class="co">## 4                 B mem typehealthy   type  1.05301165  1.856365920  2.64635530</span></span>
<span><span class="co">## 5         CD4 cm S100A4 (Intercept)   &lt;NA&gt;  1.30961099  1.659969916  1.97885848</span></span>
<span><span class="co">## 6         CD4 cm S100A4 typehealthy   type  0.49008189  0.941824585  1.40162675</span></span>
<span><span class="co">## 7  CD4 cm high cytokine (Intercept)   &lt;NA&gt; -1.05277304 -0.537842451  0.03148057</span></span>
<span><span class="co">## 8  CD4 cm high cytokine typehealthy   type -2.99765852 -1.251223937  1.15173298</span></span>
<span><span class="co">## 9       CD4 cm ribosome (Intercept)   &lt;NA&gt; -0.04752921  0.305897482  0.68170916</span></span>
<span><span class="co">## 10      CD4 cm ribosome typehealthy   type -1.77260446 -0.954001809  0.05527450</span></span>
<span><span class="co">## 11 CD4 em high cytokine (Intercept)   &lt;NA&gt; -1.16915623 -0.717399712 -0.20381060</span></span>
<span><span class="co">## 12 CD4 em high cytokine typehealthy   type -2.29657362 -1.267126439  0.17663722</span></span>
<span><span class="co">## 13            CD4 naive (Intercept)   &lt;NA&gt;  0.76602013  1.216767358  1.65026195</span></span>
<span><span class="co">## 14            CD4 naive typehealthy   type  0.46416513  1.134278709  1.82818479</span></span>
<span><span class="co">## 15         CD4 ribosome (Intercept)   &lt;NA&gt; -0.32123598  0.043460966  0.43737491</span></span>
<span><span class="co">## 16         CD4 ribosome typehealthy   type  1.47778901  2.061694519  2.63789867</span></span>
<span><span class="co">## 17             CD8 em 1 (Intercept)   &lt;NA&gt;  0.27797702  0.675783232  1.07498584</span></span>
<span><span class="co">## 18             CD8 em 1 typehealthy   type -0.02815355  0.794877803  1.58247095</span></span>
<span><span class="co">## 19             CD8 em 2 (Intercept)   &lt;NA&gt; -0.89998755  0.128294072  1.23661304</span></span>
<span><span class="co">## 20             CD8 em 2 typehealthy   type -3.56392919 -1.205064147  1.24496777</span></span>
<span><span class="co">## 21             CD8 em 3 (Intercept)   &lt;NA&gt; -1.41894909 -0.643472971  0.21901258</span></span>
<span><span class="co">## 22             CD8 em 3 typehealthy   type -1.76497616 -0.822691114  0.06320423</span></span>
<span><span class="co">## 23            CD8 naive (Intercept)   &lt;NA&gt;  0.60277389  1.017493586  1.42975549</span></span>
<span><span class="co">## 24            CD8 naive typehealthy   type -0.70820478  0.010860435  0.73822923</span></span>
<span><span class="co">## 25     CD8 transitional (Intercept)   &lt;NA&gt;  0.90589231  1.214952369  1.51827011</span></span>
<span><span class="co">## 26     CD8 transitional typehealthy   type  0.15791635  0.678810066  1.19877234</span></span>
<span><span class="co">## 27                 MAIT (Intercept)   &lt;NA&gt; -1.05026455 -0.598896254 -0.12461070</span></span>
<span><span class="co">## 28                 MAIT typehealthy   type  0.44482258  1.223999579  2.00334613</span></span>
<span><span class="co">## 29               Mac M1 (Intercept)   &lt;NA&gt;  0.60467438  0.949699075  1.28556804</span></span>
<span><span class="co">## 30               Mac M1 typehealthy   type  0.02066449  0.598931465  1.17875036</span></span>
<span><span class="co">## 31        Mac M1 FCER1A (Intercept)   &lt;NA&gt; -1.34486684 -0.986036407 -0.55436162</span></span>
<span><span class="co">## 32        Mac M1 FCER1A typehealthy   type -1.81372815 -1.067093455 -0.21103252</span></span>
<span><span class="co">## 33                 Mono (Intercept)   &lt;NA&gt; -1.81858681 -1.350404752 -0.79124262</span></span>
<span><span class="co">## 34                 Mono typehealthy   type -0.83783930 -0.107635994  0.65896538</span></span>
<span><span class="co">## 35          Mono NKG7 1 (Intercept)   &lt;NA&gt; -1.45439376 -0.871294748 -0.17999055</span></span>
<span><span class="co">## 36          Mono NKG7 1 typehealthy   type -3.97376935 -2.907860852 -0.99630437</span></span>
<span><span class="co">## 37          Mono NKG7 2 (Intercept)   &lt;NA&gt; -0.40620260 -0.005646587  0.41224426</span></span>
<span><span class="co">## 38          Mono NKG7 2 typehealthy   type -2.46542232 -1.269049419  0.48881037</span></span>
<span><span class="co">## 39  Mono classic inflam (Intercept)   &lt;NA&gt;  1.48410453  1.784088564  2.04272705</span></span>
<span><span class="co">## 40  Mono classic inflam typehealthy   type -1.27175415 -0.461272413  0.33578812</span></span>
<span><span class="co">## 41     Mono non-classic (Intercept)   &lt;NA&gt;  0.61148113  0.959930878  1.31357242</span></span>
<span><span class="co">## 42     Mono non-classic typehealthy   type -0.64418171  0.008394654  0.68180779</span></span>
<span><span class="co">## 43    Myeloid migratory (Intercept)   &lt;NA&gt; -0.36058494  0.016873642  0.42393211</span></span>
<span><span class="co">## 44    Myeloid migratory typehealthy   type -1.32473720 -0.549045770  0.33420981</span></span>
<span><span class="co">## 45                   NK (Intercept)   &lt;NA&gt;  1.13437415  1.432013859  1.71824646</span></span>
<span><span class="co">## 46                   NK typehealthy   type -0.04528934  0.539377358  1.10910531</span></span>
<span><span class="co">## 47           NK cycling (Intercept)   &lt;NA&gt; -1.66258924 -1.221269093 -0.72743058</span></span>
<span><span class="co">## 48           NK cycling typehealthy   type -0.77289565 -0.188197685  0.42692169</span></span>
<span><span class="co">## 49     NK high cytokine (Intercept)   &lt;NA&gt; -2.22910600 -1.633069024 -0.96205044</span></span>
<span><span class="co">## 50     NK high cytokine typehealthy   type -1.23954275  0.050287039  1.70842787</span></span>
<span><span class="co">## 51                 Stem (Intercept)   &lt;NA&gt; -1.06861578 -0.575676868 -0.03569317</span></span>
<span><span class="co">## 52                 Stem typehealthy   type -3.16972097 -2.234455343 -0.75526951</span></span>
<span><span class="co">## 53                T gd1 (Intercept)   &lt;NA&gt; -1.06369488 -0.686254132 -0.27835173</span></span>
<span><span class="co">## 54                T gd1 typehealthy   type  0.06333396  0.928377053  1.87684144</span></span>
<span><span class="co">## 55                T gd2 (Intercept)   &lt;NA&gt; -1.24303746 -0.676879729 -0.03618023</span></span>
<span><span class="co">## 56                T gd2 typehealthy   type -0.43179832  0.533413416  1.55302185</span></span>
<span><span class="co">## 57                 cDC2 (Intercept)   &lt;NA&gt; -0.77560074 -0.290368412  0.23251013</span></span>
<span><span class="co">## 58                 cDC2 typehealthy   type -0.39754284  0.316949775  1.03006931</span></span>
<span><span class="co">## 59                  pDC (Intercept)   &lt;NA&gt; -1.13136073 -0.679354301 -0.13289483</span></span>
<span><span class="co">## 60                  pDC typehealthy   type -0.10743343  0.673496174  1.52042292</span></span>
<span><span class="co">##      c_pH0        c_FDR  c_n_eff c_R_k_hat     v_lower     v_effect    v_upper</span></span>
<span><span class="co">## 1  0.00425 4.038462e-04 6220.274 1.0000567 -4.47237075 -3.824352259 -3.0207516</span></span>
<span><span class="co">## 2  0.00000 0.000000e+00 4332.127 1.0009625 -1.67868909 -0.502049505  0.7393221</span></span>
<span><span class="co">## 3  0.03125 6.069444e-03 4714.315 0.9997993 -5.28525007 -4.552157844 -3.5411521</span></span>
<span><span class="co">## 4  0.00000 0.000000e+00 4341.276 0.9997797 -2.06078106 -0.963360273  0.2402856</span></span>
<span><span class="co">## 5  0.00000 0.000000e+00 5953.640 1.0003693 -3.81303199 -3.212726353 -2.5002773</span></span>
<span><span class="co">## 6  0.00125 3.125000e-04 5185.990 1.0017941 -2.07437120 -0.917829168  0.2591152</span></span>
<span><span class="co">## 7  0.11875 2.071739e-02 5061.810 0.9997510 -5.44810926 -4.716511179 -3.7449439</span></span>
<span><span class="co">## 8  0.18125 5.028750e-02 4029.244 0.9997598 -0.44769588  1.534202172  3.4746158</span></span>
<span><span class="co">## 9  0.27350 3.530000e-02 4059.581 0.9998429 -5.34916087 -4.609796896 -3.7646895</span></span>
<span><span class="co">## 10 0.06200 1.662500e-02 3441.277 0.9997915 -0.53590282  0.658860646  1.8849363</span></span>
<span><span class="co">## 11 0.02400 3.125000e-03 4607.828 0.9997555 -5.94522327 -5.163299329 -4.2519093</span></span>
<span><span class="co">## 12 0.06475 2.032692e-02 4275.527 0.9998338 -1.05236604  0.366851489  1.8492453</span></span>
<span><span class="co">## 13 0.00000 0.000000e+00 6071.679 0.9997540 -3.66253473 -3.058437865 -2.3011414</span></span>
<span><span class="co">## 14 0.00525 1.916667e-03 6143.807 0.9998453 -0.85519079  0.214062311  1.2616965</span></span>
<span><span class="co">## 15 0.79900 9.189286e-02 5463.801 1.0001771 -5.70470299 -4.891802690 -4.0023215</span></span>
<span><span class="co">## 16 0.00000 0.000000e+00 5290.990 0.9997525 -0.62593856  0.551654045  1.7729957</span></span>
<span><span class="co">## 17 0.00975 1.071429e-03 4077.818 1.0000116 -4.75986422 -4.054367972 -3.2433195</span></span>
<span><span class="co">## 18 0.07600 2.430357e-02 5845.838 0.9997503  0.03494679  1.080948068  2.1546025</span></span>
<span><span class="co">## 19 0.54775 6.570370e-02 4124.336 1.0010147 -3.68379962 -1.990610961 -0.6008783</span></span>
<span><span class="co">## 20 0.22325 6.528409e-02 4549.068 1.0003754 -1.59814988  0.506022553  2.6546100</span></span>
<span><span class="co">## 21 0.13250 2.537500e-02 5764.159 0.9999679 -4.73262516 -3.937001997 -2.5883987</span></span>
<span><span class="co">## 22 0.08025 2.803333e-02 4679.710 1.0005038 -4.09560122 -2.718240461 -1.3631696</span></span>
<span><span class="co">## 23 0.00025 2.500000e-05 4484.366 1.0000537 -4.16749641 -3.516552059 -2.7424955</span></span>
<span><span class="co">## 24 0.69250 1.622586e-01 4447.936 0.9998039 -0.99899018  0.024581340  1.1029183</span></span>
<span><span class="co">## 25 0.00000 0.000000e+00 4900.734 0.9999956 -4.83388876 -4.087305805 -3.2232430</span></span>
<span><span class="co">## 26 0.03425 8.575000e-03 4944.997 1.0002617 -1.07841296  0.084020792  1.3254906</span></span>
<span><span class="co">## 27 0.05575 1.012500e-02 4695.802 0.9997528 -5.84288261 -5.073883716 -4.1015296</span></span>
<span><span class="co">## 28 0.00500 1.250000e-03 6181.109 0.9998369 -0.73083224  0.381455569  1.5516070</span></span>
<span><span class="co">## 29 0.00000 0.000000e+00 4794.841 0.9998616 -4.97071341 -4.218965822 -3.3897791</span></span>
<span><span class="co">## 30 0.08400 3.153125e-02 5852.191 0.9999626 -0.78985839  0.307101068  1.5015152</span></span>
<span><span class="co">## 31 0.00050 8.333333e-05 4420.827 0.9997505 -6.96718324 -5.995970549 -4.8867844</span></span>
<span><span class="co">## 32 0.02375 5.722222e-03 4398.675 0.9997629 -1.46798941 -0.023714439  1.4230765</span></span>
<span><span class="co">## 33 0.00000 0.000000e+00 5036.154 0.9997597 -6.58556042 -5.746475667 -4.7653395</span></span>
<span><span class="co">## 34 0.60050 1.433214e-01 4726.974 0.9998012 -2.16881309 -0.767202452  0.5776295</span></span>
<span><span class="co">## 35 0.02800 4.588235e-03 6131.530 0.9998128 -5.23848922 -4.498184720 -3.5658850</span></span>
<span><span class="co">## 36 0.00875 3.468750e-03 2163.545 0.9998492 -5.29633121 -2.364370532  0.7140195</span></span>
<span><span class="co">## 37 0.81775 1.169224e-01 3997.214 0.9997986 -5.39818091 -4.676946527 -3.8455683</span></span>
<span><span class="co">## 38 0.08850 3.488235e-02 3636.555 0.9999503 -0.18847451  1.221428727  2.8182485</span></span>
<span><span class="co">## 39 0.00000 0.000000e+00 3948.267 0.9998234 -4.33433591 -3.597237869 -2.7510414</span></span>
<span><span class="co">## 40 0.25975 8.087500e-02 5934.663 0.9997950  0.32546689  1.446996448  2.5359410</span></span>
<span><span class="co">## 41 0.00000 0.000000e+00 4848.210 1.0002095 -4.72084080 -4.006002100 -3.1893655</span></span>
<span><span class="co">## 42 0.71850 1.808000e-01 5195.839 0.9999028 -0.84551327  0.256530203  1.4361228</span></span>
<span><span class="co">## 43 0.82525 1.405333e-01 4541.975 0.9998352 -5.58370440 -4.821603586 -3.9528164</span></span>
<span><span class="co">## 44 0.20725 5.776190e-02 5070.041 0.9997580 -0.80084691  0.363930049  1.6019834</span></span>
<span><span class="co">## 45 0.00000 0.000000e+00 4471.975 1.0002373 -4.53053116 -3.818570530 -3.0310863</span></span>
<span><span class="co">## 46 0.11650 4.339474e-02 6161.884 1.0002674 -0.53002267  0.623164715  1.7919097</span></span>
<span><span class="co">## 47 0.00025 4.545455e-05 5679.186 0.9998315 -6.61539417 -5.791058020 -4.8351112</span></span>
<span><span class="co">## 48 0.51625 1.086346e-01 4728.248 0.9998761 -3.27999119 -1.596454357 -0.1158578</span></span>
<span><span class="co">## 49 0.00000 0.000000e+00 4640.230 0.9999547 -6.41840677 -5.557094623 -4.5064191</span></span>
<span><span class="co">## 50 0.58800 1.263889e-01 3152.123 1.0011501 -0.70677066  0.681787175  2.3029439</span></span>
<span><span class="co">## 51 0.08700 1.626136e-02 5493.202 0.9998262 -5.48066634 -4.762275273 -3.8736963</span></span>
<span><span class="co">## 52 0.00750 2.714286e-03 2655.027 0.9997546 -2.85291034 -0.844922308  0.9554046</span></span>
<span><span class="co">## 53 0.01100 1.733333e-03 4692.063 0.9998629 -6.49971938 -5.581845190 -4.6251623</span></span>
<span><span class="co">## 54 0.05175 1.250000e-02 5162.244 1.0000080  0.14613606  1.345473598  2.6501132</span></span>
<span><span class="co">## 55 0.06825 1.289286e-02 5213.015 0.9997910 -5.30755563 -4.557489802 -3.6217868</span></span>
<span><span class="co">## 56 0.24500 7.309783e-02 4536.369 0.9997606 -1.22267568 -0.105757513  1.0889201</span></span>
<span><span class="co">## 57 0.34375 4.716346e-02 5083.723 1.0000189 -5.40622759 -4.629345298 -3.6948889</span></span>
<span><span class="co">## 58 0.36725 9.233000e-02 4676.448 0.9997855 -1.77423497 -0.632074061  0.5585467</span></span>
<span><span class="co">## 59 0.03750 7.723684e-03 4865.020 0.9997543 -5.79788651 -5.047521830 -4.1262999</span></span>
<span><span class="co">## 60 0.11500 3.933333e-02 4877.551 0.9997517 -1.09883647 -0.004138393  1.1670847</span></span>
<span><span class="co">##      v_pH0        v_FDR  v_n_eff v_R_k_hat   count_data</span></span>
<span><span class="co">## 1  0.00000 0.0000000000 4648.568 0.9999651 c("10x_6....</span></span>
<span><span class="co">## 2  0.30750 0.1440694444 4983.104 0.9998813 c("10x_6....</span></span>
<span><span class="co">## 3  0.00000 0.0000000000 3588.356 0.9998841 c("10x_6....</span></span>
<span><span class="co">## 4  0.09875 0.0530833333 4507.871 0.9997503 c("10x_6....</span></span>
<span><span class="co">## 5  0.00000 0.0000000000 5773.108 1.0000837 c("10x_6....</span></span>
<span><span class="co">## 6  0.11675 0.0594500000 5255.133 0.9998035 c("10x_6....</span></span>
<span><span class="co">## 7  0.00000 0.0000000000 8330.805 0.9997608 c("10x_6....</span></span>
<span><span class="co">## 8  0.09150 0.0473750000 4558.335 0.9997525 c("10x_6....</span></span>
<span><span class="co">## 9  0.00000 0.0000000000 7273.853 0.9997801 c("10x_6....</span></span>
<span><span class="co">## 10 0.23000 0.0970384615 4457.578 0.9997640 c("10x_6....</span></span>
<span><span class="co">## 11 0.00000 0.0000000000 6844.765 0.9997584 c("10x_6....</span></span>
<span><span class="co">## 12 0.41025 0.1897386364 4186.488 0.9999649 c("10x_6....</span></span>
<span><span class="co">## 13 0.00000 0.0000000000 5044.457 0.9997574 c("10x_6....</span></span>
<span><span class="co">## 14 0.48450 0.2220700000 5008.526 0.9997728 c("10x_6....</span></span>
<span><span class="co">## 15 0.00000 0.0000000000 4821.616 0.9998303 c("10x_6....</span></span>
<span><span class="co">## 16 0.28000 0.1344558824 4963.327 1.0009502 c("10x_6....</span></span>
<span><span class="co">## 17 0.00000 0.0000000000 6571.326 0.9997509 c("10x_6....</span></span>
<span><span class="co">## 18 0.05350 0.0256000000 5022.178 0.9997558 c("10x_6....</span></span>
<span><span class="co">## 19 0.00350 0.0001166667 3326.859 1.0010595 c("10x_6....</span></span>
<span><span class="co">## 20 0.39550 0.1683250000 3852.842 0.9997892 c("10x_6....</span></span>
<span><span class="co">## 21 0.00000 0.0000000000 3258.397 1.0000369 c("10x_6....</span></span>
<span><span class="co">## 22 0.00000 0.0000000000 3924.869 0.9997989 c("10x_6....</span></span>
<span><span class="co">## 23 0.00000 0.0000000000 6059.154 0.9998790 c("10x_6....</span></span>
<span><span class="co">## 24 0.63600 0.2848500000 5303.628 0.9999280 c("10x_6....</span></span>
<span><span class="co">## 25 0.00000 0.0000000000 5922.819 1.0001866 c("10x_6....</span></span>
<span><span class="co">## 26 0.57375 0.2476944444 5102.646 0.9997594 c("10x_6....</span></span>
<span><span class="co">## 27 0.00000 0.0000000000 6514.562 0.9997677 c("10x_6....</span></span>
<span><span class="co">## 28 0.37775 0.1563684211 5186.102 0.9997747 c("10x_6....</span></span>
<span><span class="co">## 29 0.00000 0.0000000000 6381.840 0.9997632 c("10x_6....</span></span>
<span><span class="co">## 30 0.43000 0.2001847826 5657.069 0.9997626 c("10x_6....</span></span>
<span><span class="co">## 31 0.00000 0.0000000000 5146.706 0.9997500 c("10x_6....</span></span>
<span><span class="co">## 32 0.59325 0.2600357143 4149.512 0.9997535 c("10x_6....</span></span>
<span><span class="co">## 33 0.00000 0.0000000000 5654.691 0.9998075 c("10x_6....</span></span>
<span><span class="co">## 34 0.20825 0.0729772727 4847.805 0.9999185 c("10x_6....</span></span>
<span><span class="co">## 35 0.00000 0.0000000000 6011.824 0.9998174 c("10x_6....</span></span>
<span><span class="co">## 36 0.07825 0.0343750000 2792.663 0.9998129 c("10x_6....</span></span>
<span><span class="co">## 37 0.00000 0.0000000000 6729.472 0.9999330 c("10x_6....</span></span>
<span><span class="co">## 38 0.08125 0.0410714286 3585.441 1.0000874 c("10x_6....</span></span>
<span><span class="co">## 39 0.00000 0.0000000000 5963.930 0.9997508 c("10x_6....</span></span>
<span><span class="co">## 40 0.01375 0.0068750000 4667.244 0.9997609 c("10x_6....</span></span>
<span><span class="co">## 41 0.00000 0.0000000000 5398.340 0.9997738 c("10x_6....</span></span>
<span><span class="co">## 42 0.46300 0.2111354167 5443.585 0.9999238 c("10x_6....</span></span>
<span><span class="co">## 43 0.00000 0.0000000000 6012.038 1.0001387 c("10x_6....</span></span>
<span><span class="co">## 44 0.39750 0.1792380952 4432.890 1.0003462 c("10x_6....</span></span>
<span><span class="co">## 45 0.00000 0.0000000000 6870.258 1.0000855 c("10x_6....</span></span>
<span><span class="co">## 46 0.22875 0.0859583333 5766.389 1.0007165 c("10x_6....</span></span>
<span><span class="co">## 47 0.00000 0.0000000000 4602.109 0.9998410 c("10x_6....</span></span>
<span><span class="co">## 48 0.03175 0.0186250000 5650.538 1.0002573 c("10x_6....</span></span>
<span><span class="co">## 49 0.00000 0.0000000000 6826.189 0.9997678 c("10x_6....</span></span>
<span><span class="co">## 50 0.25500 0.1253593750 3496.478 1.0009728 c("10x_6....</span></span>
<span><span class="co">## 51 0.00000 0.0000000000 6932.251 1.0009846 c("10x_6....</span></span>
<span><span class="co">## 52 0.24975 0.1167166667 3342.113 0.9997574 c("10x_6....</span></span>
<span><span class="co">## 53 0.00000 0.0000000000 6083.848 0.9997505 c("10x_6....</span></span>
<span><span class="co">## 54 0.02900 0.0142500000 5115.012 0.9997707 c("10x_6....</span></span>
<span><span class="co">## 55 0.00000 0.0000000000 5381.725 0.9997732 c("10x_6....</span></span>
<span><span class="co">## 56 0.56225 0.2351538462 3914.919 0.9997939 c("10x_6....</span></span>
<span><span class="co">## 57 0.00000 0.0000000000 5368.806 1.0001335 c("10x_6....</span></span>
<span><span class="co">## 58 0.23950 0.1072142857 4835.618 1.0001678 c("10x_6....</span></span>
<span><span class="co">## 59 0.00000 0.0000000000 5053.551 0.9997529 c("10x_6....</span></span>
<span><span class="co">## 60 0.62850 0.2727413793 5158.118 0.9997621 c("10x_6....</span></span></code></pre>
<p>A plot of group proportion, faceted by groups. The blue boxplots
represent the posterior predictive check. If the model is likely to be
descriptively adequate to the data, the blue box plot should roughly
overlay with the black box plot, which represents the observed data. The
outliers are coloured in red. A box plot will be returned for every
(discrete) covariate present in <code>formula_composition</code>. The
colour coding represents the significant associations for composition
and/or variability.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">boxplot</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `boxplot`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<p>A plot of estimates of differential composition (c_) on the x-axis
and differential variability (v_) on the y-axis. The error bars
represent 95% credible intervals. The dashed lines represent the minimal
effect that the hypothesis test is based on. An effect is labelled as
significant if bigger than the minimal effect according to the 95%
credible interval. Facets represent the covariates in the model.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_1D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualisation-of-the-mcmc-chains-from-the-posterior-distribution">Visualisation of the MCMC chains from the posterior
distribution<a class="anchor" aria-label="anchor" href="#visualisation-of-the-mcmc-chains-from-the-posterior-distribution"></a>
</h3>
<p>It is possible to directly evaluate the posterior distribution. In
this example, we plot the Monte Carlo chain for the slope parameter of
the first cell type. We can see that it has converged and is negative
with probability 1.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="st">"fit"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">rstan</span><span class="fu">::</span><span class="fu"><a href="https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html" class="external-link">traceplot</a></span><span class="op">(</span><span class="st">"beta[2,1]"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Plot 1D significance plot</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">=</span> <span class="va">res</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/plot_summary.html">plot_summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              cell_group   parameter factor     c_lower     c_effect     c_upper</span></span>
<span><span class="co">## 1            B immature (Intercept)   &lt;NA&gt;  0.34326262  0.766464042  1.19539556</span></span>
<span><span class="co">## 2            B immature typehealthy   type  0.88383339  1.443109913  2.01527658</span></span>
<span><span class="co">## 3                 B mem (Intercept)   &lt;NA&gt; -1.49625916 -0.867265848 -0.16156936</span></span>
<span><span class="co">## 4                 B mem typehealthy   type  1.05301165  1.856365920  2.64635530</span></span>
<span><span class="co">## 5         CD4 cm S100A4 (Intercept)   &lt;NA&gt;  1.30961099  1.659969916  1.97885848</span></span>
<span><span class="co">## 6         CD4 cm S100A4 typehealthy   type  0.49008189  0.941824585  1.40162675</span></span>
<span><span class="co">## 7  CD4 cm high cytokine (Intercept)   &lt;NA&gt; -1.05277304 -0.537842451  0.03148057</span></span>
<span><span class="co">## 8  CD4 cm high cytokine typehealthy   type -2.99765852 -1.251223937  1.15173298</span></span>
<span><span class="co">## 9       CD4 cm ribosome (Intercept)   &lt;NA&gt; -0.04752921  0.305897482  0.68170916</span></span>
<span><span class="co">## 10      CD4 cm ribosome typehealthy   type -1.77260446 -0.954001809  0.05527450</span></span>
<span><span class="co">## 11 CD4 em high cytokine (Intercept)   &lt;NA&gt; -1.16915623 -0.717399712 -0.20381060</span></span>
<span><span class="co">## 12 CD4 em high cytokine typehealthy   type -2.29657362 -1.267126439  0.17663722</span></span>
<span><span class="co">## 13            CD4 naive (Intercept)   &lt;NA&gt;  0.76602013  1.216767358  1.65026195</span></span>
<span><span class="co">## 14            CD4 naive typehealthy   type  0.46416513  1.134278709  1.82818479</span></span>
<span><span class="co">## 15         CD4 ribosome (Intercept)   &lt;NA&gt; -0.32123598  0.043460966  0.43737491</span></span>
<span><span class="co">## 16         CD4 ribosome typehealthy   type  1.47778901  2.061694519  2.63789867</span></span>
<span><span class="co">## 17             CD8 em 1 (Intercept)   &lt;NA&gt;  0.27797702  0.675783232  1.07498584</span></span>
<span><span class="co">## 18             CD8 em 1 typehealthy   type -0.02815355  0.794877803  1.58247095</span></span>
<span><span class="co">## 19             CD8 em 2 (Intercept)   &lt;NA&gt; -0.89998755  0.128294072  1.23661304</span></span>
<span><span class="co">## 20             CD8 em 2 typehealthy   type -3.56392919 -1.205064147  1.24496777</span></span>
<span><span class="co">## 21             CD8 em 3 (Intercept)   &lt;NA&gt; -1.41894909 -0.643472971  0.21901258</span></span>
<span><span class="co">## 22             CD8 em 3 typehealthy   type -1.76497616 -0.822691114  0.06320423</span></span>
<span><span class="co">## 23            CD8 naive (Intercept)   &lt;NA&gt;  0.60277389  1.017493586  1.42975549</span></span>
<span><span class="co">## 24            CD8 naive typehealthy   type -0.70820478  0.010860435  0.73822923</span></span>
<span><span class="co">## 25     CD8 transitional (Intercept)   &lt;NA&gt;  0.90589231  1.214952369  1.51827011</span></span>
<span><span class="co">## 26     CD8 transitional typehealthy   type  0.15791635  0.678810066  1.19877234</span></span>
<span><span class="co">## 27                 MAIT (Intercept)   &lt;NA&gt; -1.05026455 -0.598896254 -0.12461070</span></span>
<span><span class="co">## 28                 MAIT typehealthy   type  0.44482258  1.223999579  2.00334613</span></span>
<span><span class="co">## 29               Mac M1 (Intercept)   &lt;NA&gt;  0.60467438  0.949699075  1.28556804</span></span>
<span><span class="co">## 30               Mac M1 typehealthy   type  0.02066449  0.598931465  1.17875036</span></span>
<span><span class="co">## 31        Mac M1 FCER1A (Intercept)   &lt;NA&gt; -1.34486684 -0.986036407 -0.55436162</span></span>
<span><span class="co">## 32        Mac M1 FCER1A typehealthy   type -1.81372815 -1.067093455 -0.21103252</span></span>
<span><span class="co">## 33                 Mono (Intercept)   &lt;NA&gt; -1.81858681 -1.350404752 -0.79124262</span></span>
<span><span class="co">## 34                 Mono typehealthy   type -0.83783930 -0.107635994  0.65896538</span></span>
<span><span class="co">## 35          Mono NKG7 1 (Intercept)   &lt;NA&gt; -1.45439376 -0.871294748 -0.17999055</span></span>
<span><span class="co">## 36          Mono NKG7 1 typehealthy   type -3.97376935 -2.907860852 -0.99630437</span></span>
<span><span class="co">## 37          Mono NKG7 2 (Intercept)   &lt;NA&gt; -0.40620260 -0.005646587  0.41224426</span></span>
<span><span class="co">## 38          Mono NKG7 2 typehealthy   type -2.46542232 -1.269049419  0.48881037</span></span>
<span><span class="co">## 39  Mono classic inflam (Intercept)   &lt;NA&gt;  1.48410453  1.784088564  2.04272705</span></span>
<span><span class="co">## 40  Mono classic inflam typehealthy   type -1.27175415 -0.461272413  0.33578812</span></span>
<span><span class="co">## 41     Mono non-classic (Intercept)   &lt;NA&gt;  0.61148113  0.959930878  1.31357242</span></span>
<span><span class="co">## 42     Mono non-classic typehealthy   type -0.64418171  0.008394654  0.68180779</span></span>
<span><span class="co">## 43    Myeloid migratory (Intercept)   &lt;NA&gt; -0.36058494  0.016873642  0.42393211</span></span>
<span><span class="co">## 44    Myeloid migratory typehealthy   type -1.32473720 -0.549045770  0.33420981</span></span>
<span><span class="co">## 45                   NK (Intercept)   &lt;NA&gt;  1.13437415  1.432013859  1.71824646</span></span>
<span><span class="co">## 46                   NK typehealthy   type -0.04528934  0.539377358  1.10910531</span></span>
<span><span class="co">## 47           NK cycling (Intercept)   &lt;NA&gt; -1.66258924 -1.221269093 -0.72743058</span></span>
<span><span class="co">## 48           NK cycling typehealthy   type -0.77289565 -0.188197685  0.42692169</span></span>
<span><span class="co">## 49     NK high cytokine (Intercept)   &lt;NA&gt; -2.22910600 -1.633069024 -0.96205044</span></span>
<span><span class="co">## 50     NK high cytokine typehealthy   type -1.23954275  0.050287039  1.70842787</span></span>
<span><span class="co">## 51                 Stem (Intercept)   &lt;NA&gt; -1.06861578 -0.575676868 -0.03569317</span></span>
<span><span class="co">## 52                 Stem typehealthy   type -3.16972097 -2.234455343 -0.75526951</span></span>
<span><span class="co">## 53                T gd1 (Intercept)   &lt;NA&gt; -1.06369488 -0.686254132 -0.27835173</span></span>
<span><span class="co">## 54                T gd1 typehealthy   type  0.06333396  0.928377053  1.87684144</span></span>
<span><span class="co">## 55                T gd2 (Intercept)   &lt;NA&gt; -1.24303746 -0.676879729 -0.03618023</span></span>
<span><span class="co">## 56                T gd2 typehealthy   type -0.43179832  0.533413416  1.55302185</span></span>
<span><span class="co">## 57                 cDC2 (Intercept)   &lt;NA&gt; -0.77560074 -0.290368412  0.23251013</span></span>
<span><span class="co">## 58                 cDC2 typehealthy   type -0.39754284  0.316949775  1.03006931</span></span>
<span><span class="co">## 59                  pDC (Intercept)   &lt;NA&gt; -1.13136073 -0.679354301 -0.13289483</span></span>
<span><span class="co">## 60                  pDC typehealthy   type -0.10743343  0.673496174  1.52042292</span></span>
<span><span class="co">##      c_pH0        c_FDR  c_n_eff c_R_k_hat     v_lower     v_effect    v_upper</span></span>
<span><span class="co">## 1  0.00425 4.038462e-04 6220.274 1.0000567 -4.47237075 -3.824352259 -3.0207516</span></span>
<span><span class="co">## 2  0.00000 0.000000e+00 4332.127 1.0009625 -1.67868909 -0.502049505  0.7393221</span></span>
<span><span class="co">## 3  0.03125 6.069444e-03 4714.315 0.9997993 -5.28525007 -4.552157844 -3.5411521</span></span>
<span><span class="co">## 4  0.00000 0.000000e+00 4341.276 0.9997797 -2.06078106 -0.963360273  0.2402856</span></span>
<span><span class="co">## 5  0.00000 0.000000e+00 5953.640 1.0003693 -3.81303199 -3.212726353 -2.5002773</span></span>
<span><span class="co">## 6  0.00125 3.125000e-04 5185.990 1.0017941 -2.07437120 -0.917829168  0.2591152</span></span>
<span><span class="co">## 7  0.11875 2.071739e-02 5061.810 0.9997510 -5.44810926 -4.716511179 -3.7449439</span></span>
<span><span class="co">## 8  0.18125 5.028750e-02 4029.244 0.9997598 -0.44769588  1.534202172  3.4746158</span></span>
<span><span class="co">## 9  0.27350 3.530000e-02 4059.581 0.9998429 -5.34916087 -4.609796896 -3.7646895</span></span>
<span><span class="co">## 10 0.06200 1.662500e-02 3441.277 0.9997915 -0.53590282  0.658860646  1.8849363</span></span>
<span><span class="co">## 11 0.02400 3.125000e-03 4607.828 0.9997555 -5.94522327 -5.163299329 -4.2519093</span></span>
<span><span class="co">## 12 0.06475 2.032692e-02 4275.527 0.9998338 -1.05236604  0.366851489  1.8492453</span></span>
<span><span class="co">## 13 0.00000 0.000000e+00 6071.679 0.9997540 -3.66253473 -3.058437865 -2.3011414</span></span>
<span><span class="co">## 14 0.00525 1.916667e-03 6143.807 0.9998453 -0.85519079  0.214062311  1.2616965</span></span>
<span><span class="co">## 15 0.79900 9.189286e-02 5463.801 1.0001771 -5.70470299 -4.891802690 -4.0023215</span></span>
<span><span class="co">## 16 0.00000 0.000000e+00 5290.990 0.9997525 -0.62593856  0.551654045  1.7729957</span></span>
<span><span class="co">## 17 0.00975 1.071429e-03 4077.818 1.0000116 -4.75986422 -4.054367972 -3.2433195</span></span>
<span><span class="co">## 18 0.07600 2.430357e-02 5845.838 0.9997503  0.03494679  1.080948068  2.1546025</span></span>
<span><span class="co">## 19 0.54775 6.570370e-02 4124.336 1.0010147 -3.68379962 -1.990610961 -0.6008783</span></span>
<span><span class="co">## 20 0.22325 6.528409e-02 4549.068 1.0003754 -1.59814988  0.506022553  2.6546100</span></span>
<span><span class="co">## 21 0.13250 2.537500e-02 5764.159 0.9999679 -4.73262516 -3.937001997 -2.5883987</span></span>
<span><span class="co">## 22 0.08025 2.803333e-02 4679.710 1.0005038 -4.09560122 -2.718240461 -1.3631696</span></span>
<span><span class="co">## 23 0.00025 2.500000e-05 4484.366 1.0000537 -4.16749641 -3.516552059 -2.7424955</span></span>
<span><span class="co">## 24 0.69250 1.622586e-01 4447.936 0.9998039 -0.99899018  0.024581340  1.1029183</span></span>
<span><span class="co">## 25 0.00000 0.000000e+00 4900.734 0.9999956 -4.83388876 -4.087305805 -3.2232430</span></span>
<span><span class="co">## 26 0.03425 8.575000e-03 4944.997 1.0002617 -1.07841296  0.084020792  1.3254906</span></span>
<span><span class="co">## 27 0.05575 1.012500e-02 4695.802 0.9997528 -5.84288261 -5.073883716 -4.1015296</span></span>
<span><span class="co">## 28 0.00500 1.250000e-03 6181.109 0.9998369 -0.73083224  0.381455569  1.5516070</span></span>
<span><span class="co">## 29 0.00000 0.000000e+00 4794.841 0.9998616 -4.97071341 -4.218965822 -3.3897791</span></span>
<span><span class="co">## 30 0.08400 3.153125e-02 5852.191 0.9999626 -0.78985839  0.307101068  1.5015152</span></span>
<span><span class="co">## 31 0.00050 8.333333e-05 4420.827 0.9997505 -6.96718324 -5.995970549 -4.8867844</span></span>
<span><span class="co">## 32 0.02375 5.722222e-03 4398.675 0.9997629 -1.46798941 -0.023714439  1.4230765</span></span>
<span><span class="co">## 33 0.00000 0.000000e+00 5036.154 0.9997597 -6.58556042 -5.746475667 -4.7653395</span></span>
<span><span class="co">## 34 0.60050 1.433214e-01 4726.974 0.9998012 -2.16881309 -0.767202452  0.5776295</span></span>
<span><span class="co">## 35 0.02800 4.588235e-03 6131.530 0.9998128 -5.23848922 -4.498184720 -3.5658850</span></span>
<span><span class="co">## 36 0.00875 3.468750e-03 2163.545 0.9998492 -5.29633121 -2.364370532  0.7140195</span></span>
<span><span class="co">## 37 0.81775 1.169224e-01 3997.214 0.9997986 -5.39818091 -4.676946527 -3.8455683</span></span>
<span><span class="co">## 38 0.08850 3.488235e-02 3636.555 0.9999503 -0.18847451  1.221428727  2.8182485</span></span>
<span><span class="co">## 39 0.00000 0.000000e+00 3948.267 0.9998234 -4.33433591 -3.597237869 -2.7510414</span></span>
<span><span class="co">## 40 0.25975 8.087500e-02 5934.663 0.9997950  0.32546689  1.446996448  2.5359410</span></span>
<span><span class="co">## 41 0.00000 0.000000e+00 4848.210 1.0002095 -4.72084080 -4.006002100 -3.1893655</span></span>
<span><span class="co">## 42 0.71850 1.808000e-01 5195.839 0.9999028 -0.84551327  0.256530203  1.4361228</span></span>
<span><span class="co">## 43 0.82525 1.405333e-01 4541.975 0.9998352 -5.58370440 -4.821603586 -3.9528164</span></span>
<span><span class="co">## 44 0.20725 5.776190e-02 5070.041 0.9997580 -0.80084691  0.363930049  1.6019834</span></span>
<span><span class="co">## 45 0.00000 0.000000e+00 4471.975 1.0002373 -4.53053116 -3.818570530 -3.0310863</span></span>
<span><span class="co">## 46 0.11650 4.339474e-02 6161.884 1.0002674 -0.53002267  0.623164715  1.7919097</span></span>
<span><span class="co">## 47 0.00025 4.545455e-05 5679.186 0.9998315 -6.61539417 -5.791058020 -4.8351112</span></span>
<span><span class="co">## 48 0.51625 1.086346e-01 4728.248 0.9998761 -3.27999119 -1.596454357 -0.1158578</span></span>
<span><span class="co">## 49 0.00000 0.000000e+00 4640.230 0.9999547 -6.41840677 -5.557094623 -4.5064191</span></span>
<span><span class="co">## 50 0.58800 1.263889e-01 3152.123 1.0011501 -0.70677066  0.681787175  2.3029439</span></span>
<span><span class="co">## 51 0.08700 1.626136e-02 5493.202 0.9998262 -5.48066634 -4.762275273 -3.8736963</span></span>
<span><span class="co">## 52 0.00750 2.714286e-03 2655.027 0.9997546 -2.85291034 -0.844922308  0.9554046</span></span>
<span><span class="co">## 53 0.01100 1.733333e-03 4692.063 0.9998629 -6.49971938 -5.581845190 -4.6251623</span></span>
<span><span class="co">## 54 0.05175 1.250000e-02 5162.244 1.0000080  0.14613606  1.345473598  2.6501132</span></span>
<span><span class="co">## 55 0.06825 1.289286e-02 5213.015 0.9997910 -5.30755563 -4.557489802 -3.6217868</span></span>
<span><span class="co">## 56 0.24500 7.309783e-02 4536.369 0.9997606 -1.22267568 -0.105757513  1.0889201</span></span>
<span><span class="co">## 57 0.34375 4.716346e-02 5083.723 1.0000189 -5.40622759 -4.629345298 -3.6948889</span></span>
<span><span class="co">## 58 0.36725 9.233000e-02 4676.448 0.9997855 -1.77423497 -0.632074061  0.5585467</span></span>
<span><span class="co">## 59 0.03750 7.723684e-03 4865.020 0.9997543 -5.79788651 -5.047521830 -4.1262999</span></span>
<span><span class="co">## 60 0.11500 3.933333e-02 4877.551 0.9997517 -1.09883647 -0.004138393  1.1670847</span></span>
<span><span class="co">##      v_pH0        v_FDR  v_n_eff v_R_k_hat   count_data</span></span>
<span><span class="co">## 1  0.00000 0.0000000000 4648.568 0.9999651 c("10x_6....</span></span>
<span><span class="co">## 2  0.30750 0.1440694444 4983.104 0.9998813 c("10x_6....</span></span>
<span><span class="co">## 3  0.00000 0.0000000000 3588.356 0.9998841 c("10x_6....</span></span>
<span><span class="co">## 4  0.09875 0.0530833333 4507.871 0.9997503 c("10x_6....</span></span>
<span><span class="co">## 5  0.00000 0.0000000000 5773.108 1.0000837 c("10x_6....</span></span>
<span><span class="co">## 6  0.11675 0.0594500000 5255.133 0.9998035 c("10x_6....</span></span>
<span><span class="co">## 7  0.00000 0.0000000000 8330.805 0.9997608 c("10x_6....</span></span>
<span><span class="co">## 8  0.09150 0.0473750000 4558.335 0.9997525 c("10x_6....</span></span>
<span><span class="co">## 9  0.00000 0.0000000000 7273.853 0.9997801 c("10x_6....</span></span>
<span><span class="co">## 10 0.23000 0.0970384615 4457.578 0.9997640 c("10x_6....</span></span>
<span><span class="co">## 11 0.00000 0.0000000000 6844.765 0.9997584 c("10x_6....</span></span>
<span><span class="co">## 12 0.41025 0.1897386364 4186.488 0.9999649 c("10x_6....</span></span>
<span><span class="co">## 13 0.00000 0.0000000000 5044.457 0.9997574 c("10x_6....</span></span>
<span><span class="co">## 14 0.48450 0.2220700000 5008.526 0.9997728 c("10x_6....</span></span>
<span><span class="co">## 15 0.00000 0.0000000000 4821.616 0.9998303 c("10x_6....</span></span>
<span><span class="co">## 16 0.28000 0.1344558824 4963.327 1.0009502 c("10x_6....</span></span>
<span><span class="co">## 17 0.00000 0.0000000000 6571.326 0.9997509 c("10x_6....</span></span>
<span><span class="co">## 18 0.05350 0.0256000000 5022.178 0.9997558 c("10x_6....</span></span>
<span><span class="co">## 19 0.00350 0.0001166667 3326.859 1.0010595 c("10x_6....</span></span>
<span><span class="co">## 20 0.39550 0.1683250000 3852.842 0.9997892 c("10x_6....</span></span>
<span><span class="co">## 21 0.00000 0.0000000000 3258.397 1.0000369 c("10x_6....</span></span>
<span><span class="co">## 22 0.00000 0.0000000000 3924.869 0.9997989 c("10x_6....</span></span>
<span><span class="co">## 23 0.00000 0.0000000000 6059.154 0.9998790 c("10x_6....</span></span>
<span><span class="co">## 24 0.63600 0.2848500000 5303.628 0.9999280 c("10x_6....</span></span>
<span><span class="co">## 25 0.00000 0.0000000000 5922.819 1.0001866 c("10x_6....</span></span>
<span><span class="co">## 26 0.57375 0.2476944444 5102.646 0.9997594 c("10x_6....</span></span>
<span><span class="co">## 27 0.00000 0.0000000000 6514.562 0.9997677 c("10x_6....</span></span>
<span><span class="co">## 28 0.37775 0.1563684211 5186.102 0.9997747 c("10x_6....</span></span>
<span><span class="co">## 29 0.00000 0.0000000000 6381.840 0.9997632 c("10x_6....</span></span>
<span><span class="co">## 30 0.43000 0.2001847826 5657.069 0.9997626 c("10x_6....</span></span>
<span><span class="co">## 31 0.00000 0.0000000000 5146.706 0.9997500 c("10x_6....</span></span>
<span><span class="co">## 32 0.59325 0.2600357143 4149.512 0.9997535 c("10x_6....</span></span>
<span><span class="co">## 33 0.00000 0.0000000000 5654.691 0.9998075 c("10x_6....</span></span>
<span><span class="co">## 34 0.20825 0.0729772727 4847.805 0.9999185 c("10x_6....</span></span>
<span><span class="co">## 35 0.00000 0.0000000000 6011.824 0.9998174 c("10x_6....</span></span>
<span><span class="co">## 36 0.07825 0.0343750000 2792.663 0.9998129 c("10x_6....</span></span>
<span><span class="co">## 37 0.00000 0.0000000000 6729.472 0.9999330 c("10x_6....</span></span>
<span><span class="co">## 38 0.08125 0.0410714286 3585.441 1.0000874 c("10x_6....</span></span>
<span><span class="co">## 39 0.00000 0.0000000000 5963.930 0.9997508 c("10x_6....</span></span>
<span><span class="co">## 40 0.01375 0.0068750000 4667.244 0.9997609 c("10x_6....</span></span>
<span><span class="co">## 41 0.00000 0.0000000000 5398.340 0.9997738 c("10x_6....</span></span>
<span><span class="co">## 42 0.46300 0.2111354167 5443.585 0.9999238 c("10x_6....</span></span>
<span><span class="co">## 43 0.00000 0.0000000000 6012.038 1.0001387 c("10x_6....</span></span>
<span><span class="co">## 44 0.39750 0.1792380952 4432.890 1.0003462 c("10x_6....</span></span>
<span><span class="co">## 45 0.00000 0.0000000000 6870.258 1.0000855 c("10x_6....</span></span>
<span><span class="co">## 46 0.22875 0.0859583333 5766.389 1.0007165 c("10x_6....</span></span>
<span><span class="co">## 47 0.00000 0.0000000000 4602.109 0.9998410 c("10x_6....</span></span>
<span><span class="co">## 48 0.03175 0.0186250000 5650.538 1.0002573 c("10x_6....</span></span>
<span><span class="co">## 49 0.00000 0.0000000000 6826.189 0.9997678 c("10x_6....</span></span>
<span><span class="co">## 50 0.25500 0.1253593750 3496.478 1.0009728 c("10x_6....</span></span>
<span><span class="co">## 51 0.00000 0.0000000000 6932.251 1.0009846 c("10x_6....</span></span>
<span><span class="co">## 52 0.24975 0.1167166667 3342.113 0.9997574 c("10x_6....</span></span>
<span><span class="co">## 53 0.00000 0.0000000000 6083.848 0.9997505 c("10x_6....</span></span>
<span><span class="co">## 54 0.02900 0.0142500000 5115.012 0.9997707 c("10x_6....</span></span>
<span><span class="co">## 55 0.00000 0.0000000000 5381.725 0.9997732 c("10x_6....</span></span>
<span><span class="co">## 56 0.56225 0.2351538462 3914.919 0.9997939 c("10x_6....</span></span>
<span><span class="co">## 57 0.00000 0.0000000000 5368.806 1.0001335 c("10x_6....</span></span>
<span><span class="co">## 58 0.23950 0.1072142857 4835.618 1.0001678 c("10x_6....</span></span>
<span><span class="co">## 59 0.00000 0.0000000000 5053.551 0.9997529 c("10x_6....</span></span>
<span><span class="co">## 60 0.62850 0.2727413793 5158.118 0.9997621 c("10x_6....</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_1D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_1D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<p>Plot 2D significance plot. Data points are cell groups. Error bars
are the 95% credible interval. The dashed lines represent the default
threshold fold change for which the probabilities (c_pH0, v_pH0) are
calculated. pH0 of 0 represent the rejection of the null hypothesis that
no effect is observed.</p>
<p>This plot is provided only if differential variability has been
tested. The differential variability estimates are reliable only if the
linear association between mean and variability for
<code>(intercept)</code> (left-hand side facet) is satisfied. A
scatterplot (besides the Intercept) is provided for each category of
interest. The for each category of interest, the composition and
variability effects should be generally uncorrelated.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">$</span><span class="va">credible_intervals_2D</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Unknown or uninitialised column: `credible_intervals_2D`.</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2023-11-22 r85609)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rstan_2.32.3       StanHeaders_2.32.2 tidyr_1.3.0        forcats_1.0.0     </span></span>
<span><span class="co">## [5] ggplot2_3.4.4      sccomp_1.7.2       dplyr_1.1.4        BiocStyle_2.31.0  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-7                gridExtra_2.3              </span></span>
<span><span class="co">##   [3] inline_0.3.19               rlang_1.1.2                </span></span>
<span><span class="co">##   [5] magrittr_2.0.3              matrixStats_1.1.0          </span></span>
<span><span class="co">##   [7] compiler_4.4.0              loo_2.6.0                  </span></span>
<span><span class="co">##   [9] systemfonts_1.0.5           callr_3.7.3                </span></span>
<span><span class="co">##  [11] vctrs_0.6.5                 stringr_1.5.1              </span></span>
<span><span class="co">##  [13] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">##  [15] fastmap_1.1.1               XVector_0.43.0             </span></span>
<span><span class="co">##  [17] labeling_0.4.3              utf8_1.2.4                 </span></span>
<span><span class="co">##  [19] rmarkdown_2.25              tzdb_0.4.0                 </span></span>
<span><span class="co">##  [21] ps_1.7.5                    ragg_1.2.6                 </span></span>
<span><span class="co">##  [23] purrr_1.0.2                 xfun_0.41                  </span></span>
<span><span class="co">##  [25] zlibbioc_1.49.0             cachem_1.0.8               </span></span>
<span><span class="co">##  [27] GenomeInfoDb_1.39.1         jsonlite_1.8.8             </span></span>
<span><span class="co">##  [29] highr_0.10                  DelayedArray_0.29.0        </span></span>
<span><span class="co">##  [31] parallel_4.4.0              prettyunits_1.2.0          </span></span>
<span><span class="co">##  [33] R6_2.5.1                    bslib_0.6.1                </span></span>
<span><span class="co">##  [35] stringi_1.8.2               parallelly_1.36.0          </span></span>
<span><span class="co">##  [37] boot_1.3-28.1               GenomicRanges_1.55.1       </span></span>
<span><span class="co">##  [39] jquerylib_0.1.4             Rcpp_1.0.11                </span></span>
<span><span class="co">##  [41] bookdown_0.37               SummarizedExperiment_1.33.1</span></span>
<span><span class="co">##  [43] knitr_1.45                  future.apply_1.11.0        </span></span>
<span><span class="co">##  [45] readr_2.1.4                 IRanges_2.37.0             </span></span>
<span><span class="co">##  [47] Matrix_1.6-4                tidyselect_1.2.0           </span></span>
<span><span class="co">##  [49] abind_1.4-5                 yaml_2.3.7                 </span></span>
<span><span class="co">##  [51] codetools_0.2-19            curl_5.1.0                 </span></span>
<span><span class="co">##  [53] processx_3.8.2              listenv_0.9.0              </span></span>
<span><span class="co">##  [55] pkgbuild_1.4.2              lattice_0.22-5             </span></span>
<span><span class="co">##  [57] tibble_3.2.1                withr_2.5.2                </span></span>
<span><span class="co">##  [59] Biobase_2.63.0              evaluate_0.23              </span></span>
<span><span class="co">##  [61] future_1.33.0               desc_1.4.2                 </span></span>
<span><span class="co">##  [63] RcppParallel_5.1.7          pillar_1.9.0               </span></span>
<span><span class="co">##  [65] BiocManager_1.30.22         MatrixGenerics_1.15.0      </span></span>
<span><span class="co">##  [67] stats4_4.4.0                generics_0.1.3             </span></span>
<span><span class="co">##  [69] rprojroot_2.0.4             sp_2.1-2                   </span></span>
<span><span class="co">##  [71] RCurl_1.98-1.13             S4Vectors_0.41.2           </span></span>
<span><span class="co">##  [73] hms_1.1.3                   rstantools_2.3.1.1         </span></span>
<span><span class="co">##  [75] munsell_0.5.0               scales_1.3.0               </span></span>
<span><span class="co">##  [77] globals_0.16.2              glue_1.6.2                 </span></span>
<span><span class="co">##  [79] tools_4.4.0                 fs_1.6.3                   </span></span>
<span><span class="co">##  [81] dotCall64_1.1-1             grid_4.4.0                 </span></span>
<span><span class="co">##  [83] QuickJSR_1.0.8              colorspace_2.1-0           </span></span>
<span><span class="co">##  [85] SingleCellExperiment_1.25.0 GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [87] patchwork_1.1.3             cli_3.6.1                  </span></span>
<span><span class="co">##  [89] textshaping_0.3.7           spam_2.10-0                </span></span>
<span><span class="co">##  [91] fansi_1.0.5                 S4Arrays_1.3.1             </span></span>
<span><span class="co">##  [93] V8_4.4.1                    gtable_0.3.4               </span></span>
<span><span class="co">##  [95] sass_0.4.8                  digest_0.6.33              </span></span>
<span><span class="co">##  [97] progressr_0.14.0            BiocGenerics_0.49.1        </span></span>
<span><span class="co">##  [99] SparseArray_1.3.1           ggrepel_0.9.4              </span></span>
<span><span class="co">## [101] farver_2.1.1                SeuratObject_5.0.1         </span></span>
<span><span class="co">## [103] memoise_2.0.1               htmltools_0.5.7            </span></span>
<span><span class="co">## [105] pkgdown_2.0.7               lifecycle_1.0.4</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
