<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>New sccomp framework • sccomp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="New sccomp framework">
<meta property="og:description" content="sccomp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sccomp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the sccomp package</a>
    </li>
    <li>
      <a href="../articles/new_framework.html">New sccomp framework</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/sccomp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>New sccomp framework</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2024-03-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/sccomp/blob/HEAD/vignettes/new_framework.Rmd" class="external-link"><code>vignettes/new_framework.Rmd</code></a></small>
      <div class="hidden name"><code>new_framework.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidyseurat/actions/" class="external-link"><img src="https://github.com/stemangiola/tidyseurat/workflows/R-CMD-check/badge.svg" alt="R build status"></a></p>
<!-- badges: end -->
<p><img src="inst/logo-01.png" height="139px" width="120px"></p>
<p>We announce the new tidy and modular interface for a
<code>sccomp</code>, which improves modularity, and clarity. The main
change is the modularisation of <code>sccomp</code> in functions which
can be linked with the pipe operator <code>|&gt;</code>.</p>
<table class="table">
<colgroup>
<col width="48%">
<col width="51%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Estimation: <code>sccomp_stimate()</code>
</td>
<td>which is usually run once in the analysis (per model).</td>
</tr>
<tr class="even">
<td>Testing: <code><a href="../reference/sccomp_test.html">sccomp_test()</a></code>
</td>
<td>which candy run multiple times, depending on how many contrasts you
want to test (e.g. age, untreated vs treated).</td>
</tr>
<tr class="odd">
<td>Outlier removal: <code><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers()</a></code>
</td>
<td>which is usually run once after <code><a href="../reference/sccomp_estimate.html">sccomp_estimate()</a></code> in
case you want to produce estimates not influenced by outlier data
points.</td>
</tr>
<tr class="even">
<td>Unwanted variation removal:
<code><a href="../reference/sccomp_remove_unwanted_variation.html">sccomp_remove_unwanted_variation()</a></code>
</td>
<td>which is run after <code><a href="../reference/sccomp_estimate.html">sccomp_estimate()</a></code> and produces a
dataset that just preserve the variability of your factor of
interest.</td>
</tr>
<tr class="odd">
<td>Data replication: <code><a href="../reference/sccomp_replicate.html">sccomp_replicate()</a></code>
</td>
<td>which is run after <code><a href="../reference/sccomp_estimate.html">sccomp_estimate()</a></code> and produces a
dataset representing the theoretical data distribution according to the
model (from the posterior distribution).</td>
</tr>
<tr class="even">
<td>Plotting: <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code>
</td>
<td>which is run after <code>sccomp_test</code> and outputs a series of
summary plots.</td>
</tr>
</tbody>
</table>
<div class="section level2">
<h2 id="a-reminder-what-is-sccomp">A reminder: what is <code>sccomp</code><a class="anchor" aria-label="anchor" href="#a-reminder-what-is-sccomp"></a>
</h2>
<p><code>sccomp</code> is a statistical model developed for differential
variability analysis in compositional data, primarily used in cellular
omics fields like single-cell genomics, proteomics, and microbiomics
(Mangiola et al. 2023). It addresses limitations of existing methods in
differential abundance analysis by incorporating several advanced
features. <code>sccomp</code> effectively models compositional count
data properties, which were previously not adequately addressed, and
tackles cell-group-specific differential variability. This model uses a
constrained Beta-binomial distribution to enable more precise analyses.
Key capabilities of <code>sccomp</code> include improved differential
abundance analyses through cross-sample information borrowing, outlier
identification and exclusion, realistic data simulation, and
facilitating cross-study knowledge transfer. By incorporating these
features, <code>sccomp</code> provides a more comprehensive and accurate
framework for analyzing cellular omics data, identifying crucial
biological drivers such as disease progression markers in cancer and
pathogen infection.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sccomp"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/sccomp"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="deprecation-of-the-function-sccomp_glm">Deprecation of the function <code>sccomp_glm()</code><a class="anchor" aria-label="anchor" href="#deprecation-of-the-function-sccomp_glm"></a>
</h2>
<p>The new framework</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outlier_free_estimate</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Estimate</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_estimate.html">sccomp_estimate</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="va">continuous_covariate</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Remove outliers</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_outliers.html">sccomp_remove_outliers</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000397 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.97 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.369 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                17.517 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                19.886 seconds (Total)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000296 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 2.96 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:     1 / 20299 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:   301 / 20299 [  1%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  1300 / 20299 [  6%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  2300 / 20299 [ 11%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  3300 / 20299 [ 16%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  4300 / 20299 [ 21%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  5300 / 20299 [ 26%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  6300 / 20299 [ 31%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  7300 / 20299 [ 35%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  8300 / 20299 [ 40%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration:  9300 / 20299 [ 45%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 10300 / 20299 [ 50%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 11300 / 20299 [ 55%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 12300 / 20299 [ 60%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 13300 / 20299 [ 65%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 14300 / 20299 [ 70%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 15300 / 20299 [ 75%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 16300 / 20299 [ 80%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 17300 / 20299 [ 85%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 18300 / 20299 [ 90%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 19300 / 20299 [ 95%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 20299 / 20299 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.309 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                84.624 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                86.933 seconds (Total)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SAMPLING FOR MODEL 'glm_multi_beta_binomial' NOW (CHAIN 1).</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Gradient evaluation took 0.000346 seconds</span></span>
<span><span class="co">## Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 3.46 seconds.</span></span>
<span><span class="co">## Chain 1: Adjust your expectations accordingly!</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1: Iteration:    1 / 4300 [  0%]  (Warmup)</span></span>
<span><span class="co">## Chain 1: Iteration:  301 / 4300 [  7%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 1300 / 4300 [ 30%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 2300 / 4300 [ 53%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 3300 / 4300 [ 76%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: Iteration: 4300 / 4300 [100%]  (Sampling)</span></span>
<span><span class="co">## Chain 1: </span></span>
<span><span class="co">## Chain 1:  Elapsed Time: 2.28 seconds (Warm-up)</span></span>
<span><span class="co">## Chain 1:                17.127 seconds (Sampling)</span></span>
<span><span class="co">## Chain 1:                19.407 seconds (Total)</span></span>
<span><span class="co">## Chain 1:</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test</span></span>
<span><span class="va">outlier_free_estimate</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="st">"typehealthy"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 30 × 18</span></span></span>
<span><span class="co">##    cell_group  parameter factor c_lower c_effect c_upper   c_pH0   c_FDR c_n_eff</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B immature  typeheal… type     0.901    1.41   1.92   0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">6</span>166.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B mem       typeheal… type     1.11     1.72   2.34   0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">5</span>384.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> CD4 cm S10… typeheal… type     0.584    0.994  1.41   2.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 5.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>   <span style="text-decoration: underline;">6</span>685.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> CD4 cm hig… typeheal… type    -<span style="color: #BB0000;">3.04</span>    -<span style="color: #BB0000;">1.97</span>  -<span style="color: #BB0000;">1.00</span>   0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">4</span>120.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CD4 cm rib… typeheal… type    -<span style="color: #BB0000;">1.76</span>    -<span style="color: #BB0000;">1.05</span>  -<span style="color: #BB0000;">0.358</span>  7.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 1.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>   <span style="text-decoration: underline;">4</span>164.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> CD4 em hig… typeheal… type    -<span style="color: #BB0000;">2.23</span>    -<span style="color: #BB0000;">1.38</span>  -<span style="color: #BB0000;">0.622</span>  2.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 4.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span>   <span style="text-decoration: underline;">4</span>119.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> CD4 naive   typeheal… type     0.173    0.827  1.43   3.00<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 9.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span>   <span style="text-decoration: underline;">5</span>723.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> CD4 riboso… typeheal… type     1.54     2.04   2.54   0   <span style="color: #949494;"> </span>   0   <span style="color: #949494;"> </span>     <span style="text-decoration: underline;">4</span>631.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> CD8 em 1    typeheal… type    -<span style="color: #BB0000;">0.543</span>    0.123  0.690  6.03<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 1.25<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span>   <span style="text-decoration: underline;">7</span>330.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> CD8 em 2    typeheal… type    -<span style="color: #BB0000;">2.19</span>    -<span style="color: #BB0000;">0.985</span>  0.012<span style="text-decoration: underline;">9</span> 6.25<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 1.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span>   <span style="text-decoration: underline;">5</span>797.</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 20 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 9 more variables: c_R_k_hat &lt;dbl&gt;, v_lower &lt;dbl&gt;, v_effect &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   v_upper &lt;dbl&gt;, v_pH0 &lt;dbl&gt;, v_FDR &lt;dbl&gt;, v_n_eff &lt;dbl&gt;, v_R_k_hat &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   count_data &lt;list&gt;</span></span></span></code></pre>
<p>Replaces the old framework (that now will receive a deprecation
warning)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Estimate</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_glm.html">sccomp_glm</a></span><span class="op">(</span> </span>
<span>    formula_composition <span class="op">=</span> <span class="op">~</span> <span class="va">type</span> <span class="op">+</span> <span class="va">continuous_covariate</span>, </span>
<span>    .sample <span class="op">=</span>  <span class="va">sample</span>, </span>
<span>    .cell_group <span class="op">=</span> <span class="va">cell_group</span>, </span>
<span>    check_outliers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    contrasts <span class="op">=</span> <span class="st">"typehealthy"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="new-functionalities">New functionalities<a class="anchor" aria-label="anchor" href="#new-functionalities"></a>
</h2>
<div class="section level3">
<h3 id="removal-of-unwanted-variation-">Removal of unwanted variation.<a class="anchor" aria-label="anchor" href="#removal-of-unwanted-variation-"></a>
</h3>
<p>For <strong>visualisation</strong> purposes, we can select factor of
interest we would like to preserve the effect for, end exclude all the
rest. For example, if we want to produce a dataset with just the
<code>type</code> effect, we can execute</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outlier_free_estimate</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_remove_unwanted_variation.html">sccomp_remove_unwanted_variation</a></span><span class="op">(</span><span class="op">~</span> <span class="va">type</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## sccomp says: calculating residuals</span></span></code></pre>
<pre><code><span><span class="co">## sccomp says: regressing out unwanted factors</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 600 × 5</span></span></span>
<span><span class="co">##    sample       cell_group adjusted_proportion adjusted_counts logit_residuals</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 10x_6K       B immature              0.054<span style="text-decoration: underline;">8</span>           257.          -<span style="color: #BB0000;">0.748</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 10x_8K       B immature              0.143           <span style="text-decoration: underline;">1</span>073.           0.328 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> GSE115189    B immature              0.113            264.           0.028<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> SCP345_580   B immature              0.089<span style="text-decoration: underline;">9</span>           518.          -<span style="color: #BB0000;">0.198</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> SCP345_860   B immature              0.151            967.           0.385 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> SCP424_pbmc1 B immature              0.113            302.          -<span style="color: #BB0000;">0.017</span><span style="color: #BB0000; text-decoration: underline;">0</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> SCP424_pbmc2 B immature              0.201            600.           0.721 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> SCP591       B immature              0.024<span style="text-decoration: underline;">7</span>            14.1         -<span style="color: #BB0000;">1.56</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> SI-GA-E5     B immature              0.024<span style="text-decoration: underline;">1</span>           101.          -<span style="color: #BB0000;">0.708</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> SI-GA-E7     B immature              0.097<span style="text-decoration: underline;">4</span>           715.           0.777 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 590 more rows</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<p>The bloating functionalities have been improved. Now, both discrete
and continuous variables can be visualised overlaying the to reticle
data distribution from the model. This helps the user understanding
whether the model is descriptively adequate to the data.</p>
<p>For example, if the theoretical data distribution from the
<code>sccomp</code> does not overlap with the observed data
distribution, this is an indication that the probability distribution
used by <code>sccomp</code> is not suitable for the data or a different
model (design matrix) should be used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outlier_free_estimate</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="st">"typehealthy"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $boxplot</span></span>
<span><span class="co">## $boxplot[[1]]</span></span></code></pre>
<p><img src="new_framework_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $credible_intervals_1D</span></span></code></pre>
<p><img src="new_framework_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<p>Now plotting the test against the continuous covariate</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outlier_free_estimate</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/sccomp_test.html">sccomp_test</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="st">"continuous_covariate"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $boxplot</span></span>
<span><span class="co">## $boxplot[[1]]</span></span></code></pre>
<p><img src="new_framework_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $credible_intervals_1D</span></span></code></pre>
<p><img src="new_framework_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-02-28 r85999)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] sccomp_1.7.6     BiocStyle_2.31.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] gridExtra_2.3               inline_0.3.19              </span></span>
<span><span class="co">##   [3] rlang_1.1.3                 magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] matrixStats_1.2.0           compiler_4.4.0             </span></span>
<span><span class="co">##   [7] mgcv_1.9-1                  loo_2.7.0                  </span></span>
<span><span class="co">##   [9] systemfonts_1.0.5           vctrs_0.6.5                </span></span>
<span><span class="co">##  [11] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [13] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">##  [15] XVector_0.43.1              labeling_0.4.3             </span></span>
<span><span class="co">##  [17] utf8_1.2.4                  rmarkdown_2.26             </span></span>
<span><span class="co">##  [19] tzdb_0.4.0                  ragg_1.2.7                 </span></span>
<span><span class="co">##  [21] purrr_1.0.2                 xfun_0.42                  </span></span>
<span><span class="co">##  [23] zlibbioc_1.49.0             cachem_1.0.8               </span></span>
<span><span class="co">##  [25] GenomeInfoDb_1.39.7         jsonlite_1.8.8             </span></span>
<span><span class="co">##  [27] highr_0.10                  DelayedArray_0.29.9        </span></span>
<span><span class="co">##  [29] parallel_4.4.0              R6_2.5.1                   </span></span>
<span><span class="co">##  [31] RColorBrewer_1.1-3          bslib_0.6.1                </span></span>
<span><span class="co">##  [33] stringi_1.8.3               StanHeaders_2.32.6         </span></span>
<span><span class="co">##  [35] parallelly_1.37.1           boot_1.3-30                </span></span>
<span><span class="co">##  [37] GenomicRanges_1.55.3        jquerylib_0.1.4            </span></span>
<span><span class="co">##  [39] Rcpp_1.0.12                 bookdown_0.38              </span></span>
<span><span class="co">##  [41] rstan_2.32.6                SummarizedExperiment_1.33.3</span></span>
<span><span class="co">##  [43] knitr_1.45                  future.apply_1.11.1        </span></span>
<span><span class="co">##  [45] readr_2.1.5                 IRanges_2.37.1             </span></span>
<span><span class="co">##  [47] splines_4.4.0               Matrix_1.6-5               </span></span>
<span><span class="co">##  [49] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">##  [51] yaml_2.3.8                  codetools_0.2-19           </span></span>
<span><span class="co">##  [53] curl_5.2.1                  listenv_0.9.1              </span></span>
<span><span class="co">##  [55] pkgbuild_1.4.3              lattice_0.22-5             </span></span>
<span><span class="co">##  [57] tibble_3.2.1                withr_3.0.0                </span></span>
<span><span class="co">##  [59] Biobase_2.63.0              evaluate_0.23              </span></span>
<span><span class="co">##  [61] future_1.33.1               desc_1.4.3                 </span></span>
<span><span class="co">##  [63] RcppParallel_5.1.7          pillar_1.9.0               </span></span>
<span><span class="co">##  [65] BiocManager_1.30.22         MatrixGenerics_1.15.0      </span></span>
<span><span class="co">##  [67] stats4_4.4.0                generics_0.1.3             </span></span>
<span><span class="co">##  [69] sp_2.1-3                    S4Vectors_0.41.4           </span></span>
<span><span class="co">##  [71] hms_1.1.3                   ggplot2_3.5.0              </span></span>
<span><span class="co">##  [73] rstantools_2.4.0            munsell_0.5.0              </span></span>
<span><span class="co">##  [75] scales_1.3.0                globals_0.16.2             </span></span>
<span><span class="co">##  [77] glue_1.7.0                  tools_4.4.0                </span></span>
<span><span class="co">##  [79] forcats_1.0.0               fs_1.6.3                   </span></span>
<span><span class="co">##  [81] dotCall64_1.1-1             grid_4.4.0                 </span></span>
<span><span class="co">##  [83] tidyr_1.3.1                 QuickJSR_1.1.3             </span></span>
<span><span class="co">##  [85] colorspace_2.1-0            SingleCellExperiment_1.25.0</span></span>
<span><span class="co">##  [87] nlme_3.1-164                GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [89] patchwork_1.2.0             cli_3.6.2                  </span></span>
<span><span class="co">##  [91] textshaping_0.3.7           spam_2.10-0                </span></span>
<span><span class="co">##  [93] fansi_1.0.6                 S4Arrays_1.3.6             </span></span>
<span><span class="co">##  [95] dplyr_1.1.4                 V8_4.4.2                   </span></span>
<span><span class="co">##  [97] gtable_0.3.4                sass_0.4.8                 </span></span>
<span><span class="co">##  [99] digest_0.6.34               progressr_0.14.0           </span></span>
<span><span class="co">## [101] BiocGenerics_0.49.1         SparseArray_1.3.4          </span></span>
<span><span class="co">## [103] ggrepel_0.9.5               farver_2.1.1               </span></span>
<span><span class="co">## [105] SeuratObject_5.0.1          memoise_2.0.1              </span></span>
<span><span class="co">## [107] htmltools_0.5.7             pkgdown_2.0.7              </span></span>
<span><span class="co">## [109] lifecycle_1.0.4</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
